From e2552950965d5e8a7589de99a2581b515fccbe55 Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Tue, 10 Oct 2023 11:16:34 +0530
Subject: [PATCH] mac80211: fix mesh checksum WARN ON during iperf traffic

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 net/mac80211/mesh.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/net/mac80211/mesh.c b/net/mac80211/mesh.c
index 44c1a24..ec37528 100644
--- a/net/mac80211/mesh.c
+++ b/net/mac80211/mesh.c
@@ -13,6 +13,8 @@
 #include "wme.h"
 #include "driver-ops.h"
 
+#define IS_HW_CSUM_NOT_ENABLED(dev)             (!((dev)->features & NETIF_F_HW_CSUM))
+
 static int mesh_allocated;
 static struct kmem_cache *rm_cache;
 
@@ -822,7 +824,7 @@ bool ieee80211_mesh_xmit_fast(struct ieee80211_sub_if_data *sdata,
 	if (skb->sk && skb_shinfo(skb)->tx_flags & SKBTX_WIFI_STATUS)
 		return false;
 
-	if (skb->ip_summed == CHECKSUM_PARTIAL) {
+	if (skb->ip_summed == CHECKSUM_PARTIAL && IS_HW_CSUM_NOT_ENABLED(sdata->dev)) {
 		skb_set_transport_header(skb, skb_checksum_start_offset(skb));
 		if (skb_checksum_help(skb))
 			return false;
-- 
2.17.1

