Index: backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/core.c
===================================================================
--- backports-20190412-4.4.60-9bcbbf4cedb7.orig/drivers/net/wireless/ath/ath11k/core.c
+++ backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/core.c
@@ -486,6 +486,7 @@ err_mac_destroy:
 static void ath11k_core_pdev_destroy(struct ath11k_base *sc)
 {
 	ath11k_mac_unregister(sc);
+	ath11k_ahb_ext_irq_disable(sc);
 	ath11k_dp_pdev_free(sc);
 }
 
@@ -783,7 +784,6 @@ void ath11k_core_deinit(struct ath11k_ba
 {
 	mutex_lock(&sc->core_lock);
 
-	ath11k_ahb_ext_irq_disable(sc);
 	ath11k_core_pdev_destroy(sc);
 	ath11k_core_stop(sc);
 
