--- a/drivers/net/wireless/ath/ath11k/qmi.h
+++ b/drivers/net/wireless/ath/ath11k/qmi.h
@@ -36,8 +36,6 @@
 
 #define QMI_WLANFW_MAX_DATA_SIZE_V01		6144
 #define ATH11K_FIRMWARE_MODE_OFF		4
-#define Q6_CALDB_ADDR 0x50000000
-#define Q6_CALDB_SIZE 0x480000
 
 struct ath11k_base;
 enum ath11k_firmware_mode;
diff --git a/drivers/net/wireless/ath/ath11k/qmi.c b/drivers/net/wireless/ath/ath11k/qmi.c
index 046614808de2..e5c53ec95476 100644
--- a/drivers/net/wireless/ath/ath11k/qmi.c
+++ b/drivers/net/wireless/ath/ath11k/qmi.c
@@ -1640,12 +1640,6 @@ static int ath11k_qmi_respond_fw_mem_request(struct ath11k_base *ab)
 		goto out;
 
 	for (i = 0; i < req.mem_seg_len ; i++) {
-		if (!ab->qmi.target_mem[i].paddr || !ab->qmi.target_mem[i].size) {
-			ath11k_warn(ab, "qmi invalid mem request for target\n");
-			ret = -EINVAL;
-			goto out;
-		}
-
 		req.mem_seg[i].addr = ab->qmi.target_mem[i].paddr;
 		req.mem_seg[i].size = ab->qmi.target_mem[i].size;
 		req.mem_seg[i].type = ab->qmi.target_mem[i].type;
@@ -1704,10 +1698,9 @@ static int ath11k_qmi_alloc_target_mem_chunk(struct ath11k_base *ab)
 				ath11k_warn(ab, "qmi mem size is low to load caldata\n");
 				return -EINVAL;
 			}
-			ab->qmi.target_mem[idx].paddr =
-					bdf_location[mode] + ATH11K_QMI_CALDATA_OFFSET;
-			ab->qmi.target_mem[idx].vaddr =
-					bdf_location[mode] + ATH11K_QMI_CALDATA_OFFSET;
+			/* TODO cold boot calibration mode is not supported */
+			ab->qmi.target_mem[idx].paddr = 0;
+			ab->qmi.target_mem[idx].vaddr = 0;
 			ab->qmi.target_mem[idx].size = ab->qmi.target_mem[i].size;
 			ab->qmi.target_mem[idx].type = ab->qmi.target_mem[i].type;
 			idx++;
