Index: backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/dp_tx.c
===================================================================
--- backports-20190412-4.4.60-9bcbbf4cedb7.orig/drivers/net/wireless/ath/ath11k/dp_tx.c	2019-04-22 17:18:42.165275404 +0530
+++ backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/dp_tx.c	2019-04-22 17:18:42.161275451 +0530
@@ -139,6 +139,10 @@
 		ath11k_dp_encap_nwifi(skb);
 		break;
 	case HAL_TCL_ENCAP_TYPE_RAW:
+		/*  TODO: for CHECKSUM_PARTIAL case in raw mode, HW checksum offload
+		 *	  is not applicable, hence manual checksum calculation using
+		 *	  skb_checksum_help() is needed
+		 */
 	case HAL_TCL_ENCAP_TYPE_ETHERNET:
 	case HAL_TCL_ENCAP_TYPE_802_3:
 		/* TODO: Take care of other encap modes as well */
Index: backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20190412-4.4.60-9bcbbf4cedb7.orig/drivers/net/wireless/ath/ath11k/mac.c	2019-04-22 17:18:42.165275404 +0530
+++ backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/mac.c	2019-04-22 17:18:42.161275451 +0530
@@ -4892,6 +4892,9 @@
 
 	ath11k_reg_init(ar);
 
+	/* advertise HW checksum offload capabilities */
+	ar->hw->netdev_features = NETIF_F_HW_CSUM;
+
 	ret = ieee80211_register_hw(ar->hw);
 	if (ret) {
 		ath11k_err(ar->ab, "ieee80211 registration failed: %d\n", ret);
