Index: backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20190412-4.4.60-9bcbbf4cedb7.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/mac.c
@@ -4221,10 +4221,16 @@ static int ath11k_set_rts_threshold(stru
 
 static int ath11k_set_frag_threshold(struct ieee80211_hw *hw, u32 value)
 {
-	struct ath11k *ar = hw->priv;
-	int param_id = WMI_VDEV_PARAM_FRAGMENTATION_THRESHOLD;
-
-	return ath11k_set_vdev_param_to_all_vifs(ar, param_id, value);
+	/* Even though there's a WMI vdev param for fragmentation threshold no
+	 * known firmware actually implements it. Moreover it is not possible to
+	 * rely frame fragmentation to mac80211 because firmware mangling the
+	 * wifi header making remote device drop the frames.
+	 *
+	 * Hence implement a dummy callback just to say fragmentation isn't
+	 * supported. This effectively prevents mac80211 from doing frame
+	 * fragmentation in software.
+	 */
+	return -EOPNOTSUPP;
 }
 
 static void ath11k_flush(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
