--- a/drivers/net/wireless/ath/ath11k/debug.h
+++ b/drivers/net/wireless/ath/ath11k/debug.h
@@ -102,6 +102,22 @@ void ath11k_debug_fw_stats_process(struct ath11k_base *ab, u8 *evt_buf,
 
 void ath11k_debug_fw_stats_init(struct ath11k *ar);
 
+static inline bool ath11k_debug_is_pktlog_lite_mode_enabled(struct ath11k *ar)
+{
+	return (ar->debug.pktlog_mode == ATH11K_PKTLOG_MODE_LITE);
+}
+
+static inline bool ath11k_debug_is_pktlog_per_peer_enabled(struct ath11k *ar)
+{
+	return !!ar->debug.pktlog_peer_valid;
+}
+
+static inline bool ath11k_debug_is_pktlog_peer_vaild(struct ath11k *ar, u8 *addr)
+{
+	return (ar->debug.pktlog_peer_valid &&
+		ether_addr_equal(addr, ar->debug.pktlog_peer_addr));
+}
+
 static inline int ath11k_debug_is_extd_tx_stats_enabled(struct ath11k *ar)
 {
 	return ar->debug.extd_tx_stats;
@@ -157,6 +173,22 @@ static inline int ath11k_debug_is_extd_rx_stats_enabled(struct ath11k *ar)
 {
 	return 0;
 }
+
+static inline bool ath11k_debug_is_pktlog_lite_mode_enabled(struct ath11k *ar)
+{
+	return false;
+}
+
+static inline bool ath11k_debug_is_pktlog_per_peer_enabled(struct ath11k *ar)
+{
+	return false;
+}
+
+static inline bool ath11k_debug_is_pktlog_peer_vaild(struct ath11k *ar, u8 *addr)
+{
+	return false;
+}
+
 #endif /* CPTCFG_ATH11K_DEBUGFS */
 
 #ifdef CPTCFG_MAC80211_DEBUGFS
--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -988,7 +988,7 @@ static int ath11k_htt_pull_ppdu_stats(struct ath11k_base *ab,
 	ppdu_id = *((u32 *)data + 1);
 	ar = ab->pdevs[pdev_id].ar;
 
-	if (ar->debug.pktlog_mode == ATH11K_PKTLOG_MODE_LITE) {
+	if (ath11k_debug_is_pktlog_lite_mode_enabled(ar)) {
 		/* TODO update the pktlog tracing */
 	}
 
@@ -2280,7 +2280,7 @@ int ath11k_dp_rx_process_mon_status(struct ath11k_base *ab, int mac_id,
 		memset(&ppdu_info, 0, sizeof(ppdu_info));
 		ppdu_info.peer_id = HAL_INVALID_PEERID;
 
-		if (!ar->debug.pktlog_peer_valid) {
+		if (!ath11k_debug_is_pktlog_per_peer_enabled(ar)) {
 			/* TODO update the pktlog tracing */
 		}
 
@@ -2309,8 +2309,7 @@ int ath11k_dp_rx_process_mon_status(struct ath11k_base *ab, int mac_id,
 		arsta = (struct ath11k_sta *)peer->sta->drv_priv;
 		ath11k_dp_rx_update_peer_stats(arsta, &ppdu_info);
 
-		if (ar->debug.pktlog_peer_valid &&
-		    ether_addr_equal(peer->addr, ar->debug.pktlog_peer_addr)) {
+		if (ath11k_debug_is_pktlog_peer_vaild(ar, peer->addr)) {
 			/* TODO update the pktlog tracing for one peer*/
 		}
 		spin_unlock_bh(&ab->data_lock);
-- 
1.9.1

