Index: backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20190412-4.4.60-9bcbbf4cedb7.orig/drivers/net/wireless/ath/ath11k/mac.c	2019-04-24 16:33:54.382778428 +0530
+++ backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/mac.c	2019-04-24 16:43:03.890777267 +0530
@@ -1166,7 +1166,6 @@
 		case 5: /* fall through */
 		case 6: /* fall through */
 		default:
-			/* see ath10k_mac_can_set_bitrate_mask() */
 			WARN_ON(1);
 			/* fall through */
 		case -1:
@@ -2283,7 +2282,7 @@
 		   "Setting Fixed VHT Rate for peer %pM. Device will not switch to any other selected rates",
 		   sta->addr);
 
-	rate_code = ATH11K_HW_RATE_CODE(vht_rate, nss,
+	rate_code = ATH11K_HW_RATE_CODE(vht_rate, nss - 1,
 					WMI_RATE_PREAMBLE_VHT);
 	ret = ath11k_wmi_set_peer_param(ar, sta->addr,
 					arvif->vdev_id,
@@ -4346,7 +4345,7 @@
 ath11k_mac_get_single_legacy_rate(struct ath11k *ar,
 				  enum nl80211_band band,
 				  const struct cfg80211_bitrate_mask *mask,
-				  u8 *rate, u8 *nss)
+				  u32 *rate, u8 *nss)
 {
 	int rate_idx;
 	u16 bitrate;
@@ -4376,7 +4375,7 @@
 }
 
 static int ath11k_mac_set_fixed_rate_params(struct ath11k_vif *arvif,
-					    u8 rate, u8 nss, u8 sgi, u8 ldpc)
+					    u32 rate, u8 nss, u8 sgi, u8 ldpc)
 {
 	struct ath11k *ar = arvif->ar;
 	u32 vdev_param;
@@ -4476,7 +4475,7 @@
 	enum nl80211_band band;
 	const u8 *ht_mcs_mask;
 	const u16 *vht_mcs_mask;
-	u8 rate;
+	u32 rate;
 	u8 nss;
 	u8 sgi;
 	u8 ldpc;
Index: backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/wmi.h
===================================================================
--- backports-20190412-4.4.60-9bcbbf4cedb7.orig/drivers/net/wireless/ath/ath11k/wmi.h	2019-04-23 19:15:37.000000000 +0530
+++ backports-20190412-4.4.60-9bcbbf4cedb7/drivers/net/wireless/ath/ath11k/wmi.h	2019-04-24 16:37:20.330777993 +0530
@@ -4884,7 +4884,7 @@
 #define WMI_CIPHER_AES_GMAC 0xa
 
 /* Value to disable fixed rate setting */
-#define WMI_FIXED_RATE_NONE	(0xff)
+#define WMI_FIXED_RATE_NONE	(0xffff)
 
 #define ATH11K_RC_VERSION_OFFSET	28
 #define ATH11K_RC_PREAMBLE_OFFSET	8
