--- a/drivers/net/wireless/ath/ath11k/wmi.h
+++ b/drivers/net/wireless/ath/ath11k/wmi.h
@@ -4885,7 +4885,7 @@ enum wmi_ap_ps_peer_param {
 #define WMI_CIPHER_AES_GMAC 0xa
 
 /* Value to disable fixed rate setting */
-#define WMI_FIXED_RATE_NONE	(0xff)
+#define WMI_FIXED_RATE_NONE	(0xffff)
 
 #define ATH11K_RC_VERSION_OFFSET	28
 #define ATH11K_RC_PREAMBLE_OFFSET	8
--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -2317,7 +2317,7 @@ ath11k_mac_set_peer_vht_fixed_rate(struc
 		   "Setting Fixed VHT Rate for peer %pM. Device will not switch to any other selected rates",
 		   sta->addr);
 
-	rate_code = ATH11K_HW_RATE_CODE(vht_rate, nss,
+	rate_code = ATH11K_HW_RATE_CODE(vht_rate, nss - 1,
 					WMI_RATE_PREAMBLE_VHT);
 	ret = ath11k_wmi_set_peer_param(ar, sta->addr,
 					arvif->vdev_id,
@@ -4436,7 +4436,7 @@ static int
 ath11k_mac_get_single_legacy_rate(struct ath11k *ar,
 				  enum nl80211_band band,
 				  const struct cfg80211_bitrate_mask *mask,
-				  u8 *rate, u8 *nss)
+				  u32 *rate, u8 *nss)
 {
 	int rate_idx;
 	u16 bitrate;
@@ -4466,7 +4466,7 @@ ath11k_mac_get_single_legacy_rate(struct
 }
 
 static int ath11k_mac_set_fixed_rate_params(struct ath11k_vif *arvif,
-					    u8 rate, u8 nss, u8 sgi, u8 ldpc)
+					   u32 rate, u8 nss, u8 sgi, u8 ldpc)
 {
 	struct ath11k *ar = arvif->ar;
 	u32 vdev_param;
@@ -4566,7 +4566,7 @@ ath11k_mac_op_set_bitrate_mask(struct ie
 	enum nl80211_band band;
 	const u8 *ht_mcs_mask;
 	const u16 *vht_mcs_mask;
-	u8 rate;
+	u32 rate;
 	u8 nss;
 	u8 sgi;
 	u8 ldpc;
