--- a/drivers/net/wireless/ath/ath11k/ce.c
+++ b/drivers/net/wireless/ath/ath11k/ce.c
@@ -3,7 +3,7 @@
  * Copyright (c) 2018-2019 The Linux Foundation. All rights reserved.
  */
 
-#include "core.h"
+#include "dp_rx.h"
 #include "debug.h"
 
 static const struct ce_attr host_ce_config_wlan[] = {
@@ -283,8 +283,12 @@ static void ath11k_ce_recv_process_cb(struct ath11k_ce_pipe *pipe)
 		__skb_queue_tail(&list, skb);
 	}
 
-	while ((skb = __skb_dequeue(&list)))
-		ath11k_htc_rx_completion_handler(ab, skb);
+	while ((skb = __skb_dequeue(&list))) {
+		if (pipe->pipe_num == PKTLOG_CE_PIPE_NUM)
+			ath11k_dp_htt_htc_t2h_msg_handler(ab, skb);
+		else
+			ath11k_htc_rx_completion_handler(ab, skb);
+	}
 
 	ret = ath11k_ce_rx_post_pipe(pipe);
 	if (ret && ret != -ENOSPC) {
--- a/drivers/net/wireless/ath/ath11k/ce.h
+++ b/drivers/net/wireless/ath/ath11k/ce.h
@@ -7,6 +7,7 @@
 #define ATH11K_CE_H
 
 #define CE_COUNT 12
+#define PKTLOG_CE_PIPE_NUM 5
 
 /* Byte swap data words */
 #define CE_ATTR_BYTE_SWAP_DATA 2
-- 
1.9.1

