diff --git a/drivers/net/wireless/ath/ath11k/debug.c b/drivers/net/wireless/ath/ath11k/debug.c
index ce7a447..cd5371c 100644
--- a/drivers/net/wireless/ath/ath11k/debug.c
+++ b/drivers/net/wireless/ath/ath11k/debug.c
@@ -54,8 +54,8 @@ void ath11k_warn(struct ath11k_base *sc, const char *fmt, ...)
 }
 
 #ifdef CPTCFG_ATH11K_DEBUG
-void ath11k_dbg(struct ath11k_base *ab, enum ath11k_debug_mask mask,
-		const char *fmt, ...)
+void __ath11k_dbg(struct ath11k_base *ab, enum ath11k_debug_mask mask,
+		  const char *fmt, ...)
 {
 	struct va_format vaf;
 	va_list args;
@@ -84,7 +84,7 @@ void ath11k_dbg_dump(struct ath11k_base *ab,
 
 	if (ath11k_debug_mask & mask) {
 		if (msg)
-			ath11k_dbg(ab, mask, "%s\n", msg);
+			__ath11k_dbg(ab, mask, "%s\n", msg);
 
 		for (ptr = buf; (ptr - buf) < len; ptr += 16) {
 			linebuflen = 0;
diff --git a/drivers/net/wireless/ath/ath11k/debug.h b/drivers/net/wireless/ath/ath11k/debug.h
index 2943c38..8cdd283 100644
--- a/drivers/net/wireless/ath/ath11k/debug.h
+++ b/drivers/net/wireless/ath/ath11k/debug.h
@@ -111,17 +111,17 @@ __printf(2, 3) void ath11k_warn(struct ath11k_base *sc, const char *fmt, ...);
 extern unsigned int ath11k_debug_mask;
 
 #ifdef CPTCFG_ATH11K_DEBUG
-__printf(3, 4) void ath11k_dbg(struct ath11k_base *ab,
-			       enum ath11k_debug_mask mask,
-			       const char *fmt, ...);
+__printf(3, 4) void __ath11k_dbg(struct ath11k_base *ab,
+				 enum ath11k_debug_mask mask,
+				 const char *fmt, ...);
 void ath11k_dbg_dump(struct ath11k_base *ab,
 		     enum ath11k_debug_mask mask,
 		     const char *msg, const char *prefix,
 		     const void *buf, size_t len);
 #else /* CPTCFG_ATH11K_DEBUG */
-static inline int ath11k_dbg(struct ath11k_base *ab,
-			     enum ath11k_debug_mask dbg_mask,
-			     const char *fmt, ...)
+static inline int __ath11k_dbg(struct ath11k_base *ab,
+			       enum ath11k_debug_mask dbg_mask,
+			       const char *fmt, ...)
 {
 	return 0;
 }
@@ -311,4 +311,11 @@ static inline void ath11k_rx_stats_buf_pktlog_process(struct ath11k *ar,
 {
 }
 #endif /* CONFIG_ATH10K_PKTLOG */
+
+#define ath11k_dbg(ar, dbg_mask, fmt, ...)			\
+	do {							\
+		if (ath11k_debug_mask & dbg_mask)		\
+		__ath11k_dbg(ar, dbg_mask, fmt, ##__VA_ARGS__);	\
+	} while (0)
+
 #endif /* _ATH11K_DEBUG_H_ */
