Index: backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/core.h
===================================================================
--- backports-20190610-4.4.60-371886ba35ca.orig/drivers/net/wireless/ath/ath11k/core.h
+++ backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/core.h
@@ -397,6 +397,7 @@ struct ath11k_fw_stats {
 struct ath11k_dbg_htt_stats {
 	u8 type;
 	u8 reset;
+	struct debug_htt_stats_req *stats_req;
 	/* protects shared stats req buffer */
 	spinlock_t lock;
 };
Index: backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debug.h
===================================================================
--- backports-20190610-4.4.60-371886ba35ca.orig/drivers/net/wireless/ath/ath11k/debug.h
+++ backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debug.h
@@ -56,7 +56,6 @@ enum ath11k_dbg_htt_ext_stats_type {
 
 struct debug_htt_stats_req {
 	bool done;
-	u8 pdev_id;
 	u8 type;
 	u8 peer_addr[ETH_ALEN];
 	struct completion cmpln;
@@ -146,8 +145,7 @@ void ath11k_debug_fw_stats_process(struc
 				   u32 len);
 
 void ath11k_debug_fw_stats_init(struct ath11k *ar);
-int ath11k_dbg_htt_stats_req(struct ath11k *ar,
-			     struct debug_htt_stats_req *stats_req);
+int ath11k_dbg_htt_stats_req(struct ath11k *ar);
 
 static inline bool ath11k_debug_is_pktlog_lite_mode_enabled(struct ath11k *ar)
 {
@@ -221,8 +219,7 @@ static inline int ath11k_debug_is_extd_r
 	return 0;
 }
 
-static inline int ath11k_dbg_htt_stats_req(struct ath11k *ar,
-					   struct debug_htt_stats_req *stats_req)
+static inline int ath11k_dbg_htt_stats_req(struct ath11k *ar)
 {
 	return 0;
 }
Index: backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debug_htt_stats.c
===================================================================
--- backports-20190610-4.4.60-371886ba35ca.orig/drivers/net/wireless/ath/ath11k/debug_htt_stats.c
+++ backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debug_htt_stats.c
@@ -4202,16 +4202,28 @@ void ath11k_dbg_htt_ext_stats_handler(st
 
 	data = data + 4;
 	cookie = *(u64 *)data;
-	stats_req = (struct debug_htt_stats_req *)(uintptr_t)cookie;
-	if (!stats_req)
+
+	if (FIELD_GET(HTT_STATS_COOKIE_MSB, cookie) != HTT_STATS_MAGIC_VALUE) {
+		ath11k_warn(ab, "received invalid htt ext stats event\n");
 		return;
+	}
+
+	pdev_id = FIELD_GET(HTT_STATS_COOKIE_LSB, cookie);
+	rcu_read_lock();
+	ar = ath11k_get_ar_by_pdev_id(ab, pdev_id);
+	if (!ar) {
+		ath11k_warn(ab, "failed to get ar for pdev_id %d\n", pdev_id);
+		goto exit;
+	}
+
+	stats_req = ar->debug.htt_stats.stats_req;
+	if (!stats_req)
+		goto exit;
 
-	pdev_id = DP_HW2SW_MACID(stats_req->pdev_id);
-	ar = ab->pdevs[pdev_id].ar;
 	spin_lock_bh(&ar->debug.htt_stats.lock);
 	if (stats_req->done) {
 		spin_unlock_bh(&ar->debug.htt_stats.lock);
-		return;
+		goto exit;
 	}
 	stats_req->done = true;
 	spin_unlock_bh(&ar->debug.htt_stats.lock);
@@ -4226,6 +4238,8 @@ void ath11k_dbg_htt_ext_stats_handler(st
 		ath11k_warn(ab, "Failed to parse tlv %d\n", ret);
 
 	complete(&stats_req->cmpln);
+exit:
+	rcu_read_unlock();
 }
 
 static ssize_t ath11k_read_htt_stats_type(struct file *file,
@@ -4317,9 +4331,9 @@ static int ath11k_prep_htt_stats_cfg_par
 	return 0;
 }
 
-int ath11k_dbg_htt_stats_req(struct ath11k *ar,
-			     struct debug_htt_stats_req *stats_req)
+int ath11k_dbg_htt_stats_req(struct ath11k *ar)
 {
+	struct debug_htt_stats_req *stats_req = ar->debug.htt_stats.stats_req;
 	u8 type = stats_req->type;
 	u64 cookie = 0;
 	int ret, pdev_id = ar->pdev->pdev_id;
@@ -4328,9 +4342,9 @@ int ath11k_dbg_htt_stats_req(struct ath1
 	init_completion(&stats_req->cmpln);
 
 	stats_req->done = false;
-	stats_req->pdev_id = pdev_id;
 
-	cookie = (u64)(uintptr_t)stats_req;
+	cookie = FIELD_PREP(HTT_STATS_COOKIE_MSB, HTT_STATS_MAGIC_VALUE) |
+		 FIELD_PREP(HTT_STATS_COOKIE_LSB, pdev_id);
 
 	ret = ath11k_prep_htt_stats_cfg_params(ar, type, stats_req->peer_addr,
 					       &cfg_params);
@@ -4339,14 +4353,12 @@ int ath11k_dbg_htt_stats_req(struct ath1
 		return ret;
 	}
 
-	mutex_lock(&ar->conf_mutex);
 	ret = ath11k_dp_htt_h2t_ext_stats_req(ar, type, &cfg_params, cookie);
 	if (ret) {
 		ath11k_warn(ar->ab, "failed to send htt stats request: %d\n", ret);
 		mutex_unlock(&ar->conf_mutex);
 		return ret;
 	}
-	mutex_unlock(&ar->conf_mutex);
 
 	while (!wait_for_completion_timeout(&stats_req->cmpln, 3 * HZ)) {
 		spin_lock_bh(&ar->debug.htt_stats.lock);
@@ -4376,8 +4388,11 @@ static int ath11k_open_htt_stats(struct
 	if (!stats_req)
 		return -ENOMEM;
 
+	mutex_lock(&ar->conf_mutex);
+	ar->debug.htt_stats.stats_req = stats_req;
 	stats_req->type = type;
-	ret = ath11k_dbg_htt_stats_req(ar, stats_req);
+	ret = ath11k_dbg_htt_stats_req(ar);
+	mutex_unlock(&ar->conf_mutex);
 	if (ret < 0)
 		goto out;
 
Index: backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debugfs_sta.c
===================================================================
--- backports-20190610-4.4.60-371886ba35ca.orig/drivers/net/wireless/ath/ath11k/debugfs_sta.c
+++ backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debugfs_sta.c
@@ -415,9 +415,12 @@ ath11k_dbg_sta_open_htt_peer_stats(struc
 	if (!stats_req)
 		return -ENOMEM;
 
+	mutex_lock(&ar->conf_mutex);
+	ar->debug.htt_stats.stats_req = stats_req;
 	stats_req->type = ATH11K_DBG_HTT_EXT_STATS_PEER_INFO;
 	memcpy(stats_req->peer_addr, sta->addr, ETH_ALEN);
-	ret = ath11k_dbg_htt_stats_req(ar, stats_req);
+	ret = ath11k_dbg_htt_stats_req(ar);
+	mutex_unlock(&ar->conf_mutex);
 	if (ret < 0)
 		goto out;
 
Index: backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debug_htt_stats.h
===================================================================
--- backports-20190610-4.4.60-371886ba35ca.orig/drivers/net/wireless/ath/ath11k/debug_htt_stats.h
+++ backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/debug_htt_stats.h
@@ -6,6 +6,10 @@
 #ifndef DEBUG_HTT_STATS_H
 #define DEBUG_HTT_STATS_H
 
+#define HTT_STATS_COOKIE_LSB    GENMASK_ULL(31, 0)
+#define HTT_STATS_COOKIE_MSB    GENMASK_ULL(63, 32)
+#define HTT_STATS_MAGIC_VALUE   0xF0F0F0F0
+
 enum htt_tlv_tag_t {
 	HTT_STATS_TX_PDEV_CMN_TAG                           = 0,
 	HTT_STATS_TX_PDEV_UNDERRUN_TAG                      = 1,
