Index: backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20190610-4.4.60-371886ba35ca.orig/drivers/net/wireless/ath/ath11k/mac.c	2019-06-19 16:42:59.884556478 +0530
+++ backports-20190610-4.4.60-371886ba35ca/drivers/net/wireless/ath/ath11k/mac.c	2019-06-19 16:46:16.416556063 +0530
@@ -5107,6 +5107,10 @@
 			return -EINVAL;
 		}
 
+		ieee80211_iterate_stations_atomic(ar->hw,
+						  ath11k_mac_disable_peer_fixed_rate,
+						  arvif);
+
 		mutex_lock(&ar->conf_mutex);
 
 		arvif->bitrate_mask = *mask;
