--- a/drivers/net/wireless/ath/ath11k/dp.c
+++ b/drivers/net/wireless/ath/ath11k/dp.c
@@ -200,7 +200,6 @@ static int ath11k_dp_srng_common_setup(s
 	struct ath11k_dp *dp = &ab->dp;
 	struct hal_srng *srng;
 	int i, ret;
-	struct hal_reo_params reo_params = {0};
 
 	ret = ath11k_dp_srng_setup(ab, &dp->wbm_desc_rel_ring,
 				   HAL_SW2WBM_RELEASE, 0, 0,
@@ -288,22 +287,7 @@ static int ath11k_dp_srng_common_setup(s
 		goto err;
 	}
 
-	/* When hash based routing of rx packet is enabled, 16 entries to map
-	 * the hash values to the ring will be configured. Each hash entry uses
-	 * three bits to map to a particular ring. The ring mapping will be
-	 * 0:TCL, 1:SW1, 2:SW2, 3:SW3, 4:SW4, 5:Release, 6:FW and 7:Not used.
-	 */
-	reo_params.ring_hash_map0_7 = HAL_HASH_ROUTING_RING_SW1 << 0 |
-				      HAL_HASH_ROUTING_RING_SW2 << 3 |
-				      HAL_HASH_ROUTING_RING_SW3 << 6 |
-				      HAL_HASH_ROUTING_RING_SW4 << 9 |
-				      HAL_HASH_ROUTING_RING_SW1 << 12 |
-				      HAL_HASH_ROUTING_RING_SW2 << 15 |
-				      HAL_HASH_ROUTING_RING_SW3 << 18 |
-				      HAL_HASH_ROUTING_RING_SW4 << 21;
-	reo_params.ring_hash_map8_15 = reo_params.ring_hash_map0_7;
-
-	ath11k_hal_reo_hw_setup(ab, &reo_params);
+	ath11k_hal_reo_hw_setup(ab);
 
 	return 0;
 
--- a/drivers/net/wireless/ath/ath11k/hal_rx.c
+++ b/drivers/net/wireless/ath/ath11k/hal_rx.c
@@ -833,8 +833,7 @@ void ath11k_hal_reo_init_cmd_ring(struct
 	}
 }
 
-void ath11k_hal_reo_hw_setup(struct ath11k_base *ab,
-			     struct hal_reo_params *params)
+void ath11k_hal_reo_hw_setup(struct ath11k_base *ab)
 {
 	u32 reo_base = HAL_SEQ_WCSS_UMAC_REO_REG;
 	u32 val;
@@ -856,13 +855,6 @@ void ath11k_hal_reo_hw_setup(struct ath1
 			   HAL_DEFAULT_REO_TIMEOUT_USEC);
 	ath11k_ahb_write32(ab, reo_base + HAL_REO1_AGING_THRESH_IX_3,
 			   HAL_DEFAULT_REO_TIMEOUT_USEC);
-
-	ath11k_ahb_write32(ab, reo_base + HAL_REO1_DEST_RING_CTRL_IX_2,
-			   FIELD_PREP(HAL_REO_DEST_RING_CTRL_HASH_RING_MAP,
-				      params->ring_hash_map0_7));
-	ath11k_ahb_write32(ab, reo_base + HAL_REO1_DEST_RING_CTRL_IX_3,
-			   FIELD_PREP(HAL_REO_DEST_RING_CTRL_HASH_RING_MAP,
-				      params->ring_hash_map8_15));
 }
 
 static enum hal_rx_mon_status
--- a/drivers/net/wireless/ath/ath11k/hal.h
+++ b/drivers/net/wireless/ath/ath11k/hal.h
@@ -717,19 +717,6 @@ enum hal_ce_desc {
 	HAL_CE_DESC_DST_STATUS,
 };
 
-#define HAL_HASH_ROUTING_RING_TCL 0
-#define HAL_HASH_ROUTING_RING_SW1 1
-#define HAL_HASH_ROUTING_RING_SW2 2
-#define HAL_HASH_ROUTING_RING_SW3 3
-#define HAL_HASH_ROUTING_RING_SW4 4
-#define HAL_HASH_ROUTING_RING_REL 5
-#define HAL_HASH_ROUTING_RING_FW  6
-
-struct hal_reo_params {
-	u32 ring_hash_map0_7;
-	u32 ring_hash_map8_15;
-};
-
 struct hal_reo_status_header {
 	u16 cmd_num;
 	enum hal_reo_cmd_status cmd_status;
@@ -863,8 +850,7 @@ void ath11k_hal_reo_qdesc_setup(void *va
 				u32 start_seqtype);
 void ath11k_hal_reo_init_cmd_ring(struct ath11k_base *ab,
 				  struct hal_srng *srng);
-void ath11k_hal_reo_hw_setup(struct ath11k_base *ab,
-			     struct hal_reo_params *params);
+void ath11k_hal_reo_hw_setup(struct ath11k_base *ab);
 void ath11k_hal_setup_link_idle_list(struct ath11k_base *ab,
 				     struct hal_wbm_idle_scatter_list *sbuf,
 				     u32 nsbufs, u32 tot_link_desc,
