Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/wmi.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/wmi.c	2019-09-05 10:37:58.014739120 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/wmi.c	2019-09-05 11:37:43.080259107 +0530
@@ -338,7 +338,7 @@
 	if (phy_id >= hal_reg_caps->num_phy)
 		return -EINVAL;
 
-	mac_phy_caps = &wmi_mac_phy_caps[phy_idx];
+	mac_phy_caps = wmi_mac_phy_caps + phy_idx;
 
 	pdev->pdev_id = mac_phy_caps->pdev_id;
 	pdev_cap->supported_bands = mac_phy_caps->supported_bands;
@@ -3287,6 +3287,14 @@
 	if (svc_rdy_ext->n_mac_phy_caps >= svc_rdy_ext->tot_phy_id)
 		return -ENOBUFS;
 
+	len = min_t(u16, len, sizeof(struct wmi_mac_phy_capabilities));
+	if (!svc_rdy_ext->n_mac_phy_caps) {
+		svc_rdy_ext->mac_phy_caps = kzalloc((svc_rdy_ext->tot_phy_id) * len,
+				GFP_ATOMIC);
+		if (!svc_rdy_ext->mac_phy_caps)
+			return -ENOMEM;
+	}
+	memcpy(svc_rdy_ext->mac_phy_caps + svc_rdy_ext->n_mac_phy_caps, ptr, len);
 	svc_rdy_ext->n_mac_phy_caps++;
 	return 0;
 }
@@ -3509,8 +3517,7 @@
 			svc_rdy_ext->hw_mode_done = true;
 		} else if (!svc_rdy_ext->mac_phy_done) {
 			svc_rdy_ext->n_mac_phy_caps = 0;
-			svc_rdy_ext->mac_phy_caps =
-					(struct wmi_mac_phy_capabilities *)ptr;
+
 			ret = ath11k_wmi_tlv_iter(ab, ptr, len,
 						  ath11k_wmi_tlv_mac_phy_caps_parse,
 						  svc_rdy_ext);
@@ -3565,6 +3572,7 @@
 
 	if (!test_bit(WMI_TLV_SERVICE_EXT2_MSG, ab->wmi_sc.svc_map))
 		complete(&ab->wmi_sc.service_ready);
+	kfree(svc_rdy_ext.mac_phy_caps);
 
 	return 0;
 
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/wmi.h
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/wmi.h	2019-09-05 10:37:58.014739120 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/wmi.h	2019-09-05 10:37:58.010739109 +0530
@@ -2260,7 +2260,6 @@
 #define WMI_MAX_HECAP_PHY_SIZE                 (3)
 
 struct wmi_mac_phy_capabilities {
-	u32 tlv_header;
 	u32 hw_mode_id;
 	u32 pdev_id;
 	u32 phy_id;
