diff --git a/drivers/net/wireless/ath/ath11k/core.c b/drivers/net/wireless/ath/ath11k/core.c
index 1c002ac..9f1427c 100644
--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -581,6 +581,7 @@ static int ath11k_core_reconfigure_on_crash(struct ath11k_base *sc)
 	int ret;
 
 	mutex_lock(&sc->core_lock);
+	ath11k_thermal_unregister(sc);
 	ath11k_ahb_ext_irq_disable(sc);
 	ath11k_dp_pdev_free(sc);
 	ath11k_spectral_deinit(sc);
