Index: backports-20190626-4.4.60-4c6476d00657/drivers/net/wireless/ath/ath11k/core.h
===================================================================
--- backports-20190626-4.4.60-4c6476d00657.orig/drivers/net/wireless/ath/ath11k/core.h
+++ backports-20190626-4.4.60-4c6476d00657/drivers/net/wireless/ath/ath11k/core.h
@@ -352,6 +352,7 @@ struct ath11k_sta {
 	struct rate_info txrate;
 	struct rate_info last_txrate;
 	u64 rx_duration;
+	u64 tx_duration;
 	u8 rssi_comb;
 	struct ath11k_htt_tx_stats *tx_stats;
 	struct ath11k_rx_peer_stats *rx_stats;
Index: backports-20190626-4.4.60-4c6476d00657/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20190626-4.4.60-4c6476d00657.orig/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ backports-20190626-4.4.60-4c6476d00657/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -1181,6 +1181,7 @@ ath11k_update_per_peer_tx_stats(struct a
 	arsta->txrate.nss = nss;
 	arsta->txrate.bw = ath11k_mac_bw_to_mac80211_bw(bw);
 	arsta->tx_info.status.rates[0].flags |= ath11k_bw_to_mac80211_bwflags(bw);
+	arsta->tx_duration += tx_duration;
 
 	memcpy(&arsta->last_txrate, &arsta->txrate, sizeof(struct rate_info));
 
Index: backports-20190626-4.4.60-4c6476d00657/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20190626-4.4.60-4c6476d00657.orig/drivers/net/wireless/ath/ath11k/mac.c
+++ backports-20190626-4.4.60-4c6476d00657/drivers/net/wireless/ath/ath11k/mac.c
@@ -5386,6 +5386,9 @@ static void ath11k_mac_op_sta_statistics
 	sinfo->rx_duration = arsta->rx_duration;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_DURATION);
 
+	sinfo->tx_duration = arsta->tx_duration;
+	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_DURATION);
+
 	if (!arsta->txrate.legacy && !arsta->txrate.nss)
 		return;
 
