--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -2531,7 +2531,8 @@ static void ath11k_dp_rx_update_peer_sta
 	rx_stats->num_mpdu_fcs_ok += ppdu_info->num_mpdu_fcs_ok;
 	rx_stats->num_mpdu_fcs_err += ppdu_info->num_mpdu_fcs_err;
 
-	rx_stats->rx_duration = ppdu_info->rx_duration;
+	rx_stats->rx_duration += ppdu_info->rx_duration;
+	arsta->rx_duration = rx_stats->rx_duration;
 
 	if (ppdu_info->nss > 0 && ppdu_info->nss <= HAL_RX_MAX_NSS) {
 		rx_stats->pkt_stats.nss_count[ppdu_info->nss - 1] += num_msdu;
@@ -2825,7 +2826,6 @@ int ath11k_dp_rx_process_mon_status(stru
 		}
 
 		arsta = (struct ath11k_sta *)peer->sta->drv_priv;
-		arsta->rx_duration = ppdu_info.rx_duration;
 		ath11k_dp_rx_update_peer_stats(arsta, &ppdu_info);
 
 		if (ath11k_debug_is_pktlog_peer_valid(ar, peer->addr)) {
