Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/core.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/core.c	2020-03-03 19:04:07.000000000 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/core.c	2020-03-03 19:04:10.901376272 +0530
@@ -721,6 +721,7 @@
 	lockdep_assert_held(&ar->conf_mutex);
 
 	ar->num_created_vdevs = 0;
+	ar->allocated_vdev_map = 0;
 
 	ath11k_mac_scan_finish(ar);
 	ath11k_mac_peer_cleanup_all(ar);
Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/core.h
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/core.h	2020-03-03 19:04:07.000000000 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/core.h	2020-03-03 19:04:10.905376272 +0530
@@ -558,6 +558,7 @@
 	int max_num_peers;
 	u32 num_started_vdevs;
 	u32 num_created_vdevs;
+	unsigned long long allocated_vdev_map;
 
 	struct idr txmgmt_idr;
 	/* protects txmgmt_idr data */
Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/mac.c	2020-03-03 19:04:07.049376321 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/mac.c	2020-03-03 19:04:41.793375881 +0530
@@ -383,8 +383,10 @@
 						   flags,
 						   ath11k_mac_get_arvif_iter,
 						   &arvif_iter);
-	if (!arvif_iter.arvif)
+	if (!arvif_iter.arvif) {
+		ath11k_warn(ar->ab, "No VIF found for vdev %d\n", vdev_id);
 		return NULL;
+	}
 
 	return arvif_iter.arvif;
 }
@@ -417,11 +419,7 @@
 	for (i = 0; i < ab->num_radios; i++) {
 		pdev = rcu_dereference(ab->pdevs_active[i]);
 		if (pdev && pdev->ar) {
-			arvif = ath11k_mac_get_arvif(pdev->ar, vdev_id);
-			if (arvif)
-				return arvif->ar;
-			/* vdev start resp for monitor interface */
-			if (pdev->ar->monitor_vdev_id == vdev_id)
+			if (pdev->ar->allocated_vdev_map & (1LL << vdev_id))
 				return pdev->ar;
 		}
 	}
@@ -447,31 +445,6 @@
 	return NULL;
 }
 
-struct ath11k *ath11k_mac_get_ar_from_arvif_list(struct ath11k_base *ab, u32 vdev_id)
-{
-	int i;
-	struct ath11k_pdev *pdev;
-	struct ath11k_vif *arvif;
-	struct ath11k *ar;
-
-	for (i = 0; i < ab->num_radios; i++) {
-		pdev = rcu_dereference(ab->pdevs_active[i]);
-		if (pdev && pdev->ar) {
-			ar = pdev->ar;
-
-			spin_lock_bh(&ar->data_lock);
-			list_for_each_entry(arvif, &ar->arvifs, list) {
-				if (arvif->vdev_id == vdev_id) {
-					spin_unlock_bh(&ar->data_lock);
-					return arvif->ar;
-				}
-			}
-			spin_unlock_bh(&ar->data_lock);
-		}
-	}
-	return NULL;
-}
-
 static void ath11k_pdev_caps_update(struct ath11k *ar)
 {
 	struct ath11k_base *ab = ar->ab;
@@ -4202,6 +4175,7 @@
 	ar->num_started_vdevs = 0;
 	ar->num_created_vdevs = 0;
 	ar->num_peers = 0;
+	ar->allocated_vdev_map = 0;
 
 	/* Configure monitor status ring with default rx_filter to get rx status
 	 * such as rssi, rx_duration.
@@ -4446,7 +4420,9 @@
 	}
 	ar->num_created_vdevs++;
 
+	ar->allocated_vdev_map |= 1LL << arvif->vdev_id;
 	ab->free_vdev_map &= ~(1LL << arvif->vdev_id);
+
 	spin_lock_bh(&ar->data_lock);
 	list_add(&arvif->list, &ar->arvifs);
 	spin_unlock_bh(&ar->data_lock);
@@ -4588,6 +4564,7 @@
 err_vdev_del:
 	ath11k_wmi_vdev_delete(ar, arvif->vdev_id);
 	ar->num_created_vdevs--;
+	ar->allocated_vdev_map &= ~(1LL << arvif->vdev_id);
 	ab->free_vdev_map |= 1LL << arvif->vdev_id;
 	spin_lock_bh(&ar->data_lock);
 	list_del(&arvif->list);
@@ -4646,6 +4623,7 @@
 	if (time_left == 0) {
 		ath11k_warn(ab, "Timeout in receiving vdev delete response\n");
 	} else {
+		ar->allocated_vdev_map &= ~(1LL << arvif->vdev_id);
 		ab->free_vdev_map |= 1LL << (arvif->vdev_id);
 		ar->num_created_vdevs--;
 	}
@@ -6153,6 +6131,7 @@
 	if (ret)
 		return ret;
 
+	ar->allocated_vdev_map |= 1LL << ar->monitor_vdev_id;
 	ar->ab->free_vdev_map &= ~(1LL << ar->monitor_vdev_id);
 	ath11k_dbg(ar->ab, ATH11K_DBG_MAC, "mac monitor vdev %d created\n",
 		   ar->monitor_vdev_id);
@@ -6181,6 +6160,7 @@
 	if (time_left == 0) {
 		ath11k_warn(ar->ab, "Timeout in receiving vdev delete response\n");
 	} else {
+		ar->allocated_vdev_map &= ~(1LL << ar->monitor_vdev_id);
 		ar->ab->free_vdev_map |= 1LL << (ar->monitor_vdev_id);
 		ar->num_created_vdevs--;
 		ar->monitor_vdev_id = -1;
Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/mac.h
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/mac.h	2020-03-03 19:04:06.000000000 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/mac.h	2020-03-03 19:04:10.905376272 +0530
@@ -139,7 +139,6 @@
 			   int coex, u32 wlan_prio_mask, u8 wlan_weight);
 struct ath11k *ath11k_mac_get_ar_by_vdev_id(struct ath11k_base *ab, u32 vdev_id);
 struct ath11k *ath11k_mac_get_ar_by_pdev_id(struct ath11k_base *ab, u32 pdev_id);
-struct ath11k *ath11k_mac_get_ar_from_arvif_list(struct ath11k_base *ab, u32 vdev_id);
 void ath11k_mac_drain_tx(struct ath11k *ar);
 void ath11k_mac_peer_cleanup_all(struct ath11k *ar);
 int ath11k_mac_tx_mgmt_pending_free(int buf_id, void *skb, void *ctx);
Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/wmi.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/wmi.c	2020-03-03 19:04:07.000000000 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/wmi.c	2020-03-03 19:04:10.909376272 +0530
@@ -5796,7 +5796,7 @@
 	}
 
 	rcu_read_lock();
-	ar = ath11k_mac_get_ar_from_arvif_list(ab, peer_del_resp.vdev_id);
+	ar = ath11k_mac_get_ar_by_vdev_id(ab, peer_del_resp.vdev_id);
 	if (!ar) {
 		ath11k_warn(ab, "invalid vdev id in peer delete resp ev %d",
 			    peer_del_resp.vdev_id);
@@ -5824,7 +5824,7 @@
 	}
 
 	rcu_read_lock();
-	ar = ath11k_mac_get_ar_from_arvif_list(ab, vdev_id);
+	ar = ath11k_mac_get_ar_by_vdev_id(ab, vdev_id);
 	if (!ar) {
 		ath11k_warn(ab, "invalid vdev id in vdev delete resp ev %d",
 			    vdev_id);
@@ -5915,7 +5915,7 @@
 	}
 
 	rcu_read_lock();
-	ar = ath11k_mac_get_ar_from_arvif_list(ab, vdev_id);
+	ar = ath11k_mac_get_ar_by_vdev_id(ab, vdev_id);
 	if (!ar) {
 		ath11k_warn(ab, "invalid vdev id in vdev stopped ev %d",
 			    vdev_id);
