Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/mac.c	2020-03-03 19:10:43.000000000 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/mac.c	2020-03-03 19:12:17.285370123 +0530
@@ -4143,15 +4143,28 @@
 
 	while ((skb = skb_dequeue(&ar->wmi_mgmt_tx_queue)) != NULL) {
 		info = IEEE80211_SKB_CB(skb);
-		arvif = ath11k_vif_to_arvif(info->control.vif);
-
-		ret = ath11k_mac_mgmt_tx_wmi(ar, arvif, skb);
-		if (ret) {
-			ath11k_warn(ar->ab, "failed to transmit management frame %d\n",
-				    ret);
+		if (!info->control.vif) {
+			ath11k_warn(ar->ab, "no vif found for mgmt frame, flags 0x%x\n",
+				    info->control.flags);
 			ieee80211_free_txskb(ar->hw, skb);
+			continue;
+		}
+
+		arvif = ath11k_vif_to_arvif(info->control.vif);
+		if (ar->allocated_vdev_map & (1LL << arvif->vdev_id)) {
+			ret = ath11k_mac_mgmt_tx_wmi(ar, arvif, skb);
+			if (ret) {
+				ath11k_warn(ar->ab, "failed to tx mgmt frame, vdev_id %d :%d\n",
+					    arvif->vdev_id, ret);
+				ieee80211_free_txskb(ar->hw, skb);
+			} else {
+				atomic_inc(&ar->num_pending_mgmt_tx);
+			}
 		} else {
-			atomic_inc(&ar->num_pending_mgmt_tx);
+			ath11k_warn(ar->ab,
+				    "dropping mgmt frame for deleted vdev %d, flags 0x%x\n",
+				    arvif->vdev_id, info->control.flags);
+			ieee80211_free_txskb(ar->hw, skb);
 		}
 	}
 }
@@ -4501,8 +4514,10 @@
 			    arvif->vdev_id, ret);
 		goto err;
 	}
-	ar->num_created_vdevs++;
+	ath11k_info(ab, "vdev %pM created, vdev_id %d\n",
+		    vif->addr, arvif->vdev_id);
 
+	ar->num_created_vdevs++;
 	ar->allocated_vdev_map |= 1LL << arvif->vdev_id;
 	ab->free_vdev_map &= ~(1LL << arvif->vdev_id);
 
@@ -4706,6 +4721,8 @@
 	if (time_left == 0) {
 		ath11k_warn(ab, "Timeout in receiving vdev delete response\n");
 	} else {
+		ath11k_info(ab, "vdev %pM deleted, vdev_id %d\n",
+			    vif->addr, arvif->vdev_id);
 		ar->allocated_vdev_map &= ~(1LL << arvif->vdev_id);
 		ab->free_vdev_map |= 1LL << (arvif->vdev_id);
 		ar->num_created_vdevs--;
@@ -4961,6 +4978,8 @@
 		return ret;
 	}
 
+	ath11k_info(ar->ab, "vdev %pM started, vdev_id %d\n",
+		    arvif->vif->addr, arvif->vdev_id);
 	ar->num_started_vdevs++;
 
 	/* Enable CAC Flag in the driver by checking the channel DFS cac time,
@@ -5013,6 +5032,8 @@
 
 	WARN_ON(ar->num_started_vdevs == 0);
 
+	ath11k_info(ar->ab, "vdev %pM stopped, vdev_id %d\n",
+		    arvif->vif->addr, arvif->vdev_id);
 	ar->num_started_vdevs--;
 
 	if (test_bit(ATH11K_CAC_RUNNING, &ar->dev_flags)) {
