--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -4329,6 +4329,7 @@ static __le16 ath11k_mac_setup_he_6ghz_c
 }
 
 static void ath11k_mac_set_hemcsmap(struct ath11k *ar,
+				    struct ath11k_pdev_cap *cap,
 				    struct ieee80211_sta_he_cap * he_cap)
 {
 	u16 txmcs_map, rxmcs_map;
@@ -4337,12 +4338,14 @@ static void ath11k_mac_set_hemcsmap(stru
 	rxmcs_map = 0;
 	txmcs_map = 0;
 	for (i = 0; i < 8; i++) {
-		if (i < ar->num_tx_chains && ar->cfg_tx_chainmask & BIT(i))
+		if (i < ar->num_tx_chains &&
+		    (ar->cfg_tx_chainmask >> cap->tx_chain_mask_shift) & BIT(i))
 			txmcs_map |= IEEE80211_HE_MCS_SUPPORT_0_11 << (i * 2);
 		else
 			txmcs_map |= IEEE80211_HE_MCS_NOT_SUPPORTED << (i * 2);
 
-		if (i < ar->num_rx_chains && ar->cfg_rx_chainmask & BIT(i))
+		if (i < ar->num_rx_chains &&
+		    (ar->cfg_rx_chainmask >> cap->tx_chain_mask_shift) & BIT(i))
 			rxmcs_map |= IEEE80211_HE_MCS_SUPPORT_0_11 << (i * 2);
 		else
 			rxmcs_map |= IEEE80211_HE_MCS_NOT_SUPPORTED << (i * 2);
@@ -4352,13 +4355,13 @@ static void ath11k_mac_set_hemcsmap(stru
 	he_cap->he_mcs_nss_supp.tx_mcs_80 =
 		cpu_to_le16(txmcs_map & 0xffff);
 	he_cap->he_mcs_nss_supp.rx_mcs_160 =
-		cpu_to_le16((rxmcs_map >> 16) & 0xffff);
+		cpu_to_le16(rxmcs_map & 0xffff);
 	he_cap->he_mcs_nss_supp.tx_mcs_160 =
-		cpu_to_le16((txmcs_map >> 16) & 0xffff);
+		cpu_to_le16(txmcs_map & 0xffff);
 	he_cap->he_mcs_nss_supp.rx_mcs_80p80 =
-		cpu_to_le16((rxmcs_map >> 16) & 0xffff);
+		cpu_to_le16(rxmcs_map & 0xffff);
 	he_cap->he_mcs_nss_supp.tx_mcs_80p80 =
-		cpu_to_le16((txmcs_map >> 16) & 0xffff);
+		cpu_to_le16(txmcs_map & 0xffff);
 }
 
 static int ath11k_mac_copy_he_cap(struct ath11k *ar,
@@ -4393,6 +4396,10 @@ static int ath11k_mac_copy_he_cap(struct
 
 		he_cap_elem->mac_cap_info[1] &=
 			IEEE80211_HE_MAC_CAP1_TF_MAC_PAD_DUR_MASK;
+		he_cap_elem->phy_cap_info[0] &=
+			~IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_160MHZ_IN_5G;
+		he_cap_elem->phy_cap_info[0] &=
+			~IEEE80211_HE_PHY_CAP0_CHANNEL_WIDTH_SET_80PLUS80_MHZ_IN_5G;
 		he_cap_elem->phy_cap_info[4] &=
 			~IEEE80211_HE_PHY_CAP4_BEAMFORMEE_MAX_STS_UNDER_80MHZ_MASK;
 		he_cap_elem->phy_cap_info[4] &=
@@ -4425,7 +4432,7 @@ static int ath11k_mac_copy_he_cap(struct
 			break;
 		}
 
-		ath11k_mac_set_hemcsmap(ar, he_cap);
+		ath11k_mac_set_hemcsmap(ar, cap, he_cap);
 
 		memset(he_cap->ppe_thres, 0, sizeof(he_cap->ppe_thres));
 		if (he_cap_elem->phy_cap_info[6] &
--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -4991,16 +4991,17 @@ static int ath11k_check_chain_mask(struct ath11k *ar, u32 ant, bool is_tx_ant)
 }
 
 static void ath11k_gen_ppe_thresh(struct ath11k_ppe_threshold *fw_ppet,
-				  u8 *he_ppet)
+				  u8 *he_ppet, u32 num_txrx_chains)
 {
 	int nss, ru;
+	u32 max_nss = min(num_txrx_chains, fw_ppet->numss_m1);
 	u8 bit = 7;
 
-	he_ppet[0] = fw_ppet->numss_m1 & IEEE80211_PPE_THRES_NSS_MASK;
+	he_ppet[0] = (max_nss - 1) & IEEE80211_PPE_THRES_NSS_MASK;
 	he_ppet[0] |= (fw_ppet->ru_bit_mask <<
 		       IEEE80211_PPE_THRES_RU_INDEX_BITMASK_POS) &
 		      IEEE80211_PPE_THRES_RU_INDEX_BITMASK_MASK;
-	for (nss = 0; nss <= fw_ppet->numss_m1; nss++) {
+	for (nss = 0; nss < max_nss; nss++) {
 		for (ru = 0; ru < 4; ru++) {
 			u8 val;
 			int i;
@@ -5161,6 +5162,7 @@ static int ath11k_mac_copy_he_cap(struct ath11k *ar,
 				  int band)
 {
 	int i, idx = 0;
+	u32 nss = min(ar->num_tx_chains, ar->num_rx_chains);
 
 	for (i = 0; i < NUM_NL80211_IFTYPES; i++) {
 		struct ieee80211_sta_he_cap *he_cap = &data[idx].he_cap;
@@ -5229,7 +5231,7 @@ static int ath11k_mac_copy_he_cap(struct ath11k *ar,
 		if (he_cap_elem->phy_cap_info[6] &
 		    IEEE80211_HE_PHY_CAP6_PPE_THRESHOLD_PRESENT)
 			ath11k_gen_ppe_thresh(&band_cap->he_ppet,
-					      he_cap->ppe_thres);
+					      he_cap->ppe_thres, nss);
 
 		if (band == NL80211_BAND_6GHZ) {
 			data[idx].he_6ghz_capa.capa =

