Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp_rx.c	2019-09-03 13:07:44.822749630 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c	2019-09-03 13:07:44.822749630 +0530
@@ -3286,7 +3286,8 @@
 
 	ath11k_dp_rx_h_ppdu(ar, desc, status);
 
-	status->flag |= RX_FLAG_MMIC_ERROR;
+	status->flag |= (RX_FLAG_MMIC_STRIPPED | RX_FLAG_MMIC_ERROR |
+			 RX_FLAG_DECRYPTED);
 
 	ath11k_dp_rx_h_undecap(ar, msdu, desc,
 			       HAL_ENCRYPT_TYPE_TKIP_MIC, status, false);
