--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -2857,6 +2857,7 @@
 	struct ath11k *ar = hw->priv;
 	struct ath11k_vif *arvif = ath11k_vif_to_arvif(vif);
 	struct ath11k_sta *arsta = (struct ath11k_sta *)sta->drv_priv;
+	struct ath11k_peer *peer;
 	int ret = 0;
 
 	/* cancel must be done outside the mutex to avoid deadlock */
@@ -2891,6 +2892,15 @@
 
 		ath11k_mac_dec_num_stations(arvif, sta);
 
+		spin_lock_bh(&ar->data_lock);
+		peer = ath11k_peer_find(ar->ab, arvif->vdev_id, sta->addr);
+		if (peer && (peer->sta == sta)) {
+			peer->sta = NULL;
+			ath11k_warn(ar->ab, "Found peer entry %pM n vdev %i after it was supposedly removed\n",
+				   vif->addr, arvif->vdev_id);
+		}
+		spin_unlock_bh(&ar->data_lock);
+
 		kfree(arsta->tx_stats);
 		arsta->tx_stats = NULL;
 		ret = ath11k_mac_ap_ps_recalc(ar);
