--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -5452,6 +5452,11 @@ static void ath11k_mac_op_sta_statistics
 	sinfo->rx_duration = arsta->rx_duration;
 	sinfo->filled |= BIT_ULL(NL80211_STA_INFO_RX_DURATION);
 
+	if (arsta->tx_stats) {
+		sinfo->tx_duration = arsta->tx_stats->tx_duration;
+		sinfo->filled |= BIT_ULL(NL80211_STA_INFO_TX_DURATION);
+	}
+
 	if (!arsta->txrate.legacy && !arsta->txrate.nss)
 		return;
 
