Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp.c	2019-09-25 11:21:28.083337748 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp.c	2019-09-25 11:26:01.079931993 +0530
@@ -42,8 +42,9 @@
 int ath11k_dp_peer_setup(struct ath11k *ar, int vdev_id, const u8 *addr)
 {
 	struct ath11k_base *ab = ar->ab;
+	struct ath11k_peer *peer;
 	u32 reo_dest;
-	int ret;
+	int ret = 0, tid;
 
 	/* NOTE: reo_dest ring id starts from 1 unlike mac_id which starts from 0 */
 	reo_dest = ar->dp.mac_id + 1;
@@ -57,20 +58,15 @@
 		return ret;
 	}
 
-	ret = ath11k_peer_rx_tid_setup(ar, addr, vdev_id,
-				       HAL_DESC_REO_NON_QOS_TID, 1, 0, HAL_PN_TYPE_NONE);
-	if (ret) {
-		ath11k_warn(ab, "failed to setup rxd tid queue for non-qos tid %d\n",
-			    ret);
-		return ret;
-	}
-
-	ret = ath11k_peer_rx_tid_setup(ar, addr, vdev_id, 0, 1, 0, HAL_PN_TYPE_NONE);
-	if (ret) {
-		ath11k_warn(ab, "failed to setup rxd tid queue for tid 0 %d\n",
-			    ret);
-		return ret;
-	}
+	for (tid = 0; tid <= IEEE80211_NUM_TIDS; tid++) {
+		ret = ath11k_peer_rx_tid_setup(ar, addr, vdev_id,
+					       tid, 1, 0, HAL_PN_TYPE_NONE);
+		if (ret) {
+			ath11k_warn(ab, "failed to setup rxd tid queue for tid %d: %d\n",
+				    tid, ret);
+			goto peer_clean;
+		}
+ 	}
 
 	ret = ath11k_peer_rx_frag_setup(ar, addr, vdev_id);
 	if (ret) {
@@ -81,6 +77,23 @@
 	/* TODO: Setup other peer specific resource used in data path */
 
 	return 0;
+
+peer_clean:
+	spin_lock_bh(&ab->base_lock);
+
+	peer = ath11k_peer_find(ab, vdev_id, addr);
+	if (!peer) {
+		ath11k_warn(ab, "failed to find the peer to del rx tid\n");
+		spin_unlock_bh(&ab->base_lock);
+		return -ENOENT;
+	}
+
+	for (; tid >= 0; tid--)
+		ath11k_peer_rx_tid_delete(ar, peer, tid);
+
+	spin_unlock_bh(&ab->base_lock);
+
+	return ret;
 }
 
 void ath11k_dp_srng_cleanup(struct ath11k_base *ab, struct dp_srng *ring)
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp.h
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp.h	2019-09-25 11:21:28.083337748 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp.h	2019-09-25 11:21:28.079337741 +0530
@@ -182,7 +182,7 @@
 #define DP_RX_RELEASE_RING_SIZE		1024
 #define DP_REO_EXCEPTION_RING_SIZE	128
 #define DP_REO_CMD_RING_SIZE		128
-#define DP_REO_STATUS_RING_SIZE		256
+#define DP_REO_STATUS_RING_SIZE		2048
 #define DP_RXDMA_BUF_RING_SIZE		4096
 #define DP_RXDMA_REFILL_RING_SIZE	2048
 #define DP_RXDMA_ERR_DST_RING_SIZE	1024
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp_rx.c	2019-09-25 11:21:28.083337748 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c	2019-09-25 11:21:28.079337741 +0530
@@ -733,8 +733,8 @@
 	__skb_queue_purge(&rx_tid->rx_frags);
 }
 
-static void ath11k_peer_rx_tid_delete(struct ath11k *ar,
-				      struct ath11k_peer *peer, u8 tid)
+void ath11k_peer_rx_tid_delete(struct ath11k *ar,
+			       struct ath11k_peer *peer, u8 tid)
 {
 	struct ath11k_hal_reo_cmd cmd = {0};
 	struct dp_rx_tid *rx_tid = &peer->rx_tid[tid];
Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.h
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp_rx.h	2019-09-25 11:21:28.083337748 +0530
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.h	2019-09-25 11:21:28.079337741 +0530
@@ -67,6 +67,8 @@
 				       enum set_key_cmd key_cmd,
 				       struct ieee80211_key_conf *key);
 void ath11k_peer_rx_tid_cleanup(struct ath11k *ar, struct ath11k_peer *peer);
+void ath11k_peer_rx_tid_delete(struct ath11k *ar,
+			       struct ath11k_peer *peer, u8 tid);
 int ath11k_peer_rx_tid_setup(struct ath11k *ar, const u8 *peer_mac, int vdev_id,
 			     u8 tid, u32 ba_win_sz, u16 ssn,
 			     enum hal_pn_type pn_type);
