Index: backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20190802-4.4.60-50ca7786c26.orig/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ backports-20190802-4.4.60-50ca7786c26/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -3022,8 +3022,9 @@ int ath11k_dp_rx_process_mon_status(stru
 		peer = ath11k_peer_find_by_id(ab, ppdu_info.peer_id);
 
 		if (!peer || !peer->sta) {
-			ath11k_warn(ab, "failed to find the peer with peer_id %d\n",
-				    ppdu_info.peer_id);
+			ath11k_dbg(ab, ATH11K_DBG_DATA,
+				   "failed to find the peer with peer_id to update peer stats %d\n",
+				   ppdu_info.peer_id);
 			spin_unlock_bh(&ab->base_lock);
 			rcu_read_unlock();
 			dev_kfree_skb_any(skb);
