--- a/drivers/net/wireless/ath/ath11k/htc.c
+++ b/drivers/net/wireless/ath/ath11k/htc.c
@@ -96,23 +96,6 @@ int ath11k_htc_send(struct ath11k_htc *h
 
 	skb_push(skb, sizeof(struct ath11k_htc_hdr));
 
-	if (ep->tx_credit_flow_enabled) {
-		credits = DIV_ROUND_UP(skb->len, htc->target_credit_size);
-		spin_lock_bh(&htc->tx_lock);
-		if (ep->tx_credits < credits) {
-			ath11k_dbg(ab, ATH11K_DBG_HTC,
-				   "htc insufficient credits ep %d required %d available %d\n",
-				   eid, credits, ep->tx_credits);
-			spin_unlock_bh(&htc->tx_lock);
-			ret = -EAGAIN;
-			goto err_pull;
-		}
-		ep->tx_credits -= credits;
-		ath11k_dbg(ab, ATH11K_DBG_HTC,
-			   "htc ep %d consumed %d credits (total %d)\n",
-			   eid, credits, ep->tx_credits);
-		spin_unlock_bh(&htc->tx_lock);
-	}
 
 	ath11k_htc_prepare_tx_skb(ep, skb);
 
@@ -121,7 +104,6 @@ int ath11k_htc_send(struct ath11k_htc *h
 	ret = dma_mapping_error(dev, skb_cb->paddr);
 	if (ret) {
 		ret = -EIO;
-		goto err_credits;
 	}
 
 	ret = ath11k_ce_send(htc->ab, skb, ep->ul_pipe_id, ep->eid);
@@ -132,57 +114,11 @@ int ath11k_htc_send(struct ath11k_htc *h
 
 err_unmap:
 	dma_unmap_single(dev, skb_cb->paddr, skb->len, DMA_TO_DEVICE);
-err_credits:
-	if (ep->tx_credit_flow_enabled) {
-		spin_lock_bh(&htc->tx_lock);
-		ep->tx_credits += credits;
-		ath11k_dbg(ab, ATH11K_DBG_HTC,
-			   "htc ep %d reverted %d credits back (total %d)\n",
-			   eid, credits, ep->tx_credits);
-		spin_unlock_bh(&htc->tx_lock);
-
-		if (ep->ep_ops.ep_tx_credits)
-			ep->ep_ops.ep_tx_credits(htc->ab);
-	}
 err_pull:
 	skb_pull(skb, sizeof(struct ath11k_htc_hdr));
 	return ret;
 }
 
-static void
-ath11k_htc_process_credit_report(struct ath11k_htc *htc,
-				 const struct ath11k_htc_credit_report *report,
-				 int len,
-				 enum ath11k_htc_ep_id eid)
-{
-	struct ath11k_base *ab = htc->ab;
-	struct ath11k_htc_ep *ep;
-	int i, n_reports;
-
-	if (len % sizeof(*report))
-		ath11k_warn(ab, "Uneven credit report len %d", len);
-
-	n_reports = len / sizeof(*report);
-
-	spin_lock_bh(&htc->tx_lock);
-	for (i = 0; i < n_reports; i++, report++) {
-		if (report->eid >= ATH11K_HTC_EP_COUNT)
-			break;
-
-		ep = &htc->endpoint[report->eid];
-		ep->tx_credits += report->credits;
-
-		ath11k_dbg(ab, ATH11K_DBG_HTC, "htc ep %d got %d credits (total %d)\n",
-			   report->eid, report->credits, ep->tx_credits);
-
-		if (ep->ep_ops.ep_tx_credits) {
-			spin_unlock_bh(&htc->tx_lock);
-			ep->ep_ops.ep_tx_credits(htc->ab);
-			spin_lock_bh(&htc->tx_lock);
-		}
-	}
-	spin_unlock_bh(&htc->tx_lock);
-}
 
 static int ath11k_htc_process_trailer(struct ath11k_htc *htc,
 				      u8 *buffer,
@@ -210,25 +146,6 @@ static int ath11k_htc_process_trailer(st
 			break;
 		}
 
-		switch (record->hdr.id) {
-		case ATH11K_HTC_RECORD_CREDITS:
-			len = sizeof(struct ath11k_htc_credit_report);
-			if (record->hdr.len < len) {
-				ath11k_warn(ab, "Credit report too long\n");
-				status = -EINVAL;
-				break;
-			}
-			ath11k_htc_process_credit_report(htc,
-							 record->credit_report,
-							 record->hdr.len,
-							 src_eid);
-			break;
-		default:
-			ath11k_warn(ab, "Unhandled record: id:%d length:%d\n",
-				    record->hdr.id, record->hdr.len);
-			break;
-		}
-
 		if (status)
 			break;
 
@@ -543,13 +460,12 @@ int ath11k_htc_connect_service(struct at
 	unsigned int max_msg_size = 0;
 	int length, status;
 	unsigned long time_left;
-	bool disable_credit_flow_ctrl = false;
+	bool disable_credit_flow_ctrl = true;
 	u16 message_id, service_id, flags = 0;
 	u8 tx_alloc = 0;
 
 	/* special case for HTC pseudo control service */
 	if (conn_req->service_id == ATH11K_HTC_SVC_ID_RSVD_CTRL) {
-		disable_credit_flow_ctrl = true;
 		assigned_eid = ATH11K_HTC_EP_0;
 		max_msg_size = ATH11K_HTC_MAX_CTRL_MSG_LEN;
 		memset(&resp_msg_dummy, 0, sizeof(resp_msg_dummy));
@@ -578,14 +494,7 @@ int ath11k_htc_connect_service(struct at
 					 ATH11K_HTC_MSG_CONNECT_SERVICE_ID);
 
 	flags |= FIELD_PREP(ATH11K_HTC_CONN_FLAGS_RECV_ALLOC, tx_alloc);
-
-	/* Only enable credit flow control for WMI ctrl service */
-	if (!(conn_req->service_id == ATH11K_HTC_SVC_ID_WMI_CONTROL ||
-	      conn_req->service_id == ATH11K_HTC_SVC_ID_WMI_CONTROL_MAC1 ||
-	      conn_req->service_id == ATH11K_HTC_SVC_ID_WMI_CONTROL_MAC2)) {
-		flags |= ATH11K_HTC_CONN_FLAGS_DISABLE_CREDIT_FLOW_CTRL;
-		disable_credit_flow_ctrl = true;
-	}
+	flags |= ATH11K_HTC_CONN_FLAGS_DISABLE_CREDIT_FLOW_CTRL;
 
 	req_msg->flags_len = FIELD_PREP(HTC_SVC_MSG_CONNECTIONFLAGS, flags);
 	req_msg->msg_svc_id |= FIELD_PREP(HTC_SVC_MSG_SERVICE_ID,
@@ -712,6 +621,7 @@ int ath11k_htc_start(struct ath11k_htc *
 	msg->msg_id = FIELD_PREP(HTC_MSG_MESSAGEID,
 				 ATH11K_HTC_MSG_SETUP_COMPLETE_EX_ID);
 
+	msg->flags |= ATH11K_GLOBAL_DISABLE_CREDIT_FLOW;
 	ath11k_dbg(ab, ATH11K_DBG_HTC, "HTC is using TX credit flow control\n");
 
 	status = ath11k_htc_send(htc, ATH11K_HTC_EP_0, skb);
--- a/drivers/net/wireless/ath/ath11k/wmi.c
+++ b/drivers/net/wireless/ath/ath11k/wmi.c
@@ -267,20 +267,7 @@ int ath11k_wmi_cmd_send(struct ath11k_pd
 	struct ath11k_wmi_base *wmi_sc = wmi->wmi_ab;
 	int ret = -EOPNOTSUPP;
 
-	might_sleep();
-
-	wait_event_timeout(wmi_sc->tx_credits_wq, ({
-		ret = ath11k_wmi_cmd_send_nowait(wmi, skb, cmd_id);
-
-		if (ret && test_bit(ATH11K_FLAG_CRASH_FLUSH, &wmi_sc->ab->dev_flags))
-			ret = -ESHUTDOWN;
-
-		(ret != -EAGAIN);
-	}), WMI_SEND_TIMEOUT_HZ);
-
-	if (ret == -EAGAIN)
-		ath11k_warn(wmi_sc->ab, "wmi command %d timeout\n", cmd_id);
-
+	ret = ath11k_wmi_cmd_send_nowait(wmi, skb, cmd_id);
 	return ret;
 }
 
--- a/drivers/net/wireless/ath/ath11k/htc.h
+++ b/drivers/net/wireless/ath/ath11k/htc.h
@@ -114,6 +114,8 @@ struct ath11k_htc_conn_svc_resp {
 	u32 svc_meta_pad;
 } __packed;
 
+#define ATH11K_GLOBAL_DISABLE_CREDIT_FLOW BIT(1)
+
 struct ath11k_htc_setup_complete_extended {
 	u32 msg_id;
 	u32 flags;
