Index: backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/dp_rx.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ backports-20191227-4.4.60-b805aa84b44f/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -2422,9 +2422,12 @@ static void ath11k_dp_rx_h_ppdu(struct a
 	} else if (channel_num >= 36 && channel_num <= 173) {
 		rx_status->band = NL80211_BAND_5GHZ;
 	} else {
-		ath11k_warn(ar->ab, "Unsupported Channel info received %d\n",
-			    channel_num);
-		return;
+		spin_lock_bh(&ar->data_lock);
+		rx_status->band = ar->rx_channel->band;
+		channel_num = ieee80211_frequency_to_channel(ar->rx_channel->center_freq);
+		spin_unlock_bh(&ar->data_lock);
+		ath11k_dbg_dump(ar->ab, ATH11K_DBG_DP_RX, NULL, "rx_desc: ",
+				rx_desc, sizeof(struct hal_rx_desc));
 	}
 
 	rx_status->freq = ieee80211_channel_to_frequency(channel_num,
