--- a/drivers/net/wireless/ath/ath11k/peer.c
+++ b/drivers/net/wireless/ath/ath11k/peer.c
@@ -26,6 +26,25 @@ struct ath11k_peer *ath11k_peer_find(str
 	return NULL;
 }
 
+struct ath11k_peer *ath11k_peer_find_by_pdev_idx(struct ath11k_base *ab, u8 pdev_idx,
+						 const u8 *addr)
+{
+	struct ath11k_peer *peer;
+
+	lockdep_assert_held(&ab->base_lock);
+
+	list_for_each_entry(peer, &ab->peers, list) {
+		if (peer->pdev_idx != pdev_idx)
+			continue;
+		if (memcmp(peer->addr, addr, ETH_ALEN))
+			continue;
+
+		return peer;
+	}
+
+	return NULL;
+}
+
 struct ath11k_peer *ath11k_peer_find_by_addr(struct ath11k_base *ab,
 					     const u8 *addr)
 {
@@ -225,21 +244,15 @@ int ath11k_peer_create(struct ath11k *ar
 
 	spin_lock_bh(&ar->ab->base_lock);
 
-	peer = ath11k_peer_find_by_addr(ar->ab, param->peer_addr);
+
+	peer = ath11k_peer_find_by_pdev_idx(ar->ab, ar->pdev_idx, param->peer_addr);
 	if(peer) {
 		ath11k_warn(ar->ab,
-			    "peer already created for %pM on vdev %i and trying to create on vdev %i\n",
-			    param->peer_addr, peer->vdev_id, param->vdev_id);
-
-		peer_ar = ath11k_mac_get_ar_by_vdev_id(ar->ab, peer->vdev_id);
-		if(peer_ar == ar){
-			ath11k_warn(ar->ab,
-				    "ignoring the peer %pM creation on same pdev %i \n",
-				    param->peer_addr, ar->pdev->pdev_id);
+				"ignoring the peer %pM creation on same pdev idx %d \n",
+				param->peer_addr, ar->pdev_idx);
 
-			spin_unlock_bh(&ar->ab->base_lock);
-			return -EINVAL;
-		}
+		spin_unlock_bh(&ar->ab->base_lock);
+		return -EINVAL;
 	}
 
 	spin_unlock_bh(&ar->ab->base_lock);
@@ -279,6 +292,7 @@ int ath11k_peer_create(struct ath11k *ar
 		return -ENOENT;
 	}
 
+	peer->pdev_idx = ar->pdev_idx;
 	peer->sta = sta;
 	arvif->ast_hash = peer->ast_hash;
 
--- a/drivers/net/wireless/ath/ath11k/peer.h
+++ b/drivers/net/wireless/ath/ath11k/peer.h
@@ -13,6 +13,7 @@ struct ath11k_peer {
 	u8 addr[ETH_ALEN];
 	int peer_id;
 	u16 ast_hash;
+	u8 pdev_idx;
 
 	/* protected by ab->data_lock */
 	struct ieee80211_key_conf *keys[WMI_MAX_KEY_INDEX + 1];
@@ -31,6 +32,8 @@ void ath11k_peer_map_event(struct ath11k
 			   u8 *mac_addr, u16 ast_hash);
 struct ath11k_peer *ath11k_peer_find(struct ath11k_base *ab, int vdev_id,
 				     const u8 *addr);
+struct ath11k_peer *ath11k_peer_find_by_pdev_idx(struct ath11k_base *ab, u8 pdev_idx,
+						 const u8 *addr);
 struct ath11k_peer *ath11k_peer_find_by_addr(struct ath11k_base *ab,
 					     const u8 *addr);
 struct ath11k_peer *ath11k_peer_find_by_id(struct ath11k_base *ab, int peer_id);
