diff --git a/drivers/net/wireless/ath/ath11k/dp_rx.c b/drivers/net/wireless/ath/ath11k/dp_rx.c
index ae3bf0e1171f..9697eacde84c 100644
--- a/drivers/net/wireless/ath/ath11k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath11k/dp_rx.c
@@ -1377,8 +1377,7 @@ ath11k_update_per_peer_tx_stats(struct ath11k *ar,
 	/* PPDU stats reported for mgmt packet doesn't have valid tx bytes.
 	 * So skip peer stats update for mgmt packets.
 	 */
-	if (tid < HTT_PPDU_STATS_NON_QOS_TID || (peer_stats->ppdu_type
-	     <= HTT_PPDU_STATS_PPDU_TYPE_MU_MIMO_OFDMA)) {
+	if (tid < HTT_PPDU_STATS_NON_QOS_TID) {
 		memset(peer_stats, 0, sizeof(*peer_stats));
 		peer_stats->succ_pkts = succ_pkts;
 		peer_stats->succ_bytes = succ_bytes;
