Index: backports-20191227-4.4.60-b805aa84b44f/net/mac80211/key.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/net/mac80211/key.c
+++ backports-20191227-4.4.60-b805aa84b44f/net/mac80211/key.c
@@ -211,7 +211,11 @@ static int ieee80211_key_enable_hw_accel
 			  key->conf.keyidx,
 			  sta ? sta->sta.addr : bcast_addr, ret);
 
-	if (sdata->hw_80211_encap)
+	if (sdata->hw_80211_encap &&
+	    (key->conf.cipher != WLAN_CIPHER_SUITE_AES_CMAC &&
+	     key->conf.cipher != WLAN_CIPHER_SUITE_BIP_GMAC_128 &&
+	     key->conf.cipher != WLAN_CIPHER_SUITE_BIP_GMAC_256 &&
+	     key->conf.cipher != WLAN_CIPHER_SUITE_BIP_CMAC_256))
 		return -EINVAL;
 
  out_unsupported:
