Index: backports-20191227-4.4.60-b805aa84b44f/net/mac80211/key.c
===================================================================
--- backports-20191227-4.4.60-b805aa84b44f.orig/net/mac80211/key.c	2020-01-24 14:58:28.840501617 +0530
+++ backports-20191227-4.4.60-b805aa84b44f/net/mac80211/key.c	2020-01-24 15:03:15.296501012 +0530
@@ -208,7 +208,11 @@
 			  key->conf.keyidx,
 			  sta ? sta->sta.addr : bcast_addr, ret);
 
-	if (sdata->hw_80211_encap)
+	if (sdata->hw_80211_encap &&
+	    (key->conf.cipher != WLAN_CIPHER_SUITE_AES_CMAC &&
+	     key->conf.cipher != WLAN_CIPHER_SUITE_BIP_GMAC_128 &&
+	     key->conf.cipher != WLAN_CIPHER_SUITE_BIP_GMAC_256 &&
+	     key->conf.cipher != WLAN_CIPHER_SUITE_BIP_CMAC_256))
 		return -EINVAL;
 
  out_unsupported:
