diff --git a/drivers/net/wireless/ath/ath11k/debug_htt_stats.c b/drivers/net/wireless/ath/ath11k/debug_htt_stats.c
index 1b17cee176c3..13a8f402073f 100644
--- a/drivers/net/wireless/ath/ath11k/debug_htt_stats.c
+++ b/drivers/net/wireless/ath/ath11k/debug_htt_stats.c
@@ -4518,7 +4518,9 @@ out:
 
 static int ath11k_release_htt_stats(struct inode *inode, struct file *file)
 {
+	struct ath11k *ar = inode->i_private;
 	vfree(file->private_data);
+	ar->debug.htt_stats.stats_req = NULL;
 	return 0;
 }
 
-- 
2.1.2

