--- a/drivers/net/wireless/ath/ath11k/wmi.c
+++ b/drivers/net/wireless/ath/ath11k/wmi.c
@@ -4577,8 +4577,10 @@ static int wmi_process_mgmt_tx_comp(stru
 
 	ieee80211_tx_status_irqsafe(ar->hw, msdu);
 
-	WARN_ON_ONCE(atomic_read(&ar->num_pending_mgmt_tx) == 0);
-	atomic_dec(&ar->num_pending_mgmt_tx);
+	if (atomic_read(&ar->num_pending_mgmt_tx))
+		atomic_dec(&ar->num_pending_mgmt_tx);
+	else
+		WARN_ON_ONCE(1);
 
 	return 0;
 }
