Index: backports-20191108-4.4.60-cdd7d0b55972/drivers/net/wireless/ath/ath11k/core.h
===================================================================
--- backports-20191108-4.4.60-cdd7d0b55972.orig/drivers/net/wireless/ath/ath11k/core.h	2019-12-27 17:40:26.393682176 +0530
+++ backports-20191108-4.4.60-cdd7d0b55972/drivers/net/wireless/ath/ath11k/core.h	2019-12-27 17:41:26.049682050 +0530
@@ -450,6 +450,7 @@
 	u32 pktlog_mode;
 	u32 pktlog_peer_valid;
 	u8 pktlog_peer_addr[ETH_ALEN];
+	u32 rx_filter;
 #ifdef CPTCFG_ATH11K_PKTLOG
 	struct dentry *debugfs_pktlog;
 	struct ath_pktlog pktlog;
Index: backports-20191108-4.4.60-cdd7d0b55972/drivers/net/wireless/ath/ath11k/debug.c
===================================================================
--- backports-20191108-4.4.60-cdd7d0b55972.orig/drivers/net/wireless/ath/ath11k/debug.c	2019-12-27 17:40:26.321682177 +0530
+++ backports-20191108-4.4.60-cdd7d0b55972/drivers/net/wireless/ath/ath11k/debug.c	2019-12-27 17:42:03.941681970 +0530
@@ -935,6 +935,8 @@
 		tlv_filter = ath11k_mac_mon_status_filter_default;
 	}
 
+	ar->debug.rx_filter = tlv_filter.rx_filter;
+
 	ring_id = ar->dp.rx_mon_status_refill_ring.refill_buf_ring.ring_id;
 	ret = ath11k_dp_tx_htt_rx_filter_setup(ar->ab, ring_id, ar->dp.mac_id,
 					       HAL_RXDMA_MONITOR_STATUS,
Index: backports-20191108-4.4.60-cdd7d0b55972/drivers/net/wireless/ath/ath11k/mac.c
===================================================================
--- backports-20191108-4.4.60-cdd7d0b55972.orig/drivers/net/wireless/ath/ath11k/mac.c	2019-12-27 17:40:26.393682176 +0530
+++ backports-20191108-4.4.60-cdd7d0b55972/drivers/net/wireless/ath/ath11k/mac.c	2019-12-27 17:43:02.197681847 +0530
@@ -4289,8 +4289,11 @@
 	struct htt_rx_ring_tlv_filter tlv_filter = {0};
 	u32 ring_id;
 
-	if (enable)
+	if (enable) {
 		tlv_filter = ath11k_mac_mon_status_filter_default;
+		if (ar->debug.extd_rx_stats)
+			tlv_filter.rx_filter = ar->debug.rx_filter;
+	}
 
 	ring_id = ar->dp.rx_mon_status_refill_ring.refill_buf_ring.ring_id;
 
