--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -1069,10 +1069,10 @@ err_mon_del:
 
 static int ath11k_mac_setup_bcn_tmpl(struct ath11k_vif *arvif)
 {
-	struct ath11k *ar = arvif->ar;
-	struct ath11k_base *ab = ar->ab;
-	struct ieee80211_hw *hw = ar->hw;
-	struct ieee80211_vif *vif = arvif->vif;
+	struct ath11k *ar;
+	struct ath11k_base *ab;
+	struct ieee80211_hw *hw;
+	struct ieee80211_vif *vif, *parent;
 	struct ieee80211_mutable_offsets offs = {};
 	struct sk_buff *bcn;
 	struct ieee80211_mgmt *mgmt;
@@ -1082,6 +1082,17 @@ static int ath11k_mac_setup_bcn_tmpl(str
 	if (arvif->vdev_type != WMI_VDEV_TYPE_AP)
 		return 0;
 
+	if (ieee80211_get_multi_bssid_mode(arvif->vif) ==
+	    NL80211_MULTIPLE_BSSID_NON_TRANSMITTED) {
+		parent = ieee80211_get_multi_bssid_parent(arvif->vif);
+		arvif = (struct ath11k_vif *)parent->drv_priv;
+	}
+
+	ar = arvif->ar;
+	ab = ar->ab;
+	hw = ar->hw;
+	vif = arvif->vif;
+
 	bcn = ieee80211_beacon_get_template(hw, vif, &offs);
 	if (!bcn) {
 		ath11k_warn(ab, "failed to get beacon template from mac80211\n");
--- a/drivers/net/wireless/ath/ath11k/wmi.h
+++ b/drivers/net/wireless/ath/ath11k/wmi.h
@@ -3549,6 +3549,11 @@ struct wmi_bcn_tmpl_cmd {
 	u32 csa_event_bitmap;
 	u32 mbssid_ie_offset;
 	u32 esp_ie_offset;
+	u32 csc_switch_count_offset;
+	u32 csc_event_bitmap;
+	u32 mu_edca_ie_offset;
+	u32 feature_enable_bitmap;
+	u32 ema_params;
 } __packed;
 
 struct wmi_key_seq_counter {
