--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -3057,6 +3057,7 @@ static void ath10k_core_register_work(struct work_struct *work)
 
 	/* peer stats are enabled by default */
 	set_bit(ATH10K_FLAG_PEER_STATS, &ar->dev_flags);
+	ar->debug.enable_extd_tx_stats = true;
 
 	status = ath10k_core_probe_fw(ar);
 	if (status) {
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -3048,14 +3048,14 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	u8 rate = 0, sgi;
 	s8 rate_idx = 0;
 	bool skip_auto_rate;
-	struct rate_info txrate;
+	struct rate_info *txrate = &arsta->txrate;
 
 	lockdep_assert_held(&ar->data_lock);
 
-	txrate.flags = ATH10K_HW_PREAMBLE(peer_stats->ratecode);
-	txrate.bw = ATH10K_HW_BW(peer_stats->flags);
-	txrate.nss = ATH10K_HW_NSS(peer_stats->ratecode);
-	txrate.mcs = ATH10K_HW_MCS_RATE(peer_stats->ratecode);
+	txrate->flags = ATH10K_HW_PREAMBLE(peer_stats->ratecode);
+	txrate->bw = ATH10K_HW_BW(peer_stats->flags);
+	txrate->nss = ATH10K_HW_NSS(peer_stats->ratecode);
+	txrate->mcs = ATH10K_HW_MCS_RATE(peer_stats->ratecode);
 	sgi = ATH10K_HW_GI(peer_stats->flags);
 	skip_auto_rate = ATH10K_FW_SKIPPED_RATE_CTRL(peer_stats->flags);
 
@@ -3065,47 +3065,46 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	if (skip_auto_rate)
 		return;
 
-	if (txrate.flags == WMI_RATE_PREAMBLE_VHT && txrate.mcs > 9) {
-		ath10k_warn(ar, "Invalid VHT mcs %hhd peer stats",  txrate.mcs);
+	if (txrate->flags == WMI_RATE_PREAMBLE_VHT && txrate->mcs > 9) {
+		ath10k_warn(ar, "Invalid VHT mcs %hhd peer stats",  txrate->mcs);
 		return;
 	}
 
-	if (txrate.flags == WMI_RATE_PREAMBLE_HT &&
-	    (txrate.mcs > 7 || txrate.nss < 1)) {
+	if (txrate->flags == WMI_RATE_PREAMBLE_HT &&
+	    (txrate->mcs > 7 || txrate->nss < 1)) {
 		ath10k_warn(ar, "Invalid HT mcs %hhd nss %hhd peer stats",
-			    txrate.mcs, txrate.nss);
+			    txrate->mcs, txrate->nss);
 		return;
 	}
 
-	memset(&arsta->txrate, 0, sizeof(arsta->txrate));
 	memset(&arsta->tx_info.status, 0, sizeof(arsta->tx_info.status));
-	if (txrate.flags == WMI_RATE_PREAMBLE_CCK ||
-	    txrate.flags == WMI_RATE_PREAMBLE_OFDM) {
+	if (txrate->flags == WMI_RATE_PREAMBLE_CCK ||
+	    txrate->flags == WMI_RATE_PREAMBLE_OFDM) {
 		rate = ATH10K_HW_LEGACY_RATE(peer_stats->ratecode);
 		/* This is hacky, FW sends CCK rate 5.5Mbps as 6 */
-		if (rate == 6 && txrate.flags == WMI_RATE_PREAMBLE_CCK)
+		if (rate == 6 && txrate->flags == WMI_RATE_PREAMBLE_CCK)
 			rate = 5;
-		rate_idx = ath10k_get_legacy_rate_idx(ar, rate);
-		if (rate_idx < 0)
-			return;
-		arsta->txrate.legacy = rate;
-	} else if (txrate.flags == WMI_RATE_PREAMBLE_HT) {
+		arsta->txrate.legacy = ath10k_get_legacy_rate_idx(ar, rate);
+	} else if (txrate->flags == WMI_RATE_PREAMBLE_HT) {
 		arsta->txrate.flags = RATE_INFO_FLAGS_MCS;
-		arsta->txrate.mcs = txrate.mcs + 8 * (txrate.nss - 1);
+		arsta->txrate.mcs = txrate->mcs + 8 * (txrate->nss - 1);
 	} else {
 		arsta->txrate.flags = RATE_INFO_FLAGS_VHT_MCS;
-		arsta->txrate.mcs = txrate.mcs;
+		arsta->txrate.mcs = txrate->mcs;
 	}
 
-	switch (txrate.flags) {
+	arsta->txrate.bw = ath10k_bw_to_mac80211_bw(txrate->bw);
+	arsta->tx_info.status.rates[0].count = 1;
+
+	switch (txrate->flags) {
 	case WMI_RATE_PREAMBLE_OFDM:
 		if (arsta->arvif && arsta->arvif->vif)
 			conf = rcu_dereference(arsta->arvif->vif->chanctx_conf);
 		if (conf && conf->def.chan->band == NL80211_BAND_5GHZ)
-			arsta->tx_info.status.rates[0].idx = rate_idx - 4;
+			arsta->tx_info.status.rates[0].idx = txrate->legacy - 4;
 		break;
 	case WMI_RATE_PREAMBLE_CCK:
-		arsta->tx_info.status.rates[0].idx = rate_idx;
+		arsta->tx_info.status.rates[0].idx = txrate->legacy;
 		if (sgi)
 			arsta->tx_info.status.rates[0].flags |=
 				(IEEE80211_TX_RC_USE_SHORT_PREAMBLE |
@@ -3113,7 +3112,7 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 		break;
 	case WMI_RATE_PREAMBLE_HT:
 		arsta->tx_info.status.rates[0].idx =
-				txrate.mcs + ((txrate.nss - 1) * 8);
+				txrate->mcs + ((txrate->nss - 1) * 8);
 		if (sgi)
 			arsta->tx_info.status.rates[0].flags |=
 					IEEE80211_TX_RC_SHORT_GI;
@@ -3121,7 +3120,7 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 		break;
 	case WMI_RATE_PREAMBLE_VHT:
 		ieee80211_rate_set_vht(&arsta->tx_info.status.rates[0],
-				       txrate.mcs, txrate.nss);
+				       txrate->mcs, txrate->nss);
 		if (sgi)
 			arsta->tx_info.status.rates[0].flags |=
 						IEEE80211_TX_RC_SHORT_GI;
@@ -3129,8 +3128,6 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 		break;
 	}
 
-	arsta->txrate.nss = txrate.nss;
-	arsta->txrate.bw = ath10k_bw_to_mac80211_bw(txrate.bw);
 	arsta->last_tx_bitrate = cfg80211_calculate_bitrate(&arsta->txrate);
 	if (sgi)
 		arsta->txrate.flags |= RATE_INFO_FLAGS_SHORT_GI;
@@ -3164,6 +3161,7 @@ static void ath10k_htt_fetch_peer_stats(struct ath10k *ar,
 	struct ath10k_per_peer_tx_stats *p_tx_stats = &ar->peer_tx_stats;
 	struct htt_per_peer_tx_stats_ind *tx_stats;
 	struct ieee80211_sta *sta;
+	struct ath10k_sta *arsta;
 	struct ath10k_peer *peer;
 	int peer_id, i;
 	u8 ppdu_len, num_ppdu;
@@ -3190,6 +3188,7 @@ static void ath10k_htt_fetch_peer_stats(struct ath10k *ar,
 	}
 
 	sta = peer->sta;
+	arsta = (struct ath10k_sta *)sta->drv_priv;
 	for (i = 0; i < num_ppdu; i++) {
 		tx_stats = (struct htt_per_peer_tx_stats_ind *)
 			   (resp->peer_tx_stats.payload + i * ppdu_len);
@@ -3206,6 +3205,9 @@ static void ath10k_htt_fetch_peer_stats(struct ath10k *ar,
 		p_tx_stats->duration = __le16_to_cpu(tx_stats->tx_duration);
 
 		ath10k_update_per_peer_tx_stats(ar, sta, p_tx_stats);
+		if (ar->debug.enable_extd_tx_stats)
+			arsta->tx_stats->tx_duration +=
+					__le16_to_cpu(tx_stats->tx_duration);
 	}
 
 out:
@@ -3219,6 +3221,7 @@ static void ath10k_fetch_10_2_tx_stats(struct ath10k *ar, u8 *data)
 	struct ath10k_per_peer_tx_stats *p_tx_stats = &ar->peer_tx_stats;
 	struct ath10k_10_2_peer_tx_stats *tx_stats;
 	struct ieee80211_sta *sta;
+	struct ath10k_sta *arsta;
 	struct ath10k_peer *peer;
 	u16 log_type = __le16_to_cpu(hdr->log_type);
 	u32 peer_id = 0, i;
@@ -3244,6 +3247,7 @@ static void ath10k_fetch_10_2_tx_stats(struct ath10k *ar, u8 *data)
 	}
 
 	sta = peer->sta;
+	arsta = (struct ath10k_sta *)sta->drv_priv;
 	for (i = 0; i < tx_stats->tx_ppdu_cnt; i++) {
 		p_tx_stats->succ_bytes =
 			__le16_to_cpu(tx_stats->success_bytes[i]);
@@ -3259,6 +3263,10 @@ static void ath10k_fetch_10_2_tx_stats(struct ath10k *ar, u8 *data)
 
 		ath10k_update_per_peer_tx_stats(ar, sta, p_tx_stats);
 	}
+
+	if (ar->debug.enable_extd_tx_stats)
+		arsta->tx_stats->tx_duration +=
+				 __le32_to_cpu(tx_stats->tx_duration);
 	spin_unlock_bh(&ar->data_lock);
 	rcu_read_unlock();
 

--- a/drivers/net/wireless/ath/ath10k/txrx.c
+++ b/drivers/net/wireless/ath/ath10k/txrx.c
@@ -61,6 +61,7 @@ int ath10k_txrx_tx_unref(struct ath10k_htt *htt,
 	struct ath10k_skb_cb *skb_cb;
 	struct ath10k_txq *artxq;
 	struct sk_buff *msdu;
+	struct ieee80211_sta *sta = NULL;
 
 	ath10k_dbg(ar, ATH10K_DBG_HTT,
 		   "htt tx completion msdu_id %u status %d\n",
@@ -87,6 +89,7 @@ int ath10k_txrx_tx_unref(struct ath10k_htt *htt,
 	if (txq) {
 		artxq = (void *)txq->drv_priv;
 		artxq->num_fw_queued--;
+		sta = txq->sta;
 	}
 
 	ath10k_htt_tx_free_msdu_id(htt, tx_done->msdu_id);
