--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -3721,8 +3721,12 @@ static void ath10k_mac_tx_push(struct ie
 
 	ret = ath10k_mac_tx_push_txq(hw, txq);
 	if (unlikely(ret < 0)) {
-		ath10k_warn(ar, "failed to push tx to %pM tid %hhu: %d\n",
-			    txq->sta->addr, txq->tid, ret);
+		if (txq->sta)
+			ath10k_warn(ar, "failed to push tx to %pM tid %hhu: %d\n",
+				    txq->sta->addr, txq->tid, ret);
+		else
+			ath10k_warn(ar, "failed to push tx on tid %hhu: %d\n",
+				    txq->tid, ret);
 		return;
 	}
 
