--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -2702,15 +2702,16 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	struct ieee80211_chanctx_conf *conf = NULL;
 	u8 rate = 0, sgi;
 	struct rate_info *txrate = &arsta->txrate;
+	struct rate_info txrate_update;
 	u32 tx_bitrate;
 	bool skip_auto_rate;
 
 	lockdep_assert_held(&ar->data_lock);
 
-	txrate->flags = ATH10K_HW_PREAMBLE(peer_stats->ratecode);
-	txrate->bw = ATH10K_HW_BW(peer_stats->flags);
-	txrate->nss = ATH10K_HW_NSS(peer_stats->ratecode);
-	txrate->mcs = ATH10K_HW_MCS_RATE(peer_stats->ratecode);
+	txrate_update.flags = ATH10K_HW_PREAMBLE(peer_stats->ratecode);
+	txrate_update.bw = ATH10K_HW_BW(peer_stats->flags);
+	txrate_update.nss = ATH10K_HW_NSS(peer_stats->ratecode);
+	txrate_update.mcs = ATH10K_HW_MCS_RATE(peer_stats->ratecode);
 	sgi = ATH10K_HW_GI(peer_stats->flags);
 	skip_auto_rate = ATH10K_FW_SKIPPED_RATE_CTRL(peer_stats->flags);
 
@@ -2720,18 +2721,23 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	if (skip_auto_rate)
 		return;
 
-	if (txrate->flags == WMI_RATE_PREAMBLE_VHT && txrate->mcs > 9) {
-		ath10k_warn(ar, "Invalid VHT mcs %hhd peer stats",  txrate->mcs);
+	if (txrate_update.flags == WMI_RATE_PREAMBLE_VHT && txrate_update.mcs > 9) {
+		ath10k_warn(ar, "Invalid VHT mcs %hhd peer stats",  txrate_update.mcs);
 		return;
 	}
 
-	if (txrate->flags == WMI_RATE_PREAMBLE_HT &&
-	    (txrate->mcs > 7 || txrate->nss < 1)) {
+	if (txrate_update.flags == WMI_RATE_PREAMBLE_HT &&
+	    (txrate_update.mcs > 7 || txrate_update.nss < 1)) {
 		ath10k_warn(ar, "Invalid HT mcs %hhd nss %hhd peer stats",
-			    txrate->mcs, txrate->nss);
+			    txrate_update.mcs, txrate_update.nss);
 		return;
 	}
 
+	txrate->flags = txrate_update.flags;
+	txrate->bw = txrate_update.bw;
+	txrate->nss = txrate_update.nss;
+	txrate->mcs = txrate_update.mcs;
+
 	if (txrate->flags == WMI_RATE_PREAMBLE_CCK ||
 	    txrate->flags == WMI_RATE_PREAMBLE_OFDM) {
 		rate = ATH10K_HW_LEGACY_RATE(peer_stats->ratecode);
