--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -29,8 +29,10 @@ static unsigned int ath10k_cryptmode_par
 static bool uart_print;
 static bool skip_otp;
 static bool rawmode;
+bool bug_on_fw_assert = true;
 
 unsigned long ath10k_coredump_mask = BIT(ATH10K_FW_CRASH_DUMP_REGISTERS) |
+				     BIT(ATH10K_FW_CRASH_DUMP_RAM_DATA) |
 				     BIT(ATH10K_FW_CRASH_DUMP_CE_DATA);
 
 /* FIXME: most of these should be readonly */
@@ -40,6 +42,7 @@ module_param(uart_print, bool, 0644);
 module_param(skip_otp, bool, 0644);
 module_param(rawmode, bool, 0644);
 module_param_named(coredump_mask, ath10k_coredump_mask, ulong, 0444);
+module_param(bug_on_fw_assert, bool, 0644);
 
 MODULE_PARM_DESC(debug_mask, "Debugging mask");
 MODULE_PARM_DESC(uart_print, "Uart target debugging");
@@ -47,6 +50,7 @@ MODULE_PARM_DESC(skip_otp, "Skip otp fai
 MODULE_PARM_DESC(cryptmode, "Crypto mode: 0-hardware, 1-software");
 MODULE_PARM_DESC(rawmode, "Use raw 802.11 frame datapath");
 MODULE_PARM_DESC(coredump_mask, "Bitfield of what to include in firmware crash file");
+MODULE_PARM_DESC(bug_on_fw_assert, "Bug on firmware assert");
 
 static const struct ath10k_hw_params ath10k_hw_params_list[] = {
 	{
--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -2674,3 +2674,29 @@ void ath10k_dbg_dump(struct ath10k *ar,
 EXPORT_SYMBOL(ath10k_dbg_dump);
 
 #endif /* CPTCFG_ATH10K_DEBUG */
+
+void ath10k_dbg_bug_on_fw_assert(struct ath10k *ar, u8 *buf)
+{
+	u8 *addr;
+	int len = ar->hw_values->fw_crash_dump_sz;
+
+	if (!bug_on_fw_assert || ar->stats.fw_crash_counter == 0 ||
+	    !ar->coredump.fw_crash_data->ramdump_buf)
+		return;
+
+	addr = (u8 *)ioremap(ar->hw_values->fw_crash_dump_addr, len);
+	if (!addr) {
+		ath10k_err(ar, "Cannot access memory to dump WLAN Firmware data\n");
+		return;
+	}
+
+	/* copy the dump collected */
+	memcpy(addr, buf, len);
+
+	ath10k_err(ar, "Firmware DRAM dump collected successfully\n");
+	iounmap(addr);
+
+	ath10k_err(ar, "XXX: Asserting Host...\n");
+	BUG();
+}
+EXPORT_SYMBOL(ath10k_dbg_bug_on_fw_assert);
--- a/drivers/net/wireless/ath/ath10k/debug.h
+++ b/drivers/net/wireless/ath/ath10k/debug.h
@@ -72,11 +72,22 @@ struct ath10k_pktlog_hdr {
 #define ATH10K_FW_STATS_BUF_SIZE (1024 * 1024)
 
 extern unsigned int ath10k_debug_mask;
+extern bool bug_on_fw_assert;
 
 __printf(2, 3) void ath10k_info(struct ath10k *ar, const char *fmt, ...);
 __printf(2, 3) void ath10k_err(struct ath10k *ar, const char *fmt, ...);
 __printf(2, 3) void ath10k_warn(struct ath10k *ar, const char *fmt, ...);
 
+#define        ATTR_SRAM_MEM   BIT(0)
+#define        ATTR_REG_SPACE  BIT(1)
+
+#define ATH10K_FW_IO_MEM_ADDR          0x41400000
+#define ATH10K_FW_IO_MEM_SIZE          ((1 * 1024 * 1024) - 128)
+
+#define ATH10K_FW_IO_MEM_REGION_2_ADDR 0x44000000
+#define IPQ40XX_FW_IO_MEM_ADDR         0x87500000
+#define ATH10K_FW_IO_MEM_REGION_2_SIZE 0x00600000 /* 6 mb */
+
 void ath10k_debug_print_hwfw_info(struct ath10k *ar);
 void ath10k_debug_print_board_info(struct ath10k *ar);
 void ath10k_debug_print_boot_info(struct ath10k *ar);
@@ -238,6 +249,7 @@ void ath10k_sta_update_rx_tid_stats_ampd
 {
 }
 #endif /* CPTCFG_MAC80211_DEBUGFS */
+void ath10k_dbg_bug_on_fw_assert(struct ath10k *ar, u8* buf);
 
 #ifdef CPTCFG_ATH10K_DEBUG
 __printf(3, 4) void ath10k_dbg(struct ath10k *ar,
--- a/drivers/net/wireless/ath/ath10k/hw.c
+++ b/drivers/net/wireless/ath/ath10k/hw.c
@@ -146,6 +146,8 @@ const struct ath10k_hw_values qca988x_va
 	.num_target_ce_config_wlan	= 7,
 	.ce_desc_meta_data_mask		= 0xFFFC,
 	.ce_desc_meta_data_lsb		= 2,
+	.fw_crash_dump_addr		= ATH10K_FW_IO_MEM_ADDR,
+	.fw_crash_dump_sz		= ATH10K_FW_IO_MEM_SIZE,
 };
 
 const struct ath10k_hw_values qca6174_values = {
@@ -164,6 +166,8 @@ const struct ath10k_hw_values qca99x0_va
 	.num_target_ce_config_wlan	= 10,
 	.ce_desc_meta_data_mask		= 0xFFF0,
 	.ce_desc_meta_data_lsb		= 4,
+	.fw_crash_dump_addr		= ATH10K_FW_IO_MEM_REGION_2_ADDR,
+	.fw_crash_dump_sz		= ATH10K_FW_IO_MEM_REGION_2_SIZE,
 };
 
 const struct ath10k_hw_values qca9888_values = {
@@ -173,6 +177,8 @@ const struct ath10k_hw_values qca9888_va
 	.num_target_ce_config_wlan	= 10,
 	.ce_desc_meta_data_mask		= 0xFFF0,
 	.ce_desc_meta_data_lsb		= 4,
+	.fw_crash_dump_addr		= ATH10K_FW_IO_MEM_REGION_2_ADDR,
+	.fw_crash_dump_sz		= ATH10K_FW_IO_MEM_REGION_2_SIZE,
 };
 
 const struct ath10k_hw_values qca4019_values = {
@@ -180,6 +186,8 @@ const struct ath10k_hw_values qca4019_va
 	.num_target_ce_config_wlan      = 10,
 	.ce_desc_meta_data_mask         = 0xFFF0,
 	.ce_desc_meta_data_lsb          = 4,
+	.fw_crash_dump_addr		= IPQ40XX_FW_IO_MEM_ADDR,
+	.fw_crash_dump_sz		= ATH10K_FW_IO_MEM_REGION_2_SIZE,
 };
 
 const struct ath10k_hw_regs wcn3990_regs = {
--- a/drivers/net/wireless/ath/ath10k/hw.h
+++ b/drivers/net/wireless/ath/ath10k/hw.h
@@ -376,6 +376,8 @@ struct ath10k_hw_values {
 	u8 num_target_ce_config_wlan;
 	u16 ce_desc_meta_data_mask;
 	u8 ce_desc_meta_data_lsb;
+	u32 fw_crash_dump_addr;
+	u32 fw_crash_dump_sz;
 };
 
 extern const struct ath10k_hw_values qca988x_values;
--- a/drivers/net/wireless/ath/ath10k/Makefile
+++ b/drivers/net/wireless/ath/ath10k/Makefile
@@ -21,7 +21,7 @@ ath10k_core-$(CPTCFG_ATH10K_TRACING) +=
 ath10k_core-$(CONFIG_THERMAL) += thermal.o
 ath10k_core-$(CPTCFG_MAC80211_DEBUGFS) += debugfs_sta.o
 ath10k_core-$(CONFIG_PM) += wow.o
-ath10k_core-$(CONFIG_DEV_COREDUMP) += coredump.o
+ath10k_core-y += coredump.o
 ath10k_core-$(CPTCFG_ATH10K_CE) += ce.o
 
 obj-$(CPTCFG_ATH10K_PCI) += ath10k_pci.o
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -1132,11 +1132,9 @@ struct ath10k {
 
 	u32 pktlog_filter;
 
-#ifdef CONFIG_DEV_COREDUMP
 	struct {
 		struct ath10k_fw_crash_data *fw_crash_data;
 	} coredump;
-#endif
 
 	struct {
 		/* protected by conf_mutex */
--- a/drivers/net/wireless/ath/ath10k/coredump.c
+++ b/drivers/net/wireless/ath/ath10k/coredump.c
@@ -6,7 +6,9 @@
 
 #include "coredump.h"
 
+#ifdef CPTCFG_DEV_COREDUMP
 #include <linux/devcoredump.h>
+#endif
 #include <linux/kernel.h>
 #include <linux/types.h>
 #include <linux/utsname.h>
@@ -702,16 +704,6 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_REG,
-		.start = 0x98000,
-		.len = 0x50000,
-		.name = "IRAM",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
-	{
 		.type = ATH10K_MEM_REGION_TYPE_IOSRAM,
 		.start = 0xC0000,
 		.len = 0x40000,
@@ -722,10 +714,10 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x30000,
-		.len = 0x7000,
-		.name = "APB REG 1",
+		.type = ATH10K_MEM_REGION_TYPE_REG,
+		.start = 0x98000,
+		.len = 0x50000,
+		.name = "IRAM",
 		.section_table = {
 			.sections = NULL,
 			.size = 0,
@@ -733,9 +725,9 @@ static const struct ath10k_mem_region qc
 	},
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x3f000,
-		.len = 0x3000,
-		.name = "APB REG 2",
+		.start = 0x30000,
+		.len = 0x7000,
+		.name = "APB REG 1",
 		.section_table = {
 			.sections = NULL,
 			.size = 0,
@@ -771,6 +763,16 @@ static const struct ath10k_mem_region qc
 			.size = 0,
 		},
 	},
+	{
+		.type = ATH10K_MEM_REGION_TYPE_IOREG,
+		.start = 0x3f000,
+		.len = 0x3000,
+		.name = "APB REG 2",
+		.section_table = {
+			.sections = NULL,
+			.size = 0,
+		},
+	},
 };
 
 static const struct ath10k_mem_region qca9984_hw10_mem_regions[] = {
@@ -785,16 +787,6 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_REG,
-		.start = 0x98000,
-		.len = 0x50000,
-		.name = "IRAM",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
-	{
 		.type = ATH10K_MEM_REGION_TYPE_IOSRAM,
 		.start = 0xC0000,
 		.len = 0x40000,
@@ -805,10 +797,10 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x30000,
-		.len = 0x7000,
-		.name = "APB REG 1",
+		.type = ATH10K_MEM_REGION_TYPE_REG,
+		.start = 0x98000,
+		.len = 0x50000,
+		.name = "IRAM",
 		.section_table = {
 			.sections = NULL,
 			.size = 0,
@@ -816,9 +808,9 @@ static const struct ath10k_mem_region qc
 	},
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x3f000,
-		.len = 0x3000,
-		.name = "APB REG 2",
+		.start = 0x30000,
+		.len = 0x7000,
+		.name = "APB REG 1",
 		.section_table = {
 			.sections = NULL,
 			.size = 0,
@@ -854,6 +846,16 @@ static const struct ath10k_mem_region qc
 			.size = 0,
 		},
 	},
+	{
+		.type = ATH10K_MEM_REGION_TYPE_IOREG,
+		.start = 0x3f000,
+		.len = 0x3000,
+		.name = "APB REG 2",
+		.section_table = {
+			.sections = NULL,
+			.size = 0,
+		},
+	},
 };
 
 static const struct ath10k_mem_section ipq4019_soc_reg_range[] = {
@@ -911,16 +913,6 @@ static const struct ath10k_mem_region qc
 	},
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x3f000,
-		.len = 0x3000,
-		.name = "APB REG 2",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
-	{
-		.type = ATH10K_MEM_REGION_TYPE_IOREG,
 		.start = 0x43000,
 		.len = 0x3000,
 		.name = "WIFI REG",
@@ -949,6 +941,16 @@ static const struct ath10k_mem_region qc
 			.size = ARRAY_SIZE(ipq4019_soc_reg_range),
 		},
 	},
+	{
+		.type = ATH10K_MEM_REGION_TYPE_IOREG,
+		.start = 0x3f000,
+		.len = 0x3000,
+		.name = "APB REG 2",
+		.section_table = {
+			.sections = NULL,
+			.size = 0,
+		},
+	},
 };
 
 static const struct ath10k_hw_mem_layout hw_mem_layouts[] = {
@@ -1109,7 +1111,9 @@ struct ath10k_fw_crash_data *ath10k_core
 		return NULL;
 
 	guid_gen(&crash_data->guid);
+#ifdef CPTCFG_DEV_COREDUMP
 	ktime_get_real_ts64(&crash_data->timestamp);
+#endif
 
 	return crash_data;
 }
@@ -1137,6 +1141,7 @@ static struct ath10k_dump_file_data *ath
 	if (test_bit(ATH10K_FW_CRASH_DUMP_RAM_DATA, &ath10k_coredump_mask))
 		len += sizeof(*dump_tlv) + crash_data->ramdump_buf_len;
 
+	len = ar->hw_values->fw_crash_dump_sz;
 	sofar += hdr_len;
 
 	/* This is going to get big when we start dumping FW RAM and such,
@@ -1232,7 +1237,10 @@ int ath10k_coredump_submit(struct ath10k
 		return -ENODATA;
 	}
 
+	ath10k_dbg_bug_on_fw_assert(ar, (u8 *)dump);
+#ifdef CPTCFG_DEV_COREDUMP
 	dev_coredumpv(ar->dev, dump, le32_to_cpu(dump->len), GFP_KERNEL);
+#endif
 
 	return 0;
 }
--- a/drivers/net/wireless/ath/ath10k/coredump.h
+++ b/drivers/net/wireless/ath/ath10k/coredump.h
@@ -165,8 +165,6 @@ struct ath10k_hw_mem_layout {
 /* FIXME: where to put this? */
 extern unsigned long ath10k_coredump_mask;
 
-#ifdef CONFIG_DEV_COREDUMP
-
 int ath10k_coredump_submit(struct ath10k *ar);
 struct ath10k_fw_crash_data *ath10k_coredump_new(struct ath10k *ar);
 int ath10k_coredump_create(struct ath10k *ar);
@@ -176,42 +174,4 @@ void ath10k_coredump_destroy(struct ath1
 
 const struct ath10k_hw_mem_layout *ath10k_coredump_get_mem_layout(struct ath10k *ar);
 
-#else /* CONFIG_DEV_COREDUMP */
-
-static inline int ath10k_coredump_submit(struct ath10k *ar)
-{
-	return 0;
-}
-
-static inline struct ath10k_fw_crash_data *ath10k_coredump_new(struct ath10k *ar)
-{
-	return NULL;
-}
-
-static inline int ath10k_coredump_create(struct ath10k *ar)
-{
-	return 0;
-}
-
-static inline int ath10k_coredump_register(struct ath10k *ar)
-{
-	return 0;
-}
-
-static inline void ath10k_coredump_unregister(struct ath10k *ar)
-{
-}
-
-static inline void ath10k_coredump_destroy(struct ath10k *ar)
-{
-}
-
-static inline const struct ath10k_hw_mem_layout *
-ath10k_coredump_get_mem_layout(struct ath10k *ar)
-{
-	return NULL;
-}
-
-#endif /* CONFIG_DEV_COREDUMP */
-
 #endif /* _COREDUMP_H_ */
