--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -940,6 +940,28 @@ static inline u8 ath10k_bw_to_mac80211_b
 	return ret;
 }
 
+static inline u8 mac80211_bw_to_ath10k_bw(u8 bw)
+{
+	u8 ret = 0;
+
+	switch (bw) {
+	case RATE_INFO_BW_20:
+		ret = 0;
+		break;
+	case RATE_INFO_BW_40:
+		ret = 1;
+		break;
+	case RATE_INFO_BW_80:
+		ret = 2;
+		break;
+	case RATE_INFO_BW_160:
+		ret = 3;
+		break;
+	}
+
+	return ret;
+}
+
 static void ath10k_htt_rx_h_rates(struct ath10k *ar,
 				  struct ieee80211_rx_status *status,
 				  struct htt_rx_desc *rxd)
@@ -2947,25 +2969,26 @@ ath10k_accumulate_per_peer_tx_stats(stru
 		return;
 
 	tx_stats = arsta->tx_stats;
-	flags = txrate->flags;
-	gi = test_bit(ATH10K_RATE_INFO_FLAGS_SGI_BIT, &flags);
-	mcs = ATH10K_HW_MCS_RATE(pstats->ratecode);
-	bw = txrate->bw;
-	nss = txrate->nss;
-	ht_idx = mcs + (nss - 1) * 8;
+
+	gi = !!(arsta->txrate.flags & RATE_INFO_FLAGS_SHORT_GI);
+	mcs = txrate->mcs;
+	ht_idx = txrate->mcs + (txrate->nss - 1) * 8;
+	bw = mac80211_bw_to_ath10k_bw(txrate->bw);
+	nss = txrate->nss - 1;
+
 	idx = mcs * 8 + 8 * 10 * (nss - 1);
 	idx += bw * 2 + gi;
 
 #define STATS_OP_FMT(name) tx_stats->stats[ATH10K_STATS_TYPE_##name]
 
-	if (txrate->flags == RATE_INFO_FLAGS_VHT_MCS) {
+	if (txrate->flags & RATE_INFO_FLAGS_VHT_MCS) {
 		STATS_OP_FMT(SUCC).vht[0][mcs] += pstats->succ_bytes;
 		STATS_OP_FMT(SUCC).vht[1][mcs] += pstats->succ_pkts;
 		STATS_OP_FMT(FAIL).vht[0][mcs] += pstats->failed_bytes;
 		STATS_OP_FMT(FAIL).vht[1][mcs] += pstats->failed_pkts;
 		STATS_OP_FMT(RETRY).vht[0][mcs] += pstats->retry_bytes;
 		STATS_OP_FMT(RETRY).vht[1][mcs] += pstats->retry_pkts;
-	} else if (txrate->flags == RATE_INFO_FLAGS_MCS) {
+	} else if (txrate->flags & RATE_INFO_FLAGS_MCS) {
 		STATS_OP_FMT(SUCC).ht[0][ht_idx] += pstats->succ_bytes;
 		STATS_OP_FMT(SUCC).ht[1][ht_idx] += pstats->succ_pkts;
 		STATS_OP_FMT(FAIL).ht[0][ht_idx] += pstats->failed_bytes;
@@ -2986,7 +3009,7 @@ ath10k_accumulate_per_peer_tx_stats(stru
 	if (ATH10K_HW_AMPDU(pstats->flags)) {
 		tx_stats->ba_fails += ATH10K_HW_BA_FAIL(pstats->flags);
 
-		if (txrate->flags == RATE_INFO_FLAGS_MCS) {
+		if (txrate->flags & RATE_INFO_FLAGS_MCS) {
 			STATS_OP_FMT(AMPDU).ht[0][ht_idx] +=
 				pstats->succ_bytes + pstats->retry_bytes;
 			STATS_OP_FMT(AMPDU).ht[1][ht_idx] +=
