Index: backports-20190311-3.14.77-d492b001/drivers/net/wireless/ath/ath10k/txrx.c
===================================================================
--- backports-20190311-3.14.77-d492b001.orig/drivers/net/wireless/ath/ath10k/txrx.c
+++ backports-20190311-3.14.77-d492b001/drivers/net/wireless/ath/ath10k/txrx.c
@@ -189,9 +189,11 @@ int ath10k_txrx_tx_unref(struct ath10k_h
 		wake_up(&htt->empty_tx_wq);
 	spin_unlock_bh(&htt->tx_lock);
 
+	rcu_read_lock();
 	if (txq && txq->sta && skb_cb->airtime_est)
 		ieee80211_sta_register_airtime(txq->sta, txq->tid,
 					       skb_cb->airtime_est, 0);
+	rcu_read_unlock();
 
 	if (ar->bus_param.dev_type != ATH10K_DEV_TYPE_HL)
 		dma_unmap_single(dev, skb_cb->paddr, msdu->len, DMA_TO_DEVICE);
