Index: backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/bmi.h
===================================================================
--- backports-20180529-3.14.77-8442d41b.orig/drivers/net/wireless/ath/ath10k/bmi.h	2018-08-24 11:53:56.403772150 +0530
+++ backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/bmi.h	2018-08-24 11:53:56.395772152 +0530
@@ -86,9 +86,9 @@
 #define BMI_PARAM_GET_FLASH_BOARD_ID 0x8000
 #define BMI_PARAM_FLASH_SECTION_ALL 0x10000
 
-/* Dual-band Param IDs */
-#define BMI_PARAM_DUAL_BAND_2G 0x40000
-#define BMI_PARAM_DUAL_BAND_5G 0x80000
+/* Dual-band Extended Board ID */
+#define BMI_PARAM_GET_EXT_BOARD_ID 0x40000
+#define ATH10K_BMI_EXT_BOARD_ID_SUPPORT 0x40000
 
 #define ATH10K_BMI_BOARD_ID_FROM_OTP_MASK   0x7c00
 #define ATH10K_BMI_BOARD_ID_FROM_OTP_LSB    10
@@ -97,6 +97,7 @@
 #define ATH10K_BMI_CHIP_ID_FROM_OTP_LSB     15
 
 #define ATH10K_BMI_BOARD_ID_STATUS_MASK 0xff
+#define ATH10K_BMI_EBOARD_ID_STATUS_MASK 0xff
 
 struct bmi_cmd {
 	__le32 id; /* enum bmi_cmd_id */
Index: backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/core.c
===================================================================
--- backports-20180529-3.14.77-8442d41b.orig/drivers/net/wireless/ath/ath10k/core.c	2018-08-24 11:53:56.403772150 +0530
+++ backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/core.c	2018-08-24 11:54:42.935763139 +0530
@@ -42,7 +42,6 @@
 EXPORT_SYMBOL(ath10k_debug_mask);
 
 static unsigned int ath10k_cryptmode_param;
-static unsigned int ath10k_dual_band_config;
 static bool uart_print;
 static bool skip_otp;
 static bool rawmode;
@@ -58,7 +57,6 @@
 /* FIXME: most of these should be readonly */
 module_param_named(debug_mask, ath10k_debug_mask, uint, 0644);
 module_param_named(cryptmode, ath10k_cryptmode_param, uint, 0644);
-module_param_named(dual_band, ath10k_dual_band_config, uint, 0644);
 module_param(uart_print, bool, 0644);
 module_param(skip_otp, bool, 0644);
 module_param(rawmode, bool, 0644);
@@ -72,7 +70,6 @@
 MODULE_PARM_DESC(uart_print, "Uart target debugging");
 MODULE_PARM_DESC(skip_otp, "Skip otp failure for calibration in testmode");
 MODULE_PARM_DESC(cryptmode, "Crypto mode: 0-hardware, 1-software");
-MODULE_PARM_DESC(dual_band, "Ath10k dual-band config: 2 - 2.4Ghz, 5 - 5Ghz");
 MODULE_PARM_DESC(rawmode, "Use raw 802.11 frame datapath");
 MODULE_PARM_DESC(coredump_mask, "Bitfield of what to include in firmware crash file");
 MODULE_PARM_DESC(bug_on_fw_assert, "Bug on firmware assert");
@@ -374,8 +371,10 @@
 		.fw = {
 			.dir = QCA9984_HW_1_0_FW_DIR,
 			.board = QCA9984_HW_1_0_BOARD_DATA_FILE,
+			.eboard = QCA9984_HW_1_0_EBOARD_DATA_FILE,
 			.board_size = QCA99X0_BOARD_DATA_SZ,
 			.board_ext_size = QCA99X0_BOARD_EXT_DATA_SZ,
+			.ext_board_size = QCA99X0_EXT_BOARD_DATA_SZ,
 		},
 		.sw_decrypt_mcast_mgmt = true,
 		.hw_ops = &qca99x0_ops,
@@ -794,877 +793,1035 @@
 	return 0;
 }
 
-static int ath10k_download_board_data(struct ath10k *ar, const void *data,
-				      size_t data_len)
+static int ath10k_core_get_board_id_from_otp(struct ath10k *ar)
 {
-	u32 board_data_size = ar->hw_params.fw.board_size;
-	u32 address;
-	int ret;
+	u32 result, address;
+	u8 board_id, chip_id, ext_bid_support;
+	int ret, bmi_board_id_param;
 
-	ret = ath10k_push_board_ext_data(ar, data, data_len);
-	if (ret) {
-		ath10k_err(ar, "could not push board ext data (%d)\n", ret);
-		goto exit;
-	}
+	address = ar->hw_params.patch_load_addr;
 
-	ret = ath10k_bmi_read32(ar, hi_board_data, &address);
-	if (ret) {
-		ath10k_err(ar, "could not read board data addr (%d)\n", ret);
-		goto exit;
+	if (!ar->normal_mode_fw.fw_file.otp_data ||
+	    !ar->normal_mode_fw.fw_file.otp_len) {
+		ath10k_warn(ar,
+			    "failed to retrieve board id because of invalid otp\n");
+		return -ENODATA;
 	}
 
-	ret = ath10k_bmi_write_memory(ar, address, data,
-				      min_t(u32, board_data_size,
-					    data_len));
+	ath10k_dbg(ar, ATH10K_DBG_BOOT,
+		   "boot upload otp to 0x%x len %zd for board id\n",
+		   address, ar->normal_mode_fw.fw_file.otp_len);
+
+	ret = ath10k_bmi_fast_download(ar, address,
+				       ar->normal_mode_fw.fw_file.otp_data,
+				       ar->normal_mode_fw.fw_file.otp_len);
 	if (ret) {
-		ath10k_err(ar, "could not write board data (%d)\n", ret);
-		goto exit;
+		ath10k_err(ar, "could not write otp for board id check: %d\n",
+			   ret);
+		return ret;
 	}
 
-	ret = ath10k_bmi_write32(ar, hi_board_data_initialized, 1);
+	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_DT ||
+	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FILE ||
+	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FLASH_MTD)
+		bmi_board_id_param = BMI_PARAM_GET_FLASH_BOARD_ID;
+	else
+		bmi_board_id_param = BMI_PARAM_GET_EEPROM_BOARD_ID;
+
+	ret = ath10k_bmi_execute(ar, address, bmi_board_id_param, &result);
 	if (ret) {
-		ath10k_err(ar, "could not write board data bit (%d)\n", ret);
-		goto exit;
+		ath10k_err(ar, "could not execute otp for board id check: %d\n",
+			   ret);
+		return ret;
 	}
 
-exit:
-	return ret;
-}
+	board_id = MS(result, ATH10K_BMI_BOARD_ID_FROM_OTP);
+	chip_id = MS(result, ATH10K_BMI_CHIP_ID_FROM_OTP);
+	ext_bid_support = ((result & ATH10K_BMI_EXT_BOARD_ID_SUPPORT) ? 1 : 0);
 
-static int ath10k_verify_mtd_cal_data_chksum(u32 *data, u32 len)
-{
-	u16 *buf = (u16 *)data;
-	u16 sum;
-	int i;
+	ath10k_dbg(ar, ATH10K_DBG_BOOT,
+		   "boot get otp board id result 0x%08x board_id %d chip_id %d ext_bid_support %d\n",
+		   result, board_id, chip_id, ext_bid_support);
 
-	sum = 0;
+	ar->id.ext_bid_supported = ext_bid_support;
 
-	for (i = 0; i < (len / 2); i++)
-		sum ^= le16_to_cpu(*buf++);
+	if ((result & ATH10K_BMI_BOARD_ID_STATUS_MASK) != 0 ||
+	    (board_id == 0)) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "board id does not exist in otp, ignore it\n");
+		return -EOPNOTSUPP;
+	}
 
-	if (sum != 0xffff)
-		return -EINVAL;
+	ar->id.bmi_ids_valid = true;
+	ar->id.bmi_board_id = board_id;
+	ar->id.bmi_chip_id = chip_id;
 
 	return 0;
 }
 
-static int ath10k_download_cal_data_mtd(struct ath10k *ar)
+static void ath10k_core_check_bdfext(const struct dmi_header *hdr, void *data)
 {
-	struct device_node *node;
-	struct mtd_info *mtd;
-	size_t retlen;
-	u32 *buf, addr, len;
-	const char *mtd_name;
-	int ret, i;
-
-	node = ar->dev->of_node;
-	if (!node) {
-		/* Device Tree is optional, don't print any warnings if
-		 * there is no node for ath10k.
-		 */
-		return -ENOENT;
-	}
-
-	ret = of_property_read_u32(node, "qcom,cal-len", &len);
-	if (ret)
-		return -ENOENT;
-
-	ret = of_property_read_u32(node, "qcom,cal-offset", &addr);
-	if (ret)
-		return -ENOENT;
-
-	mtd_name = of_get_property(node, "qcom,mtd-name", NULL);
-	if (!mtd_name)
-		return -ENOENT;
-
-	if (len != ar->hw_params.cal_data_len) {
-		ath10k_warn(ar, "invalid cal data length in DT: %d\n", len);
-		return -EMSGSIZE;
-	}
+	struct ath10k *ar = data;
+	const char *bdf_ext;
+	const char *magic = ATH10K_SMBIOS_BDF_EXT_MAGIC;
+	u8 bdf_enabled;
+	int i;
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "mtd name: %s addr: 0x%x len: %d\n",
-		   mtd_name, addr, len);
+	if (hdr->type != ATH10K_SMBIOS_BDF_EXT_TYPE)
+		return;
 
-	buf = vmalloc(len);
-	if (!buf) {
+	if (hdr->length != ATH10K_SMBIOS_BDF_EXT_LENGTH) {
 		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "failed to alloc memory for mtd caldata\n");
-		return -ENOMEM;
+			   "wrong smbios bdf ext type length (%d).\n",
+			   hdr->length);
+		return;
 	}
 
-	mtd = get_mtd_device_nm(mtd_name);
-	if (IS_ERR_OR_NULL(mtd)) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "mtd device %s not found to read caldata\n",
-			   mtd_name);
-		ret = PTR_ERR(mtd);
-		goto release_buf;
+	bdf_enabled = *((u8 *)hdr + ATH10K_SMBIOS_BDF_EXT_OFFSET);
+	if (!bdf_enabled) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT, "bdf variant name not found.\n");
+		return;
 	}
 
-	ret = mtd_read(mtd, addr, len, &retlen, (u_char *)buf);
-	if (ret) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "failed to read caldata from mtd device %s: %d\n",
-			   mtd_name, ret);
-		goto release_mtd;
-	}
+	/* Only one string exists (per spec) */
+	bdf_ext = (char *)hdr + hdr->length;
 
-	if (retlen != len) {
+	if (memcmp(bdf_ext, magic, strlen(magic)) != 0) {
 		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "got only %d bytes from mtd (requested: %d)\n",
-			   retlen, len);
-		goto release_mtd;
+			   "bdf variant magic does not match.\n");
+		return;
 	}
 
-	ret = ath10k_verify_mtd_cal_data_chksum(buf, len);
-	if (ret) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "checksum verification failed\n");
-		goto release_mtd;
+	for (i = 0; i < strlen(bdf_ext); i++) {
+		if (!isascii(bdf_ext[i]) || !isprint(bdf_ext[i])) {
+			ath10k_dbg(ar, ATH10K_DBG_BOOT,
+				   "bdf variant name contains non ascii chars.\n");
+			return;
+		}
 	}
 
-	for (i = 0; i < (len / 4); i++)
-		buf[i] = cpu_to_le32(buf[i]);
-
-	ret = ath10k_download_board_data(ar, (const void *)buf,
-					 (size_t)len);
-	if (ret) {
-		ath10k_err(ar, "failed to download mtd caldata: %d\n", ret);
-		goto release_mtd;
+	/* Copy extension name without magic suffix */
+	if (strscpy(ar->id.bdf_ext, bdf_ext + strlen(magic),
+		    sizeof(ar->id.bdf_ext)) < 0) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "bdf variant string is longer than the buffer can accommodate (variant: %s)\n",
+			    bdf_ext);
+		return;
 	}
 
 	ath10k_dbg(ar, ATH10K_DBG_BOOT,
-		   "cal data downloaded successfully\n");
+		   "found and validated bdf variant smbios_type 0x%x bdf %s\n",
+		   ATH10K_SMBIOS_BDF_EXT_TYPE, bdf_ext);
+}
 
-	put_mtd_device(mtd);
-	vfree(buf);
+static int ath10k_core_check_smbios(struct ath10k *ar)
+{
+	ar->id.bdf_ext[0] = '\0';
+	dmi_walk(ath10k_core_check_bdfext, ar);
+
+	if (ar->id.bdf_ext[0] == '\0')
+		return -ENODATA;
 
 	return 0;
+}
 
-release_mtd:
-	put_mtd_device(mtd);
+static int ath10k_core_check_dt(struct ath10k *ar)
+{
+	struct device_node *node;
+	const char *variant = NULL;
 
-release_buf:
-	vfree(buf);
+	node = ar->dev->of_node;
+	if (!node)
+		return -ENOENT;
 
-	return ret;
+	of_property_read_string(node, "qcom,ath10k-calibration-variant",
+				&variant);
+	if (!variant)
+		return -ENODATA;
+
+	if (strscpy(ar->id.bdf_ext, variant, sizeof(ar->id.bdf_ext)) < 0)
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "bdf variant string is longer than the buffer can accommodate (variant: %s)\n",
+			    variant);
+
+	return 0;
 }
 
-static int ath10k_download_cal_file(struct ath10k *ar,
-				    const struct firmware *file)
+static int ath10k_download_fw(struct ath10k *ar)
 {
+	u32 address, data_len;
+	const void *data;
 	int ret;
 
-	if (!file)
-		return -ENOENT;
+	address = ar->hw_params.patch_load_addr;
 
-	if (IS_ERR(file))
-		return PTR_ERR(file);
+	data = ar->running_fw->fw_file.firmware_data;
+	data_len = ar->running_fw->fw_file.firmware_len;
 
-	ret = ath10k_download_board_data(ar, file->data, file->size);
+	ret = ath10k_swap_code_seg_configure(ar, &ar->running_fw->fw_file);
 	if (ret) {
-		ath10k_err(ar, "failed to download cal_file data: %d\n", ret);
+		ath10k_err(ar, "failed to configure fw code swap: %d\n",
+			   ret);
 		return ret;
 	}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot cal file downloaded\n");
+	ath10k_dbg(ar, ATH10K_DBG_BOOT,
+		   "boot uploading firmware image %pK len %d\n",
+		   data, data_len);
 
-	return 0;
+	ret = ath10k_bmi_fast_download(ar, address, data, data_len);
+	if (ret) {
+		ath10k_err(ar, "failed to download firmware: %d\n",
+			   ret);
+		return ret;
+	}
+
+	return ret;
 }
 
-static int ath10k_download_cal_dt(struct ath10k *ar, const char *dt_name)
+static void ath10k_core_free_board_files(struct ath10k *ar)
 {
-	struct device_node *node;
-	int data_len;
-	void *data;
-	int ret;
+	if (!IS_ERR(ar->normal_mode_fw.board))
+		release_firmware(ar->normal_mode_fw.board);
 
-	node = ar->dev->of_node;
-	if (!node)
-		/* Device Tree is optional, don't print any warnings if
-		 * there's no node for ath10k.
-		 */
-		return -ENOENT;
+	if (!IS_ERR(ar->normal_mode_fw.ext_board))
+		release_firmware(ar->normal_mode_fw.ext_board);
 
-	if (!of_get_property(node, dt_name, &data_len)) {
-		/* The calibration data node is optional */
-		return -ENOENT;
-	}
+	ar->normal_mode_fw.board = NULL;
+	ar->normal_mode_fw.board_data = NULL;
+	ar->normal_mode_fw.board_len = 0;
+	ar->normal_mode_fw.ext_board = NULL;
+	ar->normal_mode_fw.ext_board_data = NULL;
+	ar->normal_mode_fw.ext_board_len = 0;
+}
 
-	if (data_len != ar->hw_params.cal_data_len) {
-		ath10k_warn(ar, "invalid calibration data length in DT: %d\n",
-			    data_len);
-		ret = -EMSGSIZE;
-		goto out;
-	}
+static void ath10k_core_free_firmware_files(struct ath10k *ar)
+{
+	if (!IS_ERR(ar->normal_mode_fw.fw_file.firmware))
+		release_firmware(ar->normal_mode_fw.fw_file.firmware);
 
-	data = kmalloc(data_len, GFP_KERNEL);
-	if (!data) {
-		ret = -ENOMEM;
-		goto out;
-	}
+	if (!IS_ERR(ar->cal_file))
+		release_firmware(ar->cal_file);
 
-	ret = of_property_read_u8_array(node, dt_name, data, data_len);
-	if (ret) {
-		ath10k_warn(ar, "failed to read calibration data from DT: %d\n",
-			    ret);
-		goto out_free;
-	}
+	if (!IS_ERR(ar->pre_cal_file))
+		release_firmware(ar->pre_cal_file);
 
-	ret = ath10k_download_board_data(ar, data, data_len);
-	if (ret) {
-		ath10k_warn(ar, "failed to download calibration data from Device Tree: %d\n",
-			    ret);
-		goto out_free;
-	}
+	ath10k_swap_code_seg_release(ar, &ar->normal_mode_fw.fw_file);
 
-	ret = 0;
+	ar->normal_mode_fw.fw_file.otp_data = NULL;
+	ar->normal_mode_fw.fw_file.otp_len = 0;
 
-out_free:
-	kfree(data);
+	ar->normal_mode_fw.fw_file.firmware = NULL;
+	ar->normal_mode_fw.fw_file.firmware_data = NULL;
+	ar->normal_mode_fw.fw_file.firmware_len = 0;
 
-out:
-	return ret;
+	ar->cal_file = NULL;
+	ar->pre_cal_file = NULL;
 }
 
-static int ath10k_download_cal_eeprom(struct ath10k *ar)
+static int ath10k_fetch_cal_file(struct ath10k *ar)
 {
-	size_t data_len;
-	void *data = NULL;
-	int ret;
+	char filename[100];
 
-	ret = ath10k_hif_fetch_cal_eeprom(ar, &data, &data_len);
-	if (ret) {
-		if (ret != -EOPNOTSUPP)
-			ath10k_warn(ar, "failed to read calibration data from EEPROM: %d\n",
-				    ret);
-		goto out_free;
-	}
+	/* pre-cal-<bus>-<id>.bin */
+	scnprintf(filename, sizeof(filename), "pre-cal-%s-%s.bin",
+		  ath10k_bus_str(ar->hif.bus), dev_name(ar->dev));
 
-	ret = ath10k_download_board_data(ar, data, data_len);
-	if (ret) {
-		ath10k_warn(ar, "failed to download calibration data from EEPROM: %d\n",
-			    ret);
-		goto out_free;
-	}
+	ar->pre_cal_file = ath10k_fetch_fw_file(ar, ATH10K_FW_DIR, filename);
+	if (!IS_ERR(ar->pre_cal_file))
+		goto success;
 
-	ret = 0;
+	/* cal-<bus>-<id>.bin */
+	scnprintf(filename, sizeof(filename), "cal-%s-%s.bin",
+		  ath10k_bus_str(ar->hif.bus), dev_name(ar->dev));
 
-out_free:
-	kfree(data);
+	ar->cal_file = ath10k_fetch_fw_file(ar, ATH10K_FW_DIR, filename);
+	if (IS_ERR(ar->cal_file))
+		/* calibration file is optional, don't print any warnings */
+		return PTR_ERR(ar->cal_file);
+success:
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "found calibration file %s/%s\n",
+		   ATH10K_FW_DIR, filename);
 
-	return ret;
+	return 0;
 }
 
-static int ath10k_core_get_board_id_from_otp(struct ath10k *ar)
+static int ath10k_core_fetch_board_data_api_1(struct ath10k *ar, int bd_ie_type)
 {
-	u32 result, address;
-	u8 board_id, chip_id;
-	int ret, bmi_board_id_param;
+	if (bd_ie_type == ATH10K_BD_IE_BOARD) {
+		if (!ar->hw_params.fw.board) {
+			ath10k_err(ar, "failed to find board file fw entry\n");
+			return -EINVAL;
+		}
 
-	address = ar->hw_params.patch_load_addr;
+		ar->normal_mode_fw.board = ath10k_fetch_fw_file(ar,
+								ar->hw_params.fw.dir,
+								ar->hw_params.fw.board);
+		if (IS_ERR(ar->normal_mode_fw.board))
+			return PTR_ERR(ar->normal_mode_fw.board);
+
+		ar->normal_mode_fw.board_data = ar->normal_mode_fw.board->data;
+		ar->normal_mode_fw.board_len = ar->normal_mode_fw.board->size;
+	} else if (bd_ie_type == ATH10K_BD_IE_BOARD_EXT) {
+		if (!ar->hw_params.fw.eboard) {
+			ath10k_err(ar, "failed to find eboard file fw entry\n");
+			return -EINVAL;
+		}
 
-	if (!ar->normal_mode_fw.fw_file.otp_data ||
-	    !ar->normal_mode_fw.fw_file.otp_len) {
-		ath10k_warn(ar,
-			    "failed to retrieve board id because of invalid otp\n");
-		return -ENODATA;
+		ar->normal_mode_fw.ext_board = ath10k_fetch_fw_file(ar,
+								    ar->hw_params.fw.dir,
+								    ar->hw_params.fw.eboard);
+		if (IS_ERR(ar->normal_mode_fw.ext_board))
+			return PTR_ERR(ar->normal_mode_fw.ext_board);
+
+		ar->normal_mode_fw.ext_board_data = ar->normal_mode_fw.ext_board->data;
+		ar->normal_mode_fw.ext_board_len = ar->normal_mode_fw.ext_board->size;
 	}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT,
-		   "boot upload otp to 0x%x len %zd for board id\n",
-		   address, ar->normal_mode_fw.fw_file.otp_len);
+	return 0;
+}
 
-	ret = ath10k_bmi_fast_download(ar, address,
-				       ar->normal_mode_fw.fw_file.otp_data,
-				       ar->normal_mode_fw.fw_file.otp_len);
-	if (ret) {
-		ath10k_err(ar, "could not write otp for board id check: %d\n",
-			   ret);
-		return ret;
-	}
+static int ath10k_core_parse_bd_ie_board(struct ath10k *ar,
+					 const void *buf, size_t buf_len,
+					 const char *boardname,
+					 int bd_ie_type)
+{
+	const struct ath10k_fw_ie *hdr;
+	bool name_match_found;
+	int ret, board_ie_id;
+	size_t board_ie_len;
+	const void *board_ie_data;
 
-	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_DT ||
-	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FILE ||
-	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FLASH_MTD)
-		bmi_board_id_param = BMI_PARAM_GET_FLASH_BOARD_ID;
-	else
-		bmi_board_id_param = BMI_PARAM_GET_EEPROM_BOARD_ID;
+	name_match_found = false;
 
-        /* Select radio band from user module-param 'dual_band' */
-        if (ath10k_dual_band_config == 2) {
-            bmi_board_id_param |= BMI_PARAM_DUAL_BAND_2G;
-            ath10k_dbg(ar, ATH10K_DBG_BOOT,
-                       "Dual-band configuring with 2.4Ghz");
-        } else if (ath10k_dual_band_config == 5) {
-            bmi_board_id_param |= BMI_PARAM_DUAL_BAND_5G;
-            ath10k_dbg(ar, ATH10K_DBG_BOOT,
-                       "Dual-band configuring with 5Ghz");
-        }
+	/* go through ATH10K_BD_IE_BOARD_ elements */
+	while (buf_len > sizeof(struct ath10k_fw_ie)) {
+		hdr = buf;
+		board_ie_id = le32_to_cpu(hdr->id);
+		board_ie_len = le32_to_cpu(hdr->len);
+		board_ie_data = hdr->data;
 
-	ret = ath10k_bmi_execute(ar, address, bmi_board_id_param, &result);
-	if (ret) {
-		ath10k_err(ar, "could not execute otp for board id check: %d\n",
-			   ret);
-		return ret;
-	}
+		buf_len -= sizeof(*hdr);
+		buf += sizeof(*hdr);
 
-	board_id = MS(result, ATH10K_BMI_BOARD_ID_FROM_OTP);
-	chip_id = MS(result, ATH10K_BMI_CHIP_ID_FROM_OTP);
+		if (buf_len < ALIGN(board_ie_len, 4)) {
+			ath10k_err(ar, "invalid ATH10K_BD_IE_BOARD length: %zu < %zu\n",
+				   buf_len, ALIGN(board_ie_len, 4));
+			ret = -EINVAL;
+			goto out;
+		}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT,
-		   "boot get otp board id result 0x%08x board_id %d chip_id %d\n",
-		   result, board_id, chip_id);
+		switch (board_ie_id) {
+		case ATH10K_BD_IE_BOARD_NAME:
+			ath10k_dbg_dump(ar, ATH10K_DBG_BOOT, "board name", "",
+					board_ie_data, board_ie_len);
 
-	if ((result & ATH10K_BMI_BOARD_ID_STATUS_MASK) != 0 ||
-	    (board_id == 0)) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "board id does not exist in otp, ignore it\n");
-		return -EOPNOTSUPP;
+			if (board_ie_len != strlen(boardname))
+				break;
+
+			ret = memcmp(board_ie_data, boardname, strlen(boardname));
+			if (ret)
+				break;
+
+			name_match_found = true;
+			ath10k_dbg(ar, ATH10K_DBG_BOOT,
+				   "boot found match for name '%s'",
+				   boardname);
+			break;
+		case ATH10K_BD_IE_BOARD_DATA:
+			if (!name_match_found)
+				/* no match found */
+				break;
+
+			if (bd_ie_type == ATH10K_BD_IE_BOARD) {
+				ath10k_dbg(ar, ATH10K_DBG_BOOT,
+					   "boot found board data for '%s'",
+						boardname);
+
+				ar->normal_mode_fw.board_data = board_ie_data;
+				ar->normal_mode_fw.board_len = board_ie_len;
+			} else if (bd_ie_type == ATH10K_BD_IE_BOARD_EXT) {
+				ath10k_dbg(ar, ATH10K_DBG_BOOT,
+					   "boot found eboard data for '%s'",
+						boardname);
+
+				ar->normal_mode_fw.ext_board_data = board_ie_data;
+				ar->normal_mode_fw.ext_board_len = board_ie_len;
+			}
+
+			ret = 0;
+			goto out;
+		default:
+			ath10k_warn(ar, "unknown ATH10K_BD_IE_BOARD found: %d\n",
+				    board_ie_id);
+			break;
+		}
+
+		/* jump over the padding */
+		board_ie_len = ALIGN(board_ie_len, 4);
+
+		buf_len -= board_ie_len;
+		buf += board_ie_len;
 	}
 
-	ar->id.bmi_ids_valid = true;
-	ar->id.bmi_board_id = board_id;
-	ar->id.bmi_chip_id = chip_id;
+	/* no match found */
+	ret = -ENOENT;
 
-	return 0;
+out:
+	return ret;
 }
 
-static void ath10k_core_check_bdfext(const struct dmi_header *hdr, void *data)
+static int ath10k_core_search_bd(struct ath10k *ar,
+				 const char *boardname,
+				 const u8 *data,
+				 size_t len)
 {
-	struct ath10k *ar = data;
-	const char *bdf_ext;
-	const char *magic = ATH10K_SMBIOS_BDF_EXT_MAGIC;
-	u8 bdf_enabled;
-	int i;
+	size_t ie_len;
+	struct ath10k_fw_ie *hdr;
+	int ret = -ENOENT, ie_id;
 
-	if (hdr->type != ATH10K_SMBIOS_BDF_EXT_TYPE)
-		return;
+	while (len > sizeof(struct ath10k_fw_ie)) {
+		hdr = (struct ath10k_fw_ie *)data;
+		ie_id = le32_to_cpu(hdr->id);
+		ie_len = le32_to_cpu(hdr->len);
 
-	if (hdr->length != ATH10K_SMBIOS_BDF_EXT_LENGTH) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "wrong smbios bdf ext type length (%d).\n",
-			   hdr->length);
-		return;
-	}
+		len -= sizeof(*hdr);
+		data = hdr->data;
 
-	bdf_enabled = *((u8 *)hdr + ATH10K_SMBIOS_BDF_EXT_OFFSET);
-	if (!bdf_enabled) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT, "bdf variant name not found.\n");
-		return;
-	}
+		if (len < ALIGN(ie_len, 4)) {
+			ath10k_err(ar, "invalid length for board ie_id %d ie_len %zu len %zu\n",
+				   ie_id, ie_len, len);
+			return -EINVAL;
+		}
 
-	/* Only one string exists (per spec) */
-	bdf_ext = (char *)hdr + hdr->length;
+		switch (ie_id) {
+		case ATH10K_BD_IE_BOARD:
+			ret = ath10k_core_parse_bd_ie_board(ar, data, ie_len,
+							    boardname,
+							    ATH10K_BD_IE_BOARD);
+			if (ret == -ENOENT)
+				/* no match found, continue */
+				break;
 
-	if (memcmp(bdf_ext, magic, strlen(magic)) != 0) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "bdf variant magic does not match.\n");
-		return;
-	}
+			/* either found or error, so stop searching */
+			goto out;
+		case ATH10K_BD_IE_BOARD_EXT:
+			ret = ath10k_core_parse_bd_ie_board(ar, data, ie_len,
+							    boardname,
+							    ATH10K_BD_IE_BOARD_EXT);
+			if (ret == -ENOENT)
+				/* no match found, continue */
+				break;
 
-	for (i = 0; i < strlen(bdf_ext); i++) {
-		if (!isascii(bdf_ext[i]) || !isprint(bdf_ext[i])) {
-			ath10k_dbg(ar, ATH10K_DBG_BOOT,
-				   "bdf variant name contains non ascii chars.\n");
-			return;
+			/* either found or error, so stop searching */
+			goto out;
 		}
-	}
 
-	/* Copy extension name without magic suffix */
-	if (strscpy(ar->id.bdf_ext, bdf_ext + strlen(magic),
-		    sizeof(ar->id.bdf_ext)) < 0) {
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "bdf variant string is longer than the buffer can accommodate (variant: %s)\n",
-			    bdf_ext);
-		return;
+		/* jump over the padding */
+		ie_len = ALIGN(ie_len, 4);
+
+		len -= ie_len;
+		data += ie_len;
 	}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT,
-		   "found and validated bdf variant smbios_type 0x%x bdf %s\n",
-		   ATH10K_SMBIOS_BDF_EXT_TYPE, bdf_ext);
+out:
+	/* return result of parse_bd_ie_board() or -ENOENT */
+	return ret;
 }
 
-static int ath10k_core_check_smbios(struct ath10k *ar)
+static int ath10k_core_fetch_board_data_api_n(struct ath10k *ar,
+					      const char *boardname,
+					      const char *fallback_boardname,
+					      const char *filename)
 {
-	ar->id.bdf_ext[0] = '\0';
-	dmi_walk(ath10k_core_check_bdfext, ar);
+	size_t len, magic_len;
+	const u8 *data;
+	int ret;
 
-	if (ar->id.bdf_ext[0] == '\0')
-		return -ENODATA;
+	/* Skip if already fetched during board data download */
+	if (!ar->normal_mode_fw.board)
+		ar->normal_mode_fw.board = ath10k_fetch_fw_file(ar,
+								ar->hw_params.fw.dir,
+								filename);
+	if (IS_ERR(ar->normal_mode_fw.board))
+		return PTR_ERR(ar->normal_mode_fw.board);
 
-	return 0;
-}
+	data = ar->normal_mode_fw.board->data;
+	len = ar->normal_mode_fw.board->size;
 
-static int ath10k_core_check_dt(struct ath10k *ar)
-{
-	struct device_node *node;
-	const char *variant = NULL;
+	/* magic has extra null byte padded */
+	magic_len = strlen(ATH10K_BOARD_MAGIC) + 1;
+	if (len < magic_len) {
+		ath10k_err(ar, "failed to find magic value in %s/%s, file too short: %zu\n",
+			   ar->hw_params.fw.dir, filename, len);
+		ret = -EINVAL;
+		goto err;
+	}
 
-	node = ar->dev->of_node;
-	if (!node)
-		return -ENOENT;
+	if (memcmp(data, ATH10K_BOARD_MAGIC, magic_len)) {
+		ath10k_err(ar, "found invalid board magic\n");
+		ret = -EINVAL;
+		goto err;
+	}
 
-	of_property_read_string(node, "qcom,ath10k-calibration-variant",
-				&variant);
-	if (!variant)
-		return -ENODATA;
+	/* magic is padded to 4 bytes */
+	magic_len = ALIGN(magic_len, 4);
+	if (len < magic_len) {
+		ath10k_err(ar, "failed: %s/%s too small to contain board data, len: %zu\n",
+			   ar->hw_params.fw.dir, filename, len);
+		ret = -EINVAL;
+		goto err;
+	}
 
-	if (strscpy(ar->id.bdf_ext, variant, sizeof(ar->id.bdf_ext)) < 0)
-		ath10k_dbg(ar, ATH10K_DBG_BOOT,
-			   "bdf variant string is longer than the buffer can accommodate (variant: %s)\n",
-			    variant);
+	data += magic_len;
+	len -= magic_len;
+
+	/* attempt to find boardname in the IE list */
+	ret = ath10k_core_search_bd(ar, boardname, data, len);
+
+	/* if we didn't find it and have a fallback name, try that */
+	if (ret == -ENOENT && fallback_boardname)
+		ret = ath10k_core_search_bd(ar, fallback_boardname, data, len);
+
+	if (ret == -ENOENT) {
+		ath10k_err(ar,
+			   "failed to fetch board data for %s from %s/%s\n",
+			   boardname, ar->hw_params.fw.dir, filename);
+		ret = -ENODATA;
+	}
+
+	if (ret)
+		goto err;
 
 	return 0;
+
+err:
+	ath10k_core_free_board_files(ar);
+	return ret;
 }
 
-static int ath10k_download_and_run_otp(struct ath10k *ar)
+static int ath10k_core_create_board_name(struct ath10k *ar, char *name,
+					 size_t name_len, bool with_variant)
 {
-	u32 result, address = ar->hw_params.patch_load_addr;
-	u32 bmi_otp_exe_param = ar->hw_params.otp_exe_param;
-	int ret;
+	/* strlen(',variant=') + strlen(ar->id.bdf_ext) */
+	char variant[9 + ATH10K_SMBIOS_BDF_EXT_STR_LENGTH] = { 0 };
 
-	ret = ath10k_download_board_data(ar,
-					 ar->running_fw->board_data,
-					 ar->running_fw->board_len);
-	if (ret) {
-		ath10k_err(ar, "failed to download board data: %d\n", ret);
-		return ret;
+	if (with_variant && ar->id.bdf_ext[0] != '\0')
+		scnprintf(variant, sizeof(variant), ",variant=%s",
+			  ar->id.bdf_ext);
+
+	if (ar->id.bmi_ids_valid) {
+		scnprintf(name, name_len,
+			  "bus=%s,bmi-chip-id=%d,bmi-board-id=%d%s",
+			  ath10k_bus_str(ar->hif.bus),
+			  ar->id.bmi_chip_id,
+			  ar->id.bmi_board_id, variant);
+		goto out;
 	}
 
-	/* OTP is optional */
+	scnprintf(name, name_len,
+		  "bus=%s,vendor=%04x,device=%04x,subsystem-vendor=%04x,subsystem-device=%04x%s",
+		  ath10k_bus_str(ar->hif.bus),
+		  ar->id.vendor, ar->id.device,
+		  ar->id.subsystem_vendor, ar->id.subsystem_device, variant);
+out:
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot using board name '%s'\n", name);
 
-	if (!ar->running_fw->fw_file.otp_data ||
-	    !ar->running_fw->fw_file.otp_len) {
-		ath10k_warn(ar, "Not running otp, calibration will be incorrect (otp-data %pK otp_len %zd)!\n",
-			    ar->running_fw->fw_file.otp_data,
-			    ar->running_fw->fw_file.otp_len);
-		return 0;
-	}
+	return 0;
+}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot upload otp to 0x%x len %zd\n",
-		   address, ar->running_fw->fw_file.otp_len);
+static int ath10k_core_create_eboard_name(struct ath10k *ar, char *name,
+					  size_t name_len)
+{
+	if (ar->id.bmi_ids_valid) {
+		scnprintf(name, name_len,
+			  "bus=%s,bmi-chip-id=%d,bmi-eboard-id=%d",
+			  ath10k_bus_str(ar->hif.bus),
+			  ar->id.bmi_chip_id,
+			  ar->id.bmi_eboard_id);
 
-	ret = ath10k_bmi_fast_download(ar, address,
-				       ar->running_fw->fw_file.otp_data,
-				       ar->running_fw->fw_file.otp_len);
-	if (ret) {
-		ath10k_err(ar, "could not write otp (%d)\n", ret);
-		return ret;
+		ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot using eboard name '%s'\n", name);
+		return 0;
 	}
+	/* Fallback if returned board id is zero */
+	return -1;
+}
 
-	/* As of now pre-cal is valid for 10_4 variants */
-	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_DT ||
-	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FILE)
-		bmi_otp_exe_param = BMI_PARAM_FLASH_SECTION_ALL;
+static int ath10k_core_fetch_board_file(struct ath10k *ar, int bd_ie_type)
+{
+	char boardname[100], fallback_boardname[100];
+	int ret;
 
-	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_FLASH_MTD)
-		bmi_otp_exe_param = BMI_PARAM_FLASH_SECTION_ALL;
+	if (bd_ie_type == ATH10K_BD_IE_BOARD) {
+		ret = ath10k_core_create_board_name(ar, boardname,
+						    sizeof(boardname), true);
+		if (ret) {
+			ath10k_err(ar, "failed to create board name: %d", ret);
+			return ret;
+		}
 
-        /* Select radio band from user module-param 'dual_band' */
-        if (ath10k_dual_band_config == 2) {
-            bmi_otp_exe_param |= BMI_PARAM_DUAL_BAND_2G;
-            ath10k_dbg(ar, ATH10K_DBG_BOOT,
-                       "Dual-band OTP with 2.4Ghz");
-        } else if (ath10k_dual_band_config == 5) {
-            bmi_otp_exe_param |= BMI_PARAM_DUAL_BAND_5G;
-            ath10k_dbg(ar, ATH10K_DBG_BOOT,
-                       "Dual-band OTP with 5Ghz");
-        }
-	
-	ret = ath10k_bmi_execute(ar, address, bmi_otp_exe_param, &result);
-	if (ret) {
-		ath10k_err(ar, "could not execute otp (%d)\n", ret);
-		return ret;
+		ret = ath10k_core_create_board_name(ar, fallback_boardname,
+						    sizeof(boardname), false);
+		if (ret) {
+			ath10k_err(ar, "failed to create fallback board name: %d", ret);
+			return ret;
+		}
+	} else if (bd_ie_type == ATH10K_BD_IE_BOARD_EXT) {
+		if ((ar->bd_api == 1) ||
+			ath10k_core_create_eboard_name(ar, boardname,
+			 			       sizeof(boardname))) {
+			ath10k_dbg(ar, ATH10K_DBG_BOOT,"fallback to eboard.bin");
+			goto fallback;
+		}
 	}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot otp execute result %d\n", result);
+	ar->bd_api = 2;
+	ret = ath10k_core_fetch_board_data_api_n(ar, boardname,
+						 fallback_boardname,
+						 ATH10K_BOARD_API2_FILE);
+	if (!ret)
+		goto success;
 
-	if (!(skip_otp || test_bit(ATH10K_FW_FEATURE_IGNORE_OTP_RESULT,
-				   ar->running_fw->fw_file.fw_features)) &&
-	    result != 0) {
-		ath10k_err(ar, "otp calibration failed: %d", result);
-		return -EINVAL;
+fallback:
+	ar->bd_api = 1;
+	ret = ath10k_core_fetch_board_data_api_1(ar, bd_ie_type);
+	if (ret) {
+		ath10k_err(ar, "failed to fetch %s from %s\n",
+			   (bd_ie_type == ATH10K_BD_IE_BOARD_EXT) ? "eboard.bin" :
+			   "board-2.bin or board.bin", ar->hw_params.fw.dir);
+		return ret;
 	}
 
+success:
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "using board api %d\n", ar->bd_api);
 	return 0;
 }
 
-static int ath10k_download_fw(struct ath10k *ar)
+static int ath10k_core_get_ext_board_id_from_otp(struct ath10k *ar)
 {
-	u32 address, data_len;
-	const void *data;
+	u32 result, address;
+	u8 ext_board_id;
 	int ret;
 
 	address = ar->hw_params.patch_load_addr;
 
-	data = ar->running_fw->fw_file.firmware_data;
-	data_len = ar->running_fw->fw_file.firmware_len;
+	if (!ar->normal_mode_fw.fw_file.otp_data ||
+	    !ar->normal_mode_fw.fw_file.otp_len) {
+		ath10k_warn(ar,
+			    "failed to retrieve ext board id ,invalid otp\n");
+		return -ENODATA;
+	}
+
+	ath10k_dbg(ar, ATH10K_DBG_BOOT,
+		   "boot upload otp to 0x%x len %zd for ext board id\n",
+		   address, ar->normal_mode_fw.fw_file.otp_len);
 
-	ret = ath10k_swap_code_seg_configure(ar, &ar->running_fw->fw_file);
+	ret = ath10k_bmi_fast_download(ar, address,
+				       ar->normal_mode_fw.fw_file.otp_data,
+				       ar->normal_mode_fw.fw_file.otp_len);
 	if (ret) {
-		ath10k_err(ar, "failed to configure fw code swap: %d\n",
+		ath10k_err(ar, "could not write otp for ext board id check: %d\n",
 			   ret);
 		return ret;
 	}
 
-	ath10k_dbg(ar, ATH10K_DBG_BOOT,
-		   "boot uploading firmware image %pK len %d\n",
-		   data, data_len);
-
-	ret = ath10k_bmi_fast_download(ar, address, data, data_len);
+	ret = ath10k_bmi_execute(ar, address, BMI_PARAM_GET_EXT_BOARD_ID, &result);
 	if (ret) {
-		ath10k_err(ar, "failed to download firmware: %d\n",
+		ath10k_err(ar, "could not execute otp for ext board id check: %d\n",
 			   ret);
 		return ret;
 	}
 
-	return ret;
-}
+	if (!result) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "ext board id does not exist in otp, ignore it\n");
+		return -EOPNOTSUPP;
+	}
 
-static void ath10k_core_free_board_files(struct ath10k *ar)
-{
-	if (!IS_ERR(ar->normal_mode_fw.board))
-		release_firmware(ar->normal_mode_fw.board);
+	ext_board_id = (result & ATH10K_BMI_EBOARD_ID_STATUS_MASK);
 
-	ar->normal_mode_fw.board = NULL;
-	ar->normal_mode_fw.board_data = NULL;
-	ar->normal_mode_fw.board_len = 0;
+	ath10k_dbg(ar, ATH10K_DBG_BOOT,
+		   "boot get otp ext board id result 0x%08x ext_board_id %d\n",
+		   result, ext_board_id);
+
+	ar->id.bmi_eboard_id = ext_board_id;
+
+	return 0;
 }
 
-static void ath10k_core_free_firmware_files(struct ath10k *ar)
+static int ath10k_download_board_data(struct ath10k *ar, const void *data,
+				      size_t data_len)
 {
-	if (!IS_ERR(ar->normal_mode_fw.fw_file.firmware))
-		release_firmware(ar->normal_mode_fw.fw_file.firmware);
-
-	if (!IS_ERR(ar->cal_file))
-		release_firmware(ar->cal_file);
+	u32 board_data_size = ar->hw_params.fw.board_size;
+	u32 board_address;
+	int ret;
 
-	if (!IS_ERR(ar->pre_cal_file))
-		release_firmware(ar->pre_cal_file);
+	ret = ath10k_push_board_ext_data(ar, data, data_len);
+	if (ret) {
+		ath10k_err(ar, "could not push board ext data (%d)\n", ret);
+		goto exit;
+	}
 
-	ath10k_swap_code_seg_release(ar, &ar->normal_mode_fw.fw_file);
+	ret = ath10k_bmi_read32(ar, hi_board_data, &board_address);
+	if (ret) {
+		ath10k_err(ar, "could not read board data addr (%d)\n", ret);
+		goto exit;
+	}
 
-	ar->normal_mode_fw.fw_file.otp_data = NULL;
-	ar->normal_mode_fw.fw_file.otp_len = 0;
+	ret = ath10k_bmi_write_memory(ar, board_address, data,
+				      min_t(u32, board_data_size,
+					    data_len));
+	if (ret) {
+		ath10k_err(ar, "could not write board data (%d)\n", ret);
+		goto exit;
+	}
 
-	ar->normal_mode_fw.fw_file.firmware = NULL;
-	ar->normal_mode_fw.fw_file.firmware_data = NULL;
-	ar->normal_mode_fw.fw_file.firmware_len = 0;
+	ret = ath10k_bmi_write32(ar, hi_board_data_initialized, 1);
+	if (ret) {
+		ath10k_err(ar, "could not write board data bit (%d)\n", ret);
+		goto exit;
+	}
 
-	ar->cal_file = NULL;
-	ar->pre_cal_file = NULL;
+exit:
+       return ret;
 }
 
-static int ath10k_fetch_cal_file(struct ath10k *ar)
+static int ath10k_download_eboard_data(struct ath10k *ar)
 {
-	char filename[100];
-
-	/* pre-cal-<bus>-<id>.bin */
-	scnprintf(filename, sizeof(filename), "pre-cal-%s-%s.bin",
-		  ath10k_bus_str(ar->hif.bus), dev_name(ar->dev));
-
-	ar->pre_cal_file = ath10k_fetch_fw_file(ar, ATH10K_FW_DIR, filename);
-	if (!IS_ERR(ar->pre_cal_file))
-		goto success;
-
-	/* cal-<bus>-<id>.bin */
-	scnprintf(filename, sizeof(filename), "cal-%s-%s.bin",
-		  ath10k_bus_str(ar->hif.bus), dev_name(ar->dev));
+	u32 eboard_data_size = ar->hw_params.fw.ext_board_size;
+	u32 board_address, ext_board_address;
+	int ret;
 
-	ar->cal_file = ath10k_fetch_fw_file(ar, ATH10K_FW_DIR, filename);
-	if (IS_ERR(ar->cal_file))
-		/* calibration file is optional, don't print any warnings */
-		return PTR_ERR(ar->cal_file);
-success:
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "found calibration file %s/%s\n",
-		   ATH10K_FW_DIR, filename);
+	if (!ar->id.ext_bid_supported)
+		return -EOPNOTSUPP;
 
-	return 0;
-}
+	ret = ath10k_bmi_read32(ar, hi_board_data, &board_address);
+	if (ret) {
+		ath10k_err(ar, "could not read board data addr (%d)\n", ret);
+		goto exit;
+	}
 
-static int ath10k_core_fetch_board_data_api_1(struct ath10k *ar)
-{
-	if (!ar->hw_params.fw.board) {
-		ath10k_err(ar, "failed to find board file fw entry\n");
-		return -EINVAL;
+	/* Extended board data download */
+	ret = ath10k_core_get_ext_board_id_from_otp(ar);
+	/* Not fetching ext_board_data if ext board id is 0 */
+	if (ret == -EOPNOTSUPP) {
+		ath10k_err(ar, "otp returned ext board id 0\n");
+		return 0;
+	} else if (ret) {
+		ath10k_err(ar, "could not get ext board id, (%d)\n", ret);
+		goto exit;
 	}
 
-	ar->normal_mode_fw.board = ath10k_fetch_fw_file(ar,
-							ar->hw_params.fw.dir,
-							ar->hw_params.fw.board);
-	if (IS_ERR(ar->normal_mode_fw.board))
-		return PTR_ERR(ar->normal_mode_fw.board);
+	ret = ath10k_core_fetch_board_file(ar, ATH10K_BD_IE_BOARD_EXT);
+	if (ret)
+		goto exit;
 
-	ar->normal_mode_fw.board_data = ar->normal_mode_fw.board->data;
-	ar->normal_mode_fw.board_len = ar->normal_mode_fw.board->size;
+	if (ar->normal_mode_fw.ext_board_data) {
+		ext_board_address = board_address + EXT_BOARD_ADDRESS_OFFSET;
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "writing ext board data to addr 0x%x",
+							ext_board_address);
+		ret = ath10k_bmi_write_memory(ar, ext_board_address,
+					      ar->normal_mode_fw.ext_board_data,
+					      min_t(u32, eboard_data_size,
+					      ar->normal_mode_fw.ext_board_len));
+		if (ret)
+			ath10k_err(ar, "could not write ext board data (%d)\n", ret);
+	}
 
-	return 0;
+exit:
+	return ret;
 }
 
-static int ath10k_core_parse_bd_ie_board(struct ath10k *ar,
-					 const void *buf, size_t buf_len,
-					 const char *boardname)
+static int ath10k_verify_mtd_cal_data_chksum(u32 *data, u32 len)
 {
-	const struct ath10k_fw_ie *hdr;
-	bool name_match_found;
-	int ret, board_ie_id;
-	size_t board_ie_len;
-	const void *board_ie_data;
+	u16 *buf = (u16 *)data;
+	u16 sum;
+	int i;
 
-	name_match_found = false;
+	sum = 0;
 
-	/* go through ATH10K_BD_IE_BOARD_ elements */
-	while (buf_len > sizeof(struct ath10k_fw_ie)) {
-		hdr = buf;
-		board_ie_id = le32_to_cpu(hdr->id);
-		board_ie_len = le32_to_cpu(hdr->len);
-		board_ie_data = hdr->data;
+	for (i = 0; i < (len / 2); i++)
+		sum ^= le16_to_cpu(*buf++);
 
-		buf_len -= sizeof(*hdr);
-		buf += sizeof(*hdr);
+	if (sum != 0xffff)
+		return -EINVAL;
 
-		if (buf_len < ALIGN(board_ie_len, 4)) {
-			ath10k_err(ar, "invalid ATH10K_BD_IE_BOARD length: %zu < %zu\n",
-				   buf_len, ALIGN(board_ie_len, 4));
-			ret = -EINVAL;
-			goto out;
-		}
+	return 0;
+}
 
-		switch (board_ie_id) {
-		case ATH10K_BD_IE_BOARD_NAME:
-			ath10k_dbg_dump(ar, ATH10K_DBG_BOOT, "board name", "",
-					board_ie_data, board_ie_len);
+static int ath10k_download_cal_data_mtd(struct ath10k *ar)
+{
+	struct device_node *node;
+	struct mtd_info *mtd;
+	size_t retlen;
+	u32 *buf, addr, len;
+	const char *mtd_name;
+	int ret, i;
 
-			if (board_ie_len != strlen(boardname))
-				break;
+	node = ar->dev->of_node;
+	if (!node) {
+		/* Device Tree is optional, don't print any warnings if
+		 * there is no node for ath10k.
+		 */
+		return -ENOENT;
+	}
 
-			ret = memcmp(board_ie_data, boardname, strlen(boardname));
-			if (ret)
-				break;
+	ret = of_property_read_u32(node, "qcom,cal-len", &len);
+	if (ret)
+		return -ENOENT;
 
-			name_match_found = true;
-			ath10k_dbg(ar, ATH10K_DBG_BOOT,
-				   "boot found match for name '%s'",
-				   boardname);
-			break;
-		case ATH10K_BD_IE_BOARD_DATA:
-			if (!name_match_found)
-				/* no match found */
-				break;
+	ret = of_property_read_u32(node, "qcom,cal-offset", &addr);
+	if (ret)
+		return -ENOENT;
 
-			ath10k_dbg(ar, ATH10K_DBG_BOOT,
-				   "boot found board data for '%s'",
-				   boardname);
+	mtd_name = of_get_property(node, "qcom,mtd-name", NULL);
+	if (!mtd_name)
+		return -ENOENT;
 
-			ar->normal_mode_fw.board_data = board_ie_data;
-			ar->normal_mode_fw.board_len = board_ie_len;
+	if (len != ar->hw_params.cal_data_len) {
+		ath10k_warn(ar, "invalid cal data length in DT: %d\n", len);
+		return -EMSGSIZE;
+	}
 
-			ret = 0;
-			goto out;
-		default:
-			ath10k_warn(ar, "unknown ATH10K_BD_IE_BOARD found: %d\n",
-				    board_ie_id);
-			break;
-		}
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "mtd name: %s addr: 0x%x len: %d\n",
+		   mtd_name, addr, len);
 
-		/* jump over the padding */
-		board_ie_len = ALIGN(board_ie_len, 4);
+	buf = vmalloc(len);
+	if (!buf) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "failed to alloc memory for mtd caldata\n");
+		return -ENOMEM;
+	}
 
-		buf_len -= board_ie_len;
-		buf += board_ie_len;
+	mtd = get_mtd_device_nm(mtd_name);
+	if (IS_ERR_OR_NULL(mtd)) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "mtd device %s not found to read caldata\n",
+			   mtd_name);
+		ret = PTR_ERR(mtd);
+		goto release_buf;
 	}
 
-	/* no match found */
-	ret = -ENOENT;
+	ret = mtd_read(mtd, addr, len, &retlen, (u_char *)buf);
+	if (ret) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "failed to read caldata from mtd device %s: %d\n",
+			   mtd_name, ret);
+		goto release_mtd;
+	}
 
-out:
-	return ret;
-}
+	if (retlen != len) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "got only %d bytes from mtd (requested: %d)\n",
+			   retlen, len);
+		goto release_mtd;
+	}
 
-static int ath10k_core_search_bd(struct ath10k *ar,
-				 const char *boardname,
-				 const u8 *data,
-				 size_t len)
-{
-	size_t ie_len;
-	struct ath10k_fw_ie *hdr;
-	int ret = -ENOENT, ie_id;
+	ret = ath10k_verify_mtd_cal_data_chksum(buf, len);
+	if (ret) {
+		ath10k_dbg(ar, ATH10K_DBG_BOOT,
+			   "checksum verification failed\n");
+		goto release_mtd;
+	}
 
-	while (len > sizeof(struct ath10k_fw_ie)) {
-		hdr = (struct ath10k_fw_ie *)data;
-		ie_id = le32_to_cpu(hdr->id);
-		ie_len = le32_to_cpu(hdr->len);
+	for (i = 0; i < (len / 4); i++)
+		buf[i] = cpu_to_le32(buf[i]);
 
-		len -= sizeof(*hdr);
-		data = hdr->data;
+	ret = ath10k_download_board_data(ar, (const void *)buf,
+					 (size_t)len);
+	if (ret) {
+		ath10k_err(ar, "failed to download mtd caldata: %d\n", ret);
+		goto release_mtd;
+	}
 
-		if (len < ALIGN(ie_len, 4)) {
-			ath10k_err(ar, "invalid length for board ie_id %d ie_len %zu len %zu\n",
-				   ie_id, ie_len, len);
-			return -EINVAL;
-		}
+	ath10k_dbg(ar, ATH10K_DBG_BOOT,
+		   "cal data downloaded successfully\n");
 
-		switch (ie_id) {
-		case ATH10K_BD_IE_BOARD:
-			ret = ath10k_core_parse_bd_ie_board(ar, data, ie_len,
-							    boardname);
-			if (ret == -ENOENT)
-				/* no match found, continue */
-				break;
+	put_mtd_device(mtd);
+	vfree(buf);
 
-			/* either found or error, so stop searching */
-			goto out;
-		}
+	return 0;
 
-		/* jump over the padding */
-		ie_len = ALIGN(ie_len, 4);
+release_mtd:
+	put_mtd_device(mtd);
 
-		len -= ie_len;
-		data += ie_len;
-	}
+release_buf:
+	vfree(buf);
 
-out:
-	/* return result of parse_bd_ie_board() or -ENOENT */
 	return ret;
 }
 
-static int ath10k_core_fetch_board_data_api_n(struct ath10k *ar,
-					      const char *boardname,
-					      const char *fallback_boardname,
-					      const char *filename)
+static int ath10k_download_and_run_otp(struct ath10k *ar)
 {
-	size_t len, magic_len;
-	const u8 *data;
+	u32 result, address = ar->hw_params.patch_load_addr;
+	u32 bmi_otp_exe_param = ar->hw_params.otp_exe_param;
 	int ret;
 
-	ar->normal_mode_fw.board = ath10k_fetch_fw_file(ar,
-							ar->hw_params.fw.dir,
-							filename);
-	if (IS_ERR(ar->normal_mode_fw.board))
-		return PTR_ERR(ar->normal_mode_fw.board);
-
-	data = ar->normal_mode_fw.board->data;
-	len = ar->normal_mode_fw.board->size;
-
-	/* magic has extra null byte padded */
-	magic_len = strlen(ATH10K_BOARD_MAGIC) + 1;
-	if (len < magic_len) {
-		ath10k_err(ar, "failed to find magic value in %s/%s, file too short: %zu\n",
-			   ar->hw_params.fw.dir, filename, len);
-		ret = -EINVAL;
-		goto err;
+	ret = ath10k_download_board_data(ar,
+					 ar->running_fw->board_data,
+					 ar->running_fw->board_len);
+	if (ret) {
+		ath10k_err(ar, "failed to download board data: %d\n", ret);
+		return ret;
 	}
 
-	if (memcmp(data, ATH10K_BOARD_MAGIC, magic_len)) {
-		ath10k_err(ar, "found invalid board magic\n");
-		ret = -EINVAL;
-		goto err;
+	/* Extended board data download */
+	ret = ath10k_download_eboard_data(ar);
+	if (ret && ret != -EOPNOTSUPP) {
+		ath10k_err(ar, "failed to download eboard data: %d\n", ret);
+		return ret;
 	}
 
-	/* magic is padded to 4 bytes */
-	magic_len = ALIGN(magic_len, 4);
-	if (len < magic_len) {
-		ath10k_err(ar, "failed: %s/%s too small to contain board data, len: %zu\n",
-			   ar->hw_params.fw.dir, filename, len);
-		ret = -EINVAL;
-		goto err;
+	/* OTP is optional */
+
+	if (!ar->running_fw->fw_file.otp_data ||
+	    !ar->running_fw->fw_file.otp_len) {
+		ath10k_warn(ar, "Not running otp, calibration will be incorrect (otp-data %pK otp_len %zd)!\n",
+			    ar->running_fw->fw_file.otp_data,
+			    ar->running_fw->fw_file.otp_len);
+		return 0;
 	}
 
-	data += magic_len;
-	len -= magic_len;
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot upload otp to 0x%x len %zd\n",
+		   address, ar->running_fw->fw_file.otp_len);
 
-	/* attempt to find boardname in the IE list */
-	ret = ath10k_core_search_bd(ar, boardname, data, len);
+	ret = ath10k_bmi_fast_download(ar, address,
+				       ar->running_fw->fw_file.otp_data,
+				       ar->running_fw->fw_file.otp_len);
+	if (ret) {
+		ath10k_err(ar, "could not write otp (%d)\n", ret);
+		return ret;
+	}
 
-	/* if we didn't find it and have a fallback name, try that */
-	if (ret == -ENOENT && fallback_boardname)
-		ret = ath10k_core_search_bd(ar, fallback_boardname, data, len);
+	/* As of now pre-cal is valid for 10_4 variants */
+	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_DT ||
+	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FILE ||
+	    ar->cal_mode == ATH10K_PRE_CAL_MODE_FLASH_MTD)
+		bmi_otp_exe_param = BMI_PARAM_FLASH_SECTION_ALL;
 
-	if (ret == -ENOENT) {
-		ath10k_err(ar,
-			   "failed to fetch board data for %s from %s/%s\n",
-			   boardname, ar->hw_params.fw.dir, filename);
-		ret = -ENODATA;
+	ret = ath10k_bmi_execute(ar, address, bmi_otp_exe_param, &result);
+	if (ret) {
+		ath10k_err(ar, "could not execute otp (%d)\n", ret);
+		return ret;
 	}
 
-	if (ret)
-		goto err;
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot otp execute result %d\n", result);
 
-	return 0;
+	if (!(skip_otp || test_bit(ATH10K_FW_FEATURE_IGNORE_OTP_RESULT,
+				   ar->running_fw->fw_file.fw_features)) &&
+	    result != 0) {
+		ath10k_err(ar, "otp calibration failed: %d", result);
+		return -EINVAL;
+	}
 
-err:
-	ath10k_core_free_board_files(ar);
-	return ret;
+	return 0;
 }
 
-static int ath10k_core_create_board_name(struct ath10k *ar, char *name,
-					 size_t name_len, bool with_variant)
+static int ath10k_download_cal_file(struct ath10k *ar,
+				    const struct firmware *file)
 {
-	/* strlen(',variant=') + strlen(ar->id.bdf_ext) */
-	char variant[9 + ATH10K_SMBIOS_BDF_EXT_STR_LENGTH] = { 0 };
+	int ret;
 
-	if (with_variant && ar->id.bdf_ext[0] != '\0')
-		scnprintf(variant, sizeof(variant), ",variant=%s",
-			  ar->id.bdf_ext);
+	if (!file)
+		return -ENOENT;
 
-	if (ar->id.bmi_ids_valid) {
-		scnprintf(name, name_len,
-			  "bus=%s,bmi-chip-id=%d,bmi-board-id=%d%s",
-			  ath10k_bus_str(ar->hif.bus),
-			  ar->id.bmi_chip_id,
-			  ar->id.bmi_board_id, variant);
-		goto out;
+	if (IS_ERR(file))
+		return PTR_ERR(file);
+
+	ret = ath10k_download_board_data(ar, file->data, file->size);
+	if (ret) {
+		ath10k_err(ar, "failed to download cal_file data: %d\n", ret);
+		return ret;
 	}
 
-	scnprintf(name, name_len,
-		  "bus=%s,vendor=%04x,device=%04x,subsystem-vendor=%04x,subsystem-device=%04x%s",
-		  ath10k_bus_str(ar->hif.bus),
-		  ar->id.vendor, ar->id.device,
-		  ar->id.subsystem_vendor, ar->id.subsystem_device, variant);
-out:
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot using board name '%s'\n", name);
+	ath10k_dbg(ar, ATH10K_DBG_BOOT, "boot cal file downloaded\n");
 
 	return 0;
 }
 
-static int ath10k_core_fetch_board_file(struct ath10k *ar)
+static int ath10k_download_cal_dt(struct ath10k *ar, const char *dt_name)
 {
-	char boardname[100], fallback_boardname[100];
+	struct device_node *node;
+	int data_len;
+	void *data;
 	int ret;
 
-	ret = ath10k_core_create_board_name(ar, boardname,
-					    sizeof(boardname), true);
+	node = ar->dev->of_node;
+	if (!node)
+		/* Device Tree is optional, don't print any warnings if
+		 * there's no node for ath10k.
+		 */
+		return -ENOENT;
+
+	if (!of_get_property(node, dt_name, &data_len)) {
+		/* The calibration data node is optional */
+		return -ENOENT;
+	}
+
+	if (data_len != ar->hw_params.cal_data_len) {
+		ath10k_warn(ar, "invalid calibration data length in DT: %d\n",
+			    data_len);
+		ret = -EMSGSIZE;
+		goto out;
+	}
+
+	data = kmalloc(data_len, GFP_KERNEL);
+	if (!data) {
+		ret = -ENOMEM;
+		goto out;
+	}
+
+	ret = of_property_read_u8_array(node, dt_name, data, data_len);
 	if (ret) {
-		ath10k_err(ar, "failed to create board name: %d", ret);
-		return ret;
+		ath10k_warn(ar, "failed to read calibration data from DT: %d\n",
+			    ret);
+		goto out_free;
 	}
 
-	ret = ath10k_core_create_board_name(ar, fallback_boardname,
-					    sizeof(boardname), false);
+	ret = ath10k_download_board_data(ar, data, data_len);
 	if (ret) {
-		ath10k_err(ar, "failed to create fallback board name: %d", ret);
-		return ret;
+		ath10k_warn(ar, "failed to download calibration data from Device Tree: %d\n",
+			    ret);
+		goto out_free;
 	}
 
-	ar->bd_api = 2;
-	ret = ath10k_core_fetch_board_data_api_n(ar, boardname,
-						 fallback_boardname,
-						 ATH10K_BOARD_API2_FILE);
-	if (!ret)
-		goto success;
+	ret = 0;
 
-	ar->bd_api = 1;
-	ret = ath10k_core_fetch_board_data_api_1(ar);
+out_free:
+	kfree(data);
+
+out:
+	return ret;
+}
+
+static int ath10k_download_cal_eeprom(struct ath10k *ar)
+{
+	size_t data_len;
+	void *data = NULL;
+	int ret;
+
+	ret = ath10k_hif_fetch_cal_eeprom(ar, &data, &data_len);
 	if (ret) {
-		ath10k_err(ar, "failed to fetch board-2.bin or board.bin from %s\n",
-			   ar->hw_params.fw.dir);
-		return ret;
+		if (ret != -EOPNOTSUPP)
+			ath10k_warn(ar, "failed to read calibration data from EEPROM: %d\n",
+				    ret);
+		goto out_free;
 	}
 
-success:
-	ath10k_dbg(ar, ATH10K_DBG_BOOT, "using board api %d\n", ar->bd_api);
-	return 0;
+	ret = ath10k_download_board_data(ar, data, data_len);
+	if (ret) {
+		ath10k_warn(ar, "failed to download calibration data from EEPROM: %d\n",
+			    ret);
+		goto out_free;
+	}
+
+	ret = 0;
+
+out_free:
+	kfree(data);
+
+	return ret;
 }
 
 int ath10k_core_fetch_firmware_api_n(struct ath10k *ar, const char *name,
@@ -2835,7 +2992,7 @@
 			ath10k_dbg(ar, ATH10K_DBG_BOOT, "DT bdf variant name not set.\n");
 
 skip_get_board_id:
-		ret = ath10k_core_fetch_board_file(ar);
+		ret = ath10k_core_fetch_board_file(ar, ATH10K_BD_IE_BOARD);
 		if (ret) {
 			ath10k_err(ar, "failed to fetch board file: %d\n", ret);
 			goto err_free_firmware_files;
Index: backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/core.h
===================================================================
--- backports-20180529-3.14.77-8442d41b.orig/drivers/net/wireless/ath/ath10k/core.h	2018-08-24 11:53:56.403772150 +0530
+++ backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/core.h	2018-08-24 11:53:56.399772151 +0530
@@ -947,6 +947,9 @@
 	const struct firmware *board;
 	const void *board_data;
 	size_t board_len;
+	const struct firmware *ext_board;
+	const void *ext_board_data;
+	size_t ext_board_len;
 
 	struct ath10k_fw_file fw_file;
 };
@@ -1035,7 +1038,9 @@
 
 		bool bmi_ids_valid;
 		u8 bmi_board_id;
+		u8 bmi_eboard_id;
 		u8 bmi_chip_id;
+		bool ext_bid_supported;
 
 		char bdf_ext[ATH10K_SMBIOS_BDF_EXT_STR_LENGTH];
 	} id;
Index: backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/hw.h
===================================================================
--- backports-20180529-3.14.77-8442d41b.orig/drivers/net/wireless/ath/ath10k/hw.h	2018-08-24 11:53:56.403772150 +0530
+++ backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/hw.h	2018-08-24 11:53:56.399772151 +0530
@@ -109,6 +109,7 @@
 #define QCA9984_HW_1_0_CHIP_ID_REV	0x0
 #define QCA9984_HW_1_0_FW_DIR		ATH10K_FW_DIR "/QCA9984/hw1.0"
 #define QCA9984_HW_1_0_BOARD_DATA_FILE "board.bin"
+#define QCA9984_HW_1_0_EBOARD_DATA_FILE "eboard.bin"
 #define QCA9984_HW_1_0_PATCH_LOAD_ADDR	0x1234
 
 /* QCA9888 2.0 defines */
@@ -221,6 +222,7 @@
 enum ath10k_bd_ie_type {
 	/* contains sub IEs of enum ath10k_bd_ie_board_type */
 	ATH10K_BD_IE_BOARD = 0,
+	ATH10K_BD_IE_BOARD_EXT = 1,
 };
 
 enum ath10k_bd_ie_board_type {
@@ -548,6 +550,8 @@
 		const char *dir;
 		const char *board;
 		size_t board_size;
+		const char *eboard;
+		size_t ext_board_size;
 		size_t board_ext_size;
 	} fw;
 
Index: backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/targaddrs.h
===================================================================
--- backports-20180529-3.14.77-8442d41b.orig/drivers/net/wireless/ath/ath10k/targaddrs.h	2018-08-24 11:53:56.403772150 +0530
+++ backports-20180529-3.14.77-8442d41b/drivers/net/wireless/ath/ath10k/targaddrs.h	2018-08-24 11:53:56.399772151 +0530
@@ -484,6 +484,10 @@
 #define QCA99X0_BOARD_DATA_SZ	  12288
 #define QCA99X0_BOARD_EXT_DATA_SZ 0
 
+/* Dual band extended board data */
+#define QCA99X0_EXT_BOARD_DATA_SZ 2048
+#define EXT_BOARD_ADDRESS_OFFSET 0x3000
+
 #define QCA4019_BOARD_DATA_SZ	  12064
 #define QCA4019_BOARD_EXT_DATA_SZ 0
 
