--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -366,6 +366,10 @@ struct ath10k_tx_stats {
 	u64 ba_fails;
 };
 
+struct ath10k_txq {
+	/* TBD */
+};
+
 struct ath10k_sta {
 	struct ath10k_vif *arvif;
 
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -7640,6 +7640,7 @@ int ath10k_mac_register(struct ath10k *a
 
 	ar->hw->vif_data_size = sizeof(struct ath10k_vif);
 	ar->hw->sta_data_size = sizeof(struct ath10k_sta);
+	ar->hw->txq_data_size = sizeof(struct ath10k_txq);
 
 	ar->hw->max_listen_interval = ATH10K_MAX_HW_LISTEN_INTERVAL;
 
