--- a/drivers/net/wireless/ath/ath10k/htt.h
+++ b/drivers/net/wireless/ath/ath10k/htt.h
@@ -1659,9 +1659,11 @@ struct ath10k_htt {
 	/* This is used to group tx/rx completions separately and process them
 	 * in batches to reduce cache stalls */
 	struct tasklet_struct txrx_compl_task;
+	struct tasklet_struct tx_fetch_ind_task;
 	struct sk_buff_head tx_compl_q;
 	struct sk_buff_head rx_compl_q;
 	struct sk_buff_head rx_in_ord_compl_q;
+	struct sk_buff_head tx_fetch_ind_q;;
 
 	/* rx_status template */
 	struct ieee80211_rx_status rx_status;
@@ -1685,7 +1687,6 @@ struct ath10k_htt {
 		u16 num_tids;
 		enum htt_tx_mode_switch_mode mode;
 		enum htt_q_depth_type type;
-		struct idr resp_ids;
 	} tx_q_state;
 };
 
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -35,6 +35,7 @@
 
 static int ath10k_htt_rx_get_csum_state(struct sk_buff *skb);
 static void ath10k_htt_txrx_compl_task(unsigned long ptr);
+static void ath10k_htt_rx_tx_fetch_task(unsigned long ptr);
 
 static struct sk_buff *
 ath10k_htt_rx_find_skb_paddr(struct ath10k *ar, u32 paddr)
@@ -227,10 +228,12 @@ void ath10k_htt_rx_free(struct ath10k_ht
 	del_timer_sync(&htt->rx_ring.refill_retry_timer);
 	tasklet_kill(&htt->rx_replenish_task);
 	tasklet_kill(&htt->txrx_compl_task);
+	tasklet_kill(&htt->tx_fetch_ind_task);
 
 	skb_queue_purge(&htt->tx_compl_q);
 	skb_queue_purge(&htt->rx_compl_q);
 	skb_queue_purge(&htt->rx_in_ord_compl_q);
+	skb_queue_purge(&htt->tx_fetch_ind_q);
 
 	ath10k_htt_rx_ring_free(htt);
 
@@ -568,10 +571,13 @@ int ath10k_htt_rx_alloc(struct ath10k_ht
 
 	tasklet_init(&htt->rx_replenish_task, ath10k_htt_rx_replenish_task,
 		     (unsigned long)htt);
+	tasklet_init(&htt->tx_fetch_ind_task, ath10k_htt_rx_tx_fetch_task,
+		     (unsigned long)htt);
 
 	skb_queue_head_init(&htt->tx_compl_q);
 	skb_queue_head_init(&htt->rx_compl_q);
 	skb_queue_head_init(&htt->rx_in_ord_compl_q);
+	skb_queue_head_init(&htt->tx_fetch_ind_q);
 
 	tasklet_init(&htt->txrx_compl_task, ath10k_htt_txrx_compl_task,
 		     (unsigned long)htt);
@@ -1987,26 +1993,6 @@ static void ath10k_htt_rx_in_ord_ind(str
 	tasklet_schedule(&htt->rx_replenish_task);
 }
 
-static void ath10k_htt_rx_tx_fetch_resp_id_confirm(struct ath10k *ar,
-						   const __le32 *resp_ids,
-						   int num_resp_ids)
-{
-	int i;
-	u32 resp_id;
-
-	ath10k_dbg(ar, ATH10K_DBG_HTT, "htt rx tx fetch confirm num_resp_ids %d\n",
-		   num_resp_ids);
-
-	for (i = 0; i < num_resp_ids; i++) {
-		resp_id = le32_to_cpu(resp_ids[i]);
-
-		ath10k_dbg(ar, ATH10K_DBG_HTT, "htt rx tx fetch confirm resp_id %u\n",
-			   resp_id);
-
-		idr_remove(&ar->htt.tx_q_state.resp_ids, resp_id);
-	}
-}
-
 static void ath10k_htt_rx_tx_fetch_ind(struct ath10k *ar, struct sk_buff *skb)
 {
 	struct ieee80211_hw *hw = ar->hw;
@@ -2018,7 +2004,6 @@ static void ath10k_htt_rx_tx_fetch_ind(s
 	size_t max_num_msdus;
 	size_t num_bytes;
 	size_t num_msdus;
-	const __le32 *resp_ids;
 	u16 num_records;
 	u16 num_resp_ids;
 	u16 peer_id;
@@ -2108,9 +2093,6 @@ static void ath10k_htt_rx_tx_fetch_ind(s
 
 	rcu_read_unlock();
 
-	resp_ids = ath10k_htt_get_tx_fetch_ind_resp_ids(&resp->tx_fetch_ind);
-	ath10k_htt_rx_tx_fetch_resp_id_confirm(ar, resp_ids, num_resp_ids);
-
 	ret = ath10k_htt_tx_fetch_resp(ar,
 				       resp->tx_fetch_ind.token,
 				       resp->tx_fetch_ind.fetch_seq_num,
@@ -2125,34 +2107,6 @@ static void ath10k_htt_rx_tx_fetch_ind(s
 	ath10k_mac_txq_state_update_commit(ar);
 }
 
-static void ath10k_htt_rx_tx_fetch_confirm(struct ath10k *ar,
-					   struct sk_buff *skb)
-{
-	const struct htt_resp *resp = (void *)skb->data;
-	size_t len;
-	int num_resp_ids;
-
-	ath10k_dbg(ar, ATH10K_DBG_HTT, "htt rx tx fetch confirm\n");
-
-	len = sizeof(resp->hdr) + sizeof(resp->tx_fetch_confirm);
-	if (unlikely(skb->len < len)) {
-		ath10k_warn(ar, "received corrupted tx_fetch_confirm event: buffer too short\n");
-		return;
-	}
-
-	num_resp_ids = le16_to_cpu(resp->tx_fetch_confirm.num_resp_ids);
-	len += sizeof(resp->tx_fetch_confirm.resp_ids[0]) * num_resp_ids;
-
-	if (unlikely(skb->len < len)) {
-		ath10k_warn(ar, "received corrupted tx_fetch_confirm event: resp_ids buffer overflow\n");
-		return;
-	}
-
-	ath10k_htt_rx_tx_fetch_resp_id_confirm(ar,
-					       resp->tx_fetch_confirm.resp_ids,
-					       num_resp_ids);
-}
-
 static void ath10k_htt_rx_tx_mode_switch_ind(struct ath10k *ar,
 					     struct sk_buff *skb)
 {
@@ -2389,10 +2343,14 @@ void ath10k_htt_t2h_msg_handler(struct a
 	case HTT_T2H_MSG_TYPE_AGGR_CONF:
 		break;
 	case HTT_T2H_MSG_TYPE_TX_FETCH_IND:
-		ath10k_htt_rx_tx_fetch_ind(ar, skb);
-		break;
+		skb_queue_tail(&htt->tx_fetch_ind_q, skb);
+		tasklet_schedule(&htt->tx_fetch_ind_task);
+		return;
 	case HTT_T2H_MSG_TYPE_TX_FETCH_CONFIRM:
-		ath10k_htt_rx_tx_fetch_confirm(ar, skb);
+		/* Firmware doesn't care about response ids. It is only
+		 * provided for host convienence so if it doesn't use it then
+		 * there's no need to process these events at all.
+		 */
 		break;
 	case HTT_T2H_MSG_TYPE_TX_MODE_SWITCH_IND:
 		ath10k_htt_rx_tx_mode_switch_ind(ar, skb);
@@ -2449,3 +2407,15 @@ static void ath10k_htt_txrx_compl_task(u
 	}
 	spin_unlock_bh(&htt->rx_ring.lock);
 }
+
+static void ath10k_htt_rx_tx_fetch_task(unsigned long ptr)
+{
+	struct ath10k_htt *htt = (void *)ptr;
+	struct ath10k *ar = htt->ar;
+	struct sk_buff *skb;
+
+	while ((skb = skb_dequeue(&htt->tx_fetch_ind_q))) {
+		ath10k_htt_rx_tx_fetch_ind(ar, skb);
+		dev_kfree_skb_any(skb);
+	}
+}
--- a/drivers/net/wireless/ath/ath10k/htt_tx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_tx.c
@@ -177,7 +177,6 @@ int ath10k_htt_tx_alloc(struct ath10k_ht
 
 	spin_lock_init(&htt->tx_lock);
 	idr_init(&htt->pending_tx);
-	idr_init(&htt->tx_q_state.resp_ids);
 
 	size = htt->max_num_pending_tx * sizeof(struct ath10k_htt_txbuf);
 	htt->txbuf.vaddr = dma_alloc_coherent(ar->dev, size,
@@ -214,7 +213,6 @@ free_txbuf:
 
 free_idr:
 	idr_destroy(&htt->pending_tx);
-	idr_destroy(&htt->tx_q_state.resp_ids);
 
 	return ret;
 }
@@ -241,7 +239,6 @@ void ath10k_htt_tx_free(struct ath10k_ht
 
 	idr_for_each(&htt->pending_tx, ath10k_htt_tx_clean_up_pending, htt->ar);
 	idr_destroy(&htt->pending_tx);
-	idr_destroy(&htt->tx_q_state.resp_ids);
 
 	if (htt->txbuf.vaddr) {
 		size = htt->max_num_pending_tx *
@@ -543,7 +540,6 @@ int ath10k_htt_tx_fetch_resp(struct ath1
 {
 	struct sk_buff *skb;
 	struct htt_cmd *cmd;
-	u16 resp_id;
 	int len = 0;
 	int ret;
 
@@ -555,17 +551,10 @@ int ath10k_htt_tx_fetch_resp(struct ath1
 	if (!skb)
 		return -ENOMEM;
 
-	ret = idr_alloc(&ar->htt.tx_q_state.resp_ids, ar, 0,
-			BIT(sizeof(resp_id) * 8) - 1, GFP_ATOMIC);
-	if (ret < 0)
-		goto err_free_skb;
-
-	resp_id = ret;
-
 	skb_put(skb, len);
 	cmd = (struct htt_cmd *)skb->data;
 	cmd->hdr.msg_type = HTT_H2T_MSG_TYPE_TX_FETCH_RESP;
-	cmd->tx_fetch_resp.resp_id = cpu_to_le16(resp_id);
+	cmd->tx_fetch_resp.resp_id = 0;
 	cmd->tx_fetch_resp.fetch_seq_num = fetch_seq_num;
 	cmd->tx_fetch_resp.num_records = cpu_to_le16(num_records);
 	cmd->tx_fetch_resp.token = token;
@@ -576,14 +565,11 @@ int ath10k_htt_tx_fetch_resp(struct ath1
 	ret = ath10k_htc_send(&ar->htc, ar->htt.eid, skb);
 	if (ret) {
 		ath10k_warn(ar, "failed to submit htc command: %d\n", ret);
-		goto err_free_resp_id;
+		goto err_free_skb;
 	}
 
 	return 0;
 
-err_free_resp_id:
-	idr_remove(&ar->htt.tx_q_state.resp_ids, resp_id);
-
 err_free_skb:
 	dev_kfree_skb_any(skb);
 
