--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -54,12 +54,12 @@ struct ieee80211_local;
 /* The number of stations exposed via netdev queues. First 4 netdev queues are
  * mapped to vif's ACs. Subsequent ones, in groups of 4, are ACs for stations.
  * NUM_NDEV_STA + 1 station is wrapped around to 1st station. f.e. netdev queue
- * 5 corresponds to AC1 of STA0, STA63, STA126, ... and queue 8 corresponds
- * to AC0 of STA1, STA64, STA127, ...
+ * 5 corresponds to AC1 of STA0, STA16, STA32, ... and queue 8 corresponds
+ * to AC0 of STA1, STA17, STA33, ...
  *
  * This is used only when driver implements wake_tx_queues() op.
  */
-#define IEEE80211_NUM_NDEV_STA 63
+#define IEEE80211_NUM_NDEV_STA 16
 #define IEEE80211_NUM_NDEV_STA_Q (IEEE80211_NUM_NDEV_STA * IEEE80211_NUM_ACS)
 
 /* Required encryption head and tailroom */
