--- a/drivers/net/wireless/ath/ath10k/hw.h
+++ b/drivers/net/wireless/ath/ath10k/hw.h
@@ -337,14 +337,24 @@ enum ath10k_hw_rate_ofdm {
 	ATH10K_HW_RATE_OFDM_9M,
 };
 
-enum ath10k_hw_rate_cck {
-	ATH10K_HW_RATE_CCK_LP_11M = 0,
-	ATH10K_HW_RATE_CCK_LP_5_5M,
-	ATH10K_HW_RATE_CCK_LP_2M,
-	ATH10K_HW_RATE_CCK_LP_1M,
-	ATH10K_HW_RATE_CCK_SP_11M,
-	ATH10K_HW_RATE_CCK_SP_5_5M,
-	ATH10K_HW_RATE_CCK_SP_2M,
+enum ath10k_hw_rate_10_2_cck {
+	ATH10K_HW_RATE_10_2_CCK_LP_11M = 0,
+	ATH10K_HW_RATE_10_2_CCK_LP_5_5M,
+	ATH10K_HW_RATE_10_2_CCK_LP_2M,
+	ATH10K_HW_RATE_10_2_CCK_LP_1M,
+	ATH10K_HW_RATE_10_2_CCK_SP_11M,
+	ATH10K_HW_RATE_10_2_CCK_SP_5_5M,
+	ATH10K_HW_RATE_10_2_CCK_SP_2M,
+};
+
+enum ath10k_hw_rate_10_4_cck {
+	ATH10K_HW_RATE_10_4_CCK_LP_1M = 1,
+	ATH10K_HW_RATE_10_4_CCK_LP_2M,
+	ATH10K_HW_RATE_10_4_CCK_LP_5_5M,
+	ATH10K_HW_RATE_10_4_CCK_LP_11M,
+	ATH10K_HW_RATE_10_4_CCK_SP_2M,
+	ATH10K_HW_RATE_10_4_CCK_SP_5_5M,
+	ATH10K_HW_RATE_10_4_CCK_SP_11M,
 };
 
 enum ath10k_hw_4addr_pad {
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -37,20 +37,46 @@
 /* Rates */
 /*********/
 
-static struct ieee80211_rate ath10k_rates[] = {
+static struct ieee80211_rate ath10k_10_2_rates[] = {
 	{ .bitrate = 10,
-	  .hw_value = ATH10K_HW_RATE_CCK_LP_1M },
+	  .hw_value = ATH10K_HW_RATE_10_2_CCK_LP_1M },
 	{ .bitrate = 20,
-	  .hw_value = ATH10K_HW_RATE_CCK_LP_2M,
-	  .hw_value_short = ATH10K_HW_RATE_CCK_SP_2M,
+	  .hw_value = ATH10K_HW_RATE_10_2_CCK_LP_2M,
+	  .hw_value_short = ATH10K_HW_RATE_10_2_CCK_SP_2M,
 	  .flags = IEEE80211_RATE_SHORT_PREAMBLE },
 	{ .bitrate = 55,
-	  .hw_value = ATH10K_HW_RATE_CCK_LP_5_5M,
-	  .hw_value_short = ATH10K_HW_RATE_CCK_SP_5_5M,
+	  .hw_value = ATH10K_HW_RATE_10_2_CCK_LP_5_5M,
+	  .hw_value_short = ATH10K_HW_RATE_10_2_CCK_SP_5_5M,
 	  .flags = IEEE80211_RATE_SHORT_PREAMBLE },
 	{ .bitrate = 110,
-	  .hw_value = ATH10K_HW_RATE_CCK_LP_11M,
-	  .hw_value_short = ATH10K_HW_RATE_CCK_SP_11M,
+	  .hw_value = ATH10K_HW_RATE_10_2_CCK_LP_11M,
+	  .hw_value_short = ATH10K_HW_RATE_10_2_CCK_SP_11M,
+	  .flags = IEEE80211_RATE_SHORT_PREAMBLE },
+
+	{ .bitrate = 60, .hw_value = ATH10K_HW_RATE_OFDM_6M },
+	{ .bitrate = 90, .hw_value = ATH10K_HW_RATE_OFDM_9M },
+	{ .bitrate = 120, .hw_value = ATH10K_HW_RATE_OFDM_12M },
+	{ .bitrate = 180, .hw_value = ATH10K_HW_RATE_OFDM_18M },
+	{ .bitrate = 240, .hw_value = ATH10K_HW_RATE_OFDM_24M },
+	{ .bitrate = 360, .hw_value = ATH10K_HW_RATE_OFDM_36M },
+	{ .bitrate = 480, .hw_value = ATH10K_HW_RATE_OFDM_48M },
+	{ .bitrate = 540, .hw_value = ATH10K_HW_RATE_OFDM_54M },
+};
+
+static struct ieee80211_rate ath10k_10_4_rates[] = {
+	{ .bitrate = 10,
+	  .hw_value = ATH10K_HW_RATE_10_4_CCK_LP_1M },
+	{ .bitrate = 20,
+	  .hw_value = ATH10K_HW_RATE_10_4_CCK_LP_2M,
+	  .hw_value_short = ATH10K_HW_RATE_10_4_CCK_SP_2M,
+	  .flags = IEEE80211_RATE_SHORT_PREAMBLE },
+	{ .bitrate = 55,
+	  .hw_value = ATH10K_HW_RATE_10_4_CCK_LP_5_5M,
+	  .hw_value_short = ATH10K_HW_RATE_10_4_CCK_SP_5_5M,
+	  .flags = IEEE80211_RATE_SHORT_PREAMBLE },
+	{ .bitrate = 110,
+	  .hw_value = ATH10K_HW_RATE_10_4_CCK_LP_11M,
+	  .hw_value_short = ATH10K_HW_RATE_10_4_CCK_SP_11M,
 	  .flags = IEEE80211_RATE_SHORT_PREAMBLE },
 
 	{ .bitrate = 60, .hw_value = ATH10K_HW_RATE_OFDM_6M },
@@ -65,11 +91,14 @@ static struct ieee80211_rate ath10k_rate
 
 #define ATH10K_MAC_FIRST_OFDM_RATE_IDX 4
 
-#define ath10k_a_rates (ath10k_rates + ATH10K_MAC_FIRST_OFDM_RATE_IDX)
-#define ath10k_a_rates_size (ARRAY_SIZE(ath10k_rates) - \
+#define ath10k_a_rates (ath10k_10_2_rates + ATH10K_MAC_FIRST_OFDM_RATE_IDX)
+#define ath10k_a_rates_size (ARRAY_SIZE(ath10k_10_2_rates) - \
 			     ATH10K_MAC_FIRST_OFDM_RATE_IDX)
-#define ath10k_g_rates (ath10k_rates + 0)
-#define ath10k_g_rates_size (ARRAY_SIZE(ath10k_rates))
+#define ath10k_g_10_2_rates (ath10k_10_2_rates + 0)
+#define ath10k_g_10_2_rates_size (ARRAY_SIZE(ath10k_10_2_rates))
+
+#define ath10k_g_10_4_rates (ath10k_10_4_rates + 0)
+#define ath10k_g_10_4_rates_size (ARRAY_SIZE(ath10k_10_4_rates))
 
 static bool ath10k_mac_bitrate_is_cck(int bitrate)
 {
@@ -7757,8 +7786,14 @@ int ath10k_mac_register(struct ath10k *a
 		band = &ar->mac.sbands[IEEE80211_BAND_2GHZ];
 		band->n_channels = ARRAY_SIZE(ath10k_2ghz_channels);
 		band->channels = channels;
-		band->n_bitrates = ath10k_g_rates_size;
-		band->bitrates = ath10k_g_rates;
+
+		if (ar->wmi.op_version < ATH10K_FW_WMI_OP_VERSION_10_4) {
+			band->n_bitrates = ath10k_g_10_2_rates_size;
+			band->bitrates = ath10k_g_10_2_rates;
+		} else {
+			band->n_bitrates = ath10k_g_10_4_rates_size;
+			band->bitrates = ath10k_g_10_4_rates;
+		}
 
 		ar->hw->wiphy->bands[IEEE80211_BAND_2GHZ] = band;
 	}
--- a/drivers/net/wireless/ath/ath10k/rx_desc.h
+++ b/drivers/net/wireless/ath/ath10k/rx_desc.h
@@ -656,26 +656,6 @@ struct rx_msdu_end {
  *		Reserved: HW should fill with zero.  FW should ignore.
  */
 
-#define RX_PPDU_START_SIG_RATE_SELECT_OFDM 0
-#define RX_PPDU_START_SIG_RATE_SELECT_CCK  1
-
-#define RX_PPDU_START_SIG_RATE_OFDM_48 0
-#define RX_PPDU_START_SIG_RATE_OFDM_24 1
-#define RX_PPDU_START_SIG_RATE_OFDM_12 2
-#define RX_PPDU_START_SIG_RATE_OFDM_6  3
-#define RX_PPDU_START_SIG_RATE_OFDM_54 4
-#define RX_PPDU_START_SIG_RATE_OFDM_36 5
-#define RX_PPDU_START_SIG_RATE_OFDM_18 6
-#define RX_PPDU_START_SIG_RATE_OFDM_9  7
-
-#define RX_PPDU_START_SIG_RATE_CCK_LP_11  0
-#define RX_PPDU_START_SIG_RATE_CCK_LP_5_5 1
-#define RX_PPDU_START_SIG_RATE_CCK_LP_2   2
-#define RX_PPDU_START_SIG_RATE_CCK_LP_1   3
-#define RX_PPDU_START_SIG_RATE_CCK_SP_11  4
-#define RX_PPDU_START_SIG_RATE_CCK_SP_5_5 5
-#define RX_PPDU_START_SIG_RATE_CCK_SP_2   6
-
 #define HTT_RX_PPDU_START_PREAMBLE_LEGACY        0x04
 #define HTT_RX_PPDU_START_PREAMBLE_HT            0x08
 #define HTT_RX_PPDU_START_PREAMBLE_HT_WITH_TXBF  0x09
@@ -711,25 +691,6 @@ struct rx_msdu_end {
 /* No idea what this flag means. It seems to be always set in rate. */
 #define RX_PPDU_START_RATE_FLAG BIT(3)
 
-enum rx_ppdu_start_rate {
-	RX_PPDU_START_RATE_OFDM_48M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_48M,
-	RX_PPDU_START_RATE_OFDM_24M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_24M,
-	RX_PPDU_START_RATE_OFDM_12M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_12M,
-	RX_PPDU_START_RATE_OFDM_6M  = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_6M,
-	RX_PPDU_START_RATE_OFDM_54M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_54M,
-	RX_PPDU_START_RATE_OFDM_36M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_36M,
-	RX_PPDU_START_RATE_OFDM_18M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_18M,
-	RX_PPDU_START_RATE_OFDM_9M  = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_OFDM_9M,
-
-	RX_PPDU_START_RATE_CCK_LP_11M  = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_LP_11M,
-	RX_PPDU_START_RATE_CCK_LP_5_5M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_LP_5_5M,
-	RX_PPDU_START_RATE_CCK_LP_2M   = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_LP_2M,
-	RX_PPDU_START_RATE_CCK_LP_1M   = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_LP_1M,
-	RX_PPDU_START_RATE_CCK_SP_11M  = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_SP_11M,
-	RX_PPDU_START_RATE_CCK_SP_5_5M = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_SP_5_5M,
-	RX_PPDU_START_RATE_CCK_SP_2M   = RX_PPDU_START_RATE_FLAG | ATH10K_HW_RATE_CCK_SP_2M,
-};
-
 struct rx_ppdu_start {
 	struct {
 		u8 pri20_mhz;
