--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -29,8 +29,10 @@ static unsigned int ath10k_cryptmode_par
 static bool uart_print;
 static bool skip_otp;
 static bool rawmode;
+bool bug_on_fw_assert = true;
 
 unsigned long ath10k_coredump_mask = BIT(ATH10K_FW_CRASH_DUMP_REGISTERS) |
+				     BIT(ATH10K_FW_CRASH_DUMP_RAM_DATA) |
 				     BIT(ATH10K_FW_CRASH_DUMP_CE_DATA);
 
 /* FIXME: most of these should be readonly */
@@ -40,6 +42,7 @@ module_param(uart_print, bool, 0644);
 module_param(skip_otp, bool, 0644);
 module_param(rawmode, bool, 0644);
 module_param_named(coredump_mask, ath10k_coredump_mask, ulong, 0444);
+module_param(bug_on_fw_assert, bool, 0644);
 
 MODULE_PARM_DESC(debug_mask, "Debugging mask");
 MODULE_PARM_DESC(uart_print, "Uart target debugging");
@@ -47,6 +50,7 @@ MODULE_PARM_DESC(skip_otp, "Skip otp fai
 MODULE_PARM_DESC(cryptmode, "Crypto mode: 0-hardware, 1-software");
 MODULE_PARM_DESC(rawmode, "Use raw 802.11 frame datapath");
 MODULE_PARM_DESC(coredump_mask, "Bitfield of what to include in firmware crash file");
+MODULE_PARM_DESC(bug_on_fw_assert, "Bug on firmware assert");
 
 static const struct ath10k_hw_params ath10k_hw_params_list[] = {
 	{
--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -10,7 +10,7 @@
 #include <linux/vmalloc.h>
 #include <linux/crc32.h>
 #include <linux/firmware.h>
-
+#include <linux/of.h>
 #include "core.h"
 #include "debug.h"
 #include "hif.h"
@@ -2674,3 +2674,47 @@ void ath10k_dbg_dump(struct ath10k *ar,
 EXPORT_SYMBOL(ath10k_dbg_dump);
 
 #endif /* CPTCFG_ATH10K_DEBUG */
+
+void ath10k_dbg_bug_on_fw_assert(struct ath10k *ar, u8 *buf)
+{
+	struct device_node *node;
+	u32 crash_dump_addr, crash_dump_len;
+	u8 *addr, ret;
+
+	node = ar->dev->of_node;
+	if(!node) {
+		ath10k_err(ar, "Wifi node is not found\n");
+		return;
+	}
+
+	ret = of_property_read_u32(node, "qcom,wifi-dump", &crash_dump_addr);
+	if(ret){
+		ath10k_err(ar, "crash dump address is not assigned\n");
+		return;
+	}
+	ret = of_property_read_u32(node, "qcom,dump-len", &crash_dump_len);
+	if(ret){
+                ath10k_err(ar, "crash dump length is not assigned\n");
+                return;
+        }
+	if (!bug_on_fw_assert || ar->stats.fw_crash_counter == 0 ||
+            !ar->coredump.fw_crash_data->ramdump_buf)
+                return;
+	ath10k_info(ar, "Wifi-dump : addr %x and len %d\n",crash_dump_addr, crash_dump_len);
+
+	addr = (u8 *)ioremap(crash_dump_addr, crash_dump_len);
+	if (!addr) {
+		ath10k_err(ar, "Cannot access memory to dump WLAN Firmware data\n");
+		return;
+	}
+
+	/* copy the dump collected */
+	memcpy(addr, buf, crash_dump_len);
+
+	ath10k_err(ar, "Firmware DRAM dump collected successfully\n");
+	iounmap(addr);
+
+	ath10k_err(ar, "XXX: Asserting Host...\n");
+	BUG();
+}
+EXPORT_SYMBOL(ath10k_dbg_bug_on_fw_assert);
--- a/drivers/net/wireless/ath/ath10k/debug.h
+++ b/drivers/net/wireless/ath/ath10k/debug.h
@@ -72,11 +72,16 @@ struct ath10k_pktlog_hdr {
 #define ATH10K_FW_STATS_BUF_SIZE (1024 * 1024)
 
 extern unsigned int ath10k_debug_mask;
+extern bool bug_on_fw_assert;
 
 __printf(2, 3) void ath10k_info(struct ath10k *ar, const char *fmt, ...);
 __printf(2, 3) void ath10k_err(struct ath10k *ar, const char *fmt, ...);
 __printf(2, 3) void ath10k_warn(struct ath10k *ar, const char *fmt, ...);
 
+#define        ATTR_SRAM_MEM   BIT(0)
+#define        ATTR_REG_SPACE  BIT(1)
+
+
 void ath10k_debug_print_hwfw_info(struct ath10k *ar);
 void ath10k_debug_print_board_info(struct ath10k *ar);
 void ath10k_debug_print_boot_info(struct ath10k *ar);
@@ -238,6 +243,7 @@ void ath10k_sta_update_rx_tid_stats_ampd
 {
 }
 #endif /* CPTCFG_MAC80211_DEBUGFS */
+void ath10k_dbg_bug_on_fw_assert(struct ath10k *ar, u8* buf);
 
 #ifdef CPTCFG_ATH10K_DEBUG
 __printf(3, 4) void ath10k_dbg(struct ath10k *ar,
--- a/drivers/net/wireless/ath/ath10k/Makefile
+++ b/drivers/net/wireless/ath/ath10k/Makefile
@@ -21,7 +21,7 @@ ath10k_core-$(CPTCFG_ATH10K_TRACING) +=
 ath10k_core-$(CONFIG_THERMAL) += thermal.o
 ath10k_core-$(CPTCFG_MAC80211_DEBUGFS) += debugfs_sta.o
 ath10k_core-$(CONFIG_PM) += wow.o
-ath10k_core-$(CONFIG_DEV_COREDUMP) += coredump.o
+ath10k_core-y += coredump.o
 ath10k_core-$(CPTCFG_ATH10K_CE) += ce.o
 
 obj-$(CPTCFG_ATH10K_PCI) += ath10k_pci.o
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -1132,11 +1132,9 @@ struct ath10k {
 
 	u32 pktlog_filter;
 
-#ifdef CONFIG_DEV_COREDUMP
 	struct {
 		struct ath10k_fw_crash_data *fw_crash_data;
 	} coredump;
-#endif
 
 	struct {
 		/* protected by conf_mutex */
--- a/drivers/net/wireless/ath/ath10k/coredump.c
+++ b/drivers/net/wireless/ath/ath10k/coredump.c
@@ -5,12 +5,13 @@
  */
 
 #include "coredump.h"
-
+#ifdef CPTCFG_DEV_COREDUMP
 #include <linux/devcoredump.h>
+#endif
 #include <linux/kernel.h>
 #include <linux/types.h>
 #include <linux/utsname.h>
-
+#include <linux/of.h>
 #include "debug.h"
 #include "hw.h"
 
@@ -702,16 +703,6 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_REG,
-		.start = 0x98000,
-		.len = 0x50000,
-		.name = "IRAM",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
-	{
 		.type = ATH10K_MEM_REGION_TYPE_IOSRAM,
 		.start = 0xC0000,
 		.len = 0x40000,
@@ -722,10 +713,10 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x30000,
-		.len = 0x7000,
-		.name = "APB REG 1",
+		.type = ATH10K_MEM_REGION_TYPE_REG,
+		.start = 0x980000,
+		.len = 0x50000,
+		.name = "IRAM",
 		.section_table = {
 			.sections = NULL,
 			.size = 0,
@@ -733,9 +724,9 @@ static const struct ath10k_mem_region qc
 	},
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x3f000,
-		.len = 0x3000,
-		.name = "APB REG 2",
+		.start = 0x30000,
+		.len = 0x7000,
+		.name = "APB REG 1",
 		.section_table = {
 			.sections = NULL,
 			.size = 0,
@@ -771,6 +762,16 @@ static const struct ath10k_mem_region qc
 			.size = 0,
 		},
 	},
+	{
+		.type = ATH10K_MEM_REGION_TYPE_IOREG,
+		.start = 0x3f000,
+		.len = 0x3000,
+		.name = "APB REG 2",
+		.section_table = {
+			.sections = NULL,
+			.size = 0,
+		},
+	},
 };
 
 static const struct ath10k_mem_region qca9984_hw10_mem_regions[] = {
@@ -785,15 +786,15 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_REG,
-		.start = 0x98000,
-		.len = 0x50000,
-		.name = "IRAM",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
+                .type = ATH10K_MEM_REGION_TYPE_REG,
+                .start = 0x980000,
+                .len = 0x50000,
+                .name = "IRAM",
+                .section_table = {
+                        .sections = NULL,
+                        .size = 0,
+                },
+        },
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOSRAM,
 		.start = 0xC0000,
@@ -814,16 +815,16 @@ static const struct ath10k_mem_region qc
 			.size = 0,
 		},
 	},
-	{
-		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x3f000,
-		.len = 0x3000,
-		.name = "APB REG 2",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
+	 {
+                .type = ATH10K_MEM_REGION_TYPE_IOREG,
+                .start = 0x3f000,
+                .len = 0x3000,
+                .name = "APB REG 2",
+                .section_table = {
+                        .sections = NULL,
+                        .size = 0,
+                },
+        },
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
 		.start = 0x43000,
@@ -879,6 +880,16 @@ static const struct ath10k_mem_region qc
 			.size = 0,
 		},
 	},
+	 {
+                .type = ATH10K_MEM_REGION_TYPE_REG,
+                .start = 0x980000,
+                .len = 0x50000,
+                .name = "IRAM",
+                .section_table = {
+                        .sections = NULL,
+                        .size = 0,
+                },
+        },
 	{
 		.type = ATH10K_MEM_REGION_TYPE_REG,
 		.start = 0xC0000,
@@ -890,16 +901,6 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_REG,
-		.start = 0x98000,
-		.len = 0x50000,
-		.name = "IRAM",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
-	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
 		.start = 0x30000,
 		.len = 0x7000,
@@ -910,15 +911,15 @@ static const struct ath10k_mem_region qc
 		},
 	},
 	{
-		.type = ATH10K_MEM_REGION_TYPE_IOREG,
-		.start = 0x3f000,
-		.len = 0x3000,
-		.name = "APB REG 2",
-		.section_table = {
-			.sections = NULL,
-			.size = 0,
-		},
-	},
+                .type = ATH10K_MEM_REGION_TYPE_IOREG,
+                .start = 0x3f000,
+                .len = 0x3000,
+                .name = "APB REG 2",
+                .section_table = {
+                        .sections = NULL,
+                        .size = 0,
+                },
+        },
 	{
 		.type = ATH10K_MEM_REGION_TYPE_IOREG,
 		.start = 0x43000,
@@ -1109,7 +1110,9 @@ struct ath10k_fw_crash_data *ath10k_core
 		return NULL;
 
 	guid_gen(&crash_data->guid);
+#ifdef CPTCFG_DEV_COREDUMP
 	ktime_get_real_ts64(&crash_data->timestamp);
+#endif
 
 	return crash_data;
 }
@@ -1121,12 +1124,28 @@ static struct ath10k_dump_file_data *ath
 	struct ath10k_ce_crash_hdr *ce_hdr;
 	struct ath10k_dump_file_data *dump_data;
 	struct ath10k_tlv_dump_data *dump_tlv;
+	struct device_node *node;
 	size_t hdr_len = sizeof(*dump_data);
 	size_t len, sofar = 0;
+	u32 crash_dump_len;
+	u8 ret;
 	unsigned char *buf;
 
+	node = ar->dev->of_node;
+	if(!node){
+		ath10k_err(ar, "Wifi node not detected\n");
+		return NULL;
+	}
+
 	len = hdr_len;
 
+	ret = of_property_read_u32(node, "qcom,dump-len", &crash_dump_len);
+	if(ret){
+		ath10k_err(ar, "crash dump length not valid\n");
+		return  NULL;
+	}
+	ath10k_info(ar, "crash-dump-len %d\n",crash_dump_len);
+
 	if (test_bit(ATH10K_FW_CRASH_DUMP_REGISTERS, &ath10k_coredump_mask))
 		len += sizeof(*dump_tlv) + sizeof(crash_data->registers);
 
@@ -1137,8 +1156,8 @@ static struct ath10k_dump_file_data *ath
 	if (test_bit(ATH10K_FW_CRASH_DUMP_RAM_DATA, &ath10k_coredump_mask))
 		len += sizeof(*dump_tlv) + crash_data->ramdump_buf_len;
 
+	len = crash_dump_len;
 	sofar += hdr_len;
-
 	/* This is going to get big when we start dumping FW RAM and such,
 	 * so go ahead and use vmalloc.
 	 */
@@ -1147,7 +1166,6 @@ static struct ath10k_dump_file_data *ath
 		return NULL;
 
 	spin_lock_bh(&ar->data_lock);
-
 	dump_data = (struct ath10k_dump_file_data *)(buf);
 	strlcpy(dump_data->df_magic, "ATH10K-FW-DUMP",
 		sizeof(dump_data->df_magic));
@@ -1232,7 +1250,10 @@ int ath10k_coredump_submit(struct ath10k
 		return -ENODATA;
 	}
 
+	ath10k_dbg_bug_on_fw_assert(ar, (u8 *)dump);
+#ifdef CPTCFG_DEV_COREDUMP
 	dev_coredumpv(ar->dev, dump, le32_to_cpu(dump->len), GFP_KERNEL);
+#endif
 
 	return 0;
 }
@@ -1257,9 +1278,11 @@ int ath10k_coredump_register(struct ath1
 	if (test_bit(ATH10K_FW_CRASH_DUMP_RAM_DATA, &ath10k_coredump_mask)) {
 		crash_data->ramdump_buf_len = ath10k_coredump_get_ramdump_size(ar);
 
-		crash_data->ramdump_buf = vzalloc(crash_data->ramdump_buf_len);
-		if (!crash_data->ramdump_buf)
-			return -ENOMEM;
+		if (crash_data->ramdump_buf_len) {
+			crash_data->ramdump_buf = vzalloc(crash_data->ramdump_buf_len);
+			if (!crash_data->ramdump_buf)
+				return -ENOMEM;
+		}
 	}
 
 	return 0;
--- a/drivers/net/wireless/ath/ath10k/coredump.h
+++ b/drivers/net/wireless/ath/ath10k/coredump.h
@@ -165,8 +165,6 @@ struct ath10k_hw_mem_layout {
 /* FIXME: where to put this? */
 extern unsigned long ath10k_coredump_mask;
 
-#ifdef CONFIG_DEV_COREDUMP
-
 int ath10k_coredump_submit(struct ath10k *ar);
 struct ath10k_fw_crash_data *ath10k_coredump_new(struct ath10k *ar);
 int ath10k_coredump_create(struct ath10k *ar);
@@ -176,42 +174,4 @@ void ath10k_coredump_destroy(struct ath1
 
 const struct ath10k_hw_mem_layout *ath10k_coredump_get_mem_layout(struct ath10k *ar);
 
-#else /* CONFIG_DEV_COREDUMP */
-
-static inline int ath10k_coredump_submit(struct ath10k *ar)
-{
-	return 0;
-}
-
-static inline struct ath10k_fw_crash_data *ath10k_coredump_new(struct ath10k *ar)
-{
-	return NULL;
-}
-
-static inline int ath10k_coredump_create(struct ath10k *ar)
-{
-	return 0;
-}
-
-static inline int ath10k_coredump_register(struct ath10k *ar)
-{
-	return 0;
-}
-
-static inline void ath10k_coredump_unregister(struct ath10k *ar)
-{
-}
-
-static inline void ath10k_coredump_destroy(struct ath10k *ar)
-{
-}
-
-static inline const struct ath10k_hw_mem_layout *
-ath10k_coredump_get_mem_layout(struct ath10k *ar)
-{
-	return NULL;
-}
-
-#endif /* CONFIG_DEV_COREDUMP */
-
 #endif /* _COREDUMP_H_ */
