diff --git a/drivers/net/wireless/ath/ath10k/mac.c b/drivers/net/wireless/ath/ath10k/mac.c
index d04865a..6fb5862 100644
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -6440,11 +6433,21 @@ static int ath10k_sta_state(struct ieee80211_hw *hw,
 			goto exit;
 		}
 
+		if (ar->debug.enable_extd_tx_stats) {
+			arsta->tx_stats = kzalloc(sizeof(*arsta->tx_stats),
+							 GFP_KERNEL);
+			if (!arsta->tx_stats) {
+				ret = -ENOMEM;
+				goto exit;
+			}
+		}
+
 		ret = ath10k_peer_create(ar, vif, sta, arvif->vdev_id,
 					 sta->addr, peer_type);
 		if (ret) {
 			ath10k_warn(ar, "failed to add peer %pM for vdev %d when adding a new sta: %i\n",
 				    sta->addr, arvif->vdev_id, ret);
+			kfree(arsta->tx_stats);
 			ath10k_mac_dec_num_stations(arvif, sta);
 			for (i = 0; i < ARRAY_SIZE(sta->txq); i++)
 				ath10k_mac_txq_unref(ar, sta->txq[i]);
@@ -6459,6 +6462,7 @@ static int ath10k_sta_state(struct ieee80211_hw *hw,
 			ath10k_warn(ar, "failed to lookup peer %pM on vdev %i\n",
 				    vif->addr, arvif->vdev_id);
 			spin_unlock_bh(&ar->data_lock);
+			kfree(arsta->tx_stats);
 			ath10k_peer_delete(ar, arvif->vdev_id, sta->addr);
 			ath10k_mac_dec_num_stations(arvif, sta);
 			ret = -ENOENT;
@@ -6478,6 +6482,7 @@ static int ath10k_sta_state(struct ieee80211_hw *hw,
 		if (ret) {
 			ath10k_warn(ar, "failed to update fw tdls state on vdev %i: %i\n",
 				    arvif->vdev_id, ret);
+			kfree(arsta->tx_stats);
 			ath10k_peer_delete(ar, arvif->vdev_id,
 					   sta->addr);
 			ath10k_mac_dec_num_stations(arvif, sta);
@@ -6490,6 +6495,7 @@ static int ath10k_sta_state(struct ieee80211_hw *hw,
 			ath10k_warn(ar,
 				    "failed to update tdls peer %pM for vdev %d when adding a new sta: %i\n",
 				    sta->addr, arvif->vdev_id, ret);
+			kfree(arsta->tx_stats);
 			ath10k_peer_delete(ar, arvif->vdev_id, sta->addr);
 			ath10k_mac_dec_num_stations(arvif, sta);
 
