--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -6407,6 +6407,9 @@ static int ath10k_sta_state(struct ieee8
 			ath10k_warn(ar, "failed to add peer %pM for vdev %d when adding a new sta: %i\n",
 				    sta->addr, arvif->vdev_id, ret);
 			ath10k_mac_dec_num_stations(arvif, sta);
+			for (i = 0; i < ARRAY_SIZE(sta->txq); i++)
+				ath10k_mac_txq_unref(ar, sta->txq[i]);
+
 			goto exit;
 		}
 
