--- a/drivers/net/wireless/ath/ath11k/core.c
+++ b/drivers/net/wireless/ath/ath11k/core.c
@@ -720,6 +720,7 @@ void ath11k_core_halt(struct ath11k *ar)
 	lockdep_assert_held(&ar->conf_mutex);
 
 	ar->num_created_vdevs = 0;
+	ar->allocated_vdev_map = 0;
 
 	ath11k_mac_scan_finish(ar);
 	ath11k_mac_peer_cleanup_all(ar);
--- a/drivers/net/wireless/ath/ath11k/core.h
+++ b/drivers/net/wireless/ath/ath11k/core.h
@@ -553,6 +553,7 @@ struct ath11k {
 	int max_num_peers;
 	u32 num_started_vdevs;
 	u32 num_created_vdevs;
+	unsigned long long allocated_vdev_map;
 
 	struct idr txmgmt_idr;
 	/* protects txmgmt_idr data */
--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -371,8 +371,10 @@ struct ath11k_vif *ath11k_mac_get_arvif(
 						   flags,
 						   ath11k_mac_get_arvif_iter,
 						   &arvif_iter);
-	if (!arvif_iter.arvif)
+	if (!arvif_iter.arvif) {
+		ath11k_warn(ar->ab, "No VIF found for vdev %d\n", vdev_id);
 		return NULL;
+	}
 
 	return arvif_iter.arvif;
 }
@@ -405,11 +407,7 @@ struct ath11k *ath11k_mac_get_ar_by_vdev
 	for (i = 0; i < ab->num_radios; i++) {
 		pdev = rcu_dereference(ab->pdevs_active[i]);
 		if (pdev && pdev->ar) {
-			arvif = ath11k_mac_get_arvif(pdev->ar, vdev_id);
-			if (arvif)
-				return arvif->ar;
-			/* vdev start resp for monitor interface */
-			if (pdev->ar->monitor_vdev_id == vdev_id)
+			if (pdev->ar->allocated_vdev_map & (1LL << vdev_id))
 				return pdev->ar;
 		}
 	}
@@ -435,31 +433,6 @@ struct ath11k *ath11k_mac_get_ar_by_pdev
 	return NULL;
 }
 
-struct ath11k *ath11k_mac_get_ar_from_arvif_list(struct ath11k_base *ab, u32 vdev_id)
-{
-	int i;
-	struct ath11k_pdev *pdev;
-	struct ath11k_vif *arvif;
-	struct ath11k *ar;
-
-	for (i = 0; i < ab->num_radios; i++) {
-		pdev = rcu_dereference(ab->pdevs_active[i]);
-		if (pdev && pdev->ar) {
-			ar = pdev->ar;
-
-			spin_lock_bh(&ar->data_lock);
-			list_for_each_entry(arvif, &ar->arvifs, list) {
-				if (arvif->vdev_id == vdev_id) {
-					spin_unlock_bh(&ar->data_lock);
-					return arvif->ar;
-				}
-			}
-			spin_unlock_bh(&ar->data_lock);
-		}
-	}
-	return NULL;
-}
-
 static void ath11k_pdev_caps_update(struct ath11k *ar)
 {
 	struct ath11k_base *ab = ar->ab;
@@ -3972,6 +3945,7 @@ static int ath11k_mac_op_start(struct ie
 	ar->num_started_vdevs = 0;
 	ar->num_created_vdevs = 0;
 	ar->num_peers = 0;
+	ar->allocated_vdev_map = 0;
 
 	/* Configure monitor status ring with default rx_filter to get rx status
 	 * such as rssi, rx_duration.
@@ -4214,7 +4188,9 @@ static int ath11k_mac_op_add_interface(s
 	}
 	ar->num_created_vdevs++;
 
+	ar->allocated_vdev_map |= 1LL << arvif->vdev_id;
 	ab->free_vdev_map &= ~(1LL << arvif->vdev_id);
+
 	spin_lock_bh(&ar->data_lock);
 	list_add(&arvif->list, &ar->arvifs);
 	spin_unlock_bh(&ar->data_lock);
@@ -4342,6 +4318,7 @@ err_peer_del:
 err_vdev_del:
 	ath11k_wmi_vdev_delete(ar, arvif->vdev_id);
 	ar->num_created_vdevs--;
+	ar->allocated_vdev_map &= ~(1LL << arvif->vdev_id);
 	ab->free_vdev_map |= 1LL << arvif->vdev_id;
 	spin_lock_bh(&ar->data_lock);
 	list_del(&arvif->list);
@@ -4398,6 +4375,7 @@ static void ath11k_mac_op_remove_interfa
 	if (time_left == 0)
 		ath11k_warn(ab, "Timeout in receiving vdev delete response\n");
 	else {
+		ar->allocated_vdev_map &= ~(1LL << arvif->vdev_id);
 		ab->free_vdev_map |= 1LL << (arvif->vdev_id);
 		ar->num_created_vdevs--;
 	}
@@ -5799,6 +5777,7 @@ static int ath11k_mac_monitor_vdev_creat
 	if (ret)
 		return ret;
 
+	ar->allocated_vdev_map |= 1LL << ar->monitor_vdev_id;
 	ar->ab->free_vdev_map &= ~(1LL << ar->monitor_vdev_id);
 	ath11k_dbg(ar->ab, ATH11K_DBG_MAC, "mac monitor vdev %d created\n",
 		   ar->monitor_vdev_id);
@@ -5827,6 +5806,7 @@ static int ath11k_mac_monitor_vdev_delet
 	if (time_left == 0) {
 		ath11k_warn(ar->ab, "Timeout in receiving vdev delete response\n");
 	} else {
+		ar->allocated_vdev_map &= ~(1LL << ar->monitor_vdev_id);
 		ar->ab->free_vdev_map |= 1LL << (ar->monitor_vdev_id);
 		ar->num_created_vdevs--;
 		ar->monitor_vdev_id = -1;
--- a/drivers/net/wireless/ath/ath11k/mac.h
+++ b/drivers/net/wireless/ath/ath11k/mac.h
@@ -140,7 +140,6 @@ int ath11k_mac_coex_config(struct ath11k
 			   int coex, u32 wlan_prio_mask, u8 wlan_weight);
 struct ath11k *ath11k_mac_get_ar_by_vdev_id(struct ath11k_base *ab, u32 vdev_id);
 struct ath11k *ath11k_mac_get_ar_by_pdev_id(struct ath11k_base *ab, u32 pdev_id);
-struct ath11k *ath11k_mac_get_ar_from_arvif_list(struct ath11k_base *ab, u32 vdev_id);
 void ath11k_mac_drain_tx(struct ath11k *ar);
 void ath11k_mac_peer_cleanup_all(struct ath11k *ar);
 int ath11k_mac_tx_mgmt_pending_free(int buf_id, void *skb, void *ctx);
--- a/drivers/net/wireless/ath/ath11k/wmi.c
+++ b/drivers/net/wireless/ath/ath11k/wmi.c
@@ -5757,7 +5757,7 @@ static void ath11k_peer_delete_resp_even
 	}
 
 	rcu_read_lock();
-	ar = ath11k_mac_get_ar_from_arvif_list(ab, peer_del_resp.vdev_id);
+	ar = ath11k_mac_get_ar_by_vdev_id(ab, peer_del_resp.vdev_id);
 	if (!ar) {
 		ath11k_warn(ab, "invalid vdev id in peer delete resp ev %d",
 			    peer_del_resp.vdev_id);
@@ -5785,7 +5785,7 @@ static void ath11k_vdev_delete_resp_even
 	}
 
 	rcu_read_lock();
-	ar = ath11k_mac_get_ar_from_arvif_list(ab, vdev_id);
+	ar = ath11k_mac_get_ar_by_vdev_id(ab, vdev_id);
 	if (!ar) {
 		ath11k_warn(ab, "invalid vdev id in vdev delete resp ev %d",
 			    vdev_id);
@@ -5876,7 +5876,7 @@ static void ath11k_vdev_stopped_event(st
 	}
 
 	rcu_read_lock();
-	ar = ath11k_mac_get_ar_from_arvif_list(ab, vdev_id);
+	ar = ath11k_mac_get_ar_by_vdev_id(ab, vdev_id);
 	if (!ar) {
 		ath11k_warn(ab, "invalid vdev id in vdev stopped ev %d",
 			    vdev_id);
