--- a/backport-include/linux/mm.h
+++ b/backport-include/linux/mm.h
@@ -3,7 +3,6 @@
 #include_next <linux/mm.h>
 #include <linux/page_ref.h>
 #include <linux/sched.h>
-#include <linux/overflow.h>
 #include <linux/vmalloc.h>
 #include <linux/slab.h>
 
@@ -105,12 +104,10 @@ static inline void *kvmalloc(size_t size, gfp_t flags)
 #define kvmalloc_array LINUX_BACKPORT(kvmalloc_array)
 static inline void *kvmalloc_array(size_t n, size_t size, gfp_t flags)
 {
-	size_t bytes;
-
-	if (unlikely(check_mul_overflow(n, size, &bytes)))
+	if (size != 0 && n > SIZE_MAX / size)
 		return NULL;
 
-	return kvmalloc(bytes, flags);
+	return kvmalloc(n * size, flags);
 }
 
 #define kvzalloc LINUX_BACKPORT(kvzalloc)
@@ -47,4 +47,8 @@ static inline void *kvzalloc(size_t size, gfp_t flags)
 }
 #endif
 
+#if LINUX_VERSION_IS_GEQ(3,14,77)
+#define struct_size(p, member, n) (n * sizeof(*(p)->member) +  sizeof(*(p)))
+#endif
+
 #endif /* __BACKPORT_MM_H */
--- a/backport-include/linux/netdevice.h
+++ b/backport-include/linux/netdevice.h
@@ -333,7 +333,7 @@ static inline void netif_trans_update(struct net_device *dev)
 
 #if LINUX_VERSION_IS_LESS(4,11,9)
 #define netdev_set_priv_destructor(_dev, _destructor) \
-	(_dev)->destructor = __ ## _destructor
+	(_dev)->destructor = _destructor
 #define netdev_set_def_destructor(_dev) \
 	(_dev)->destructor = free_netdev
 #else
--- a/backport-include/linux/timekeeping.h
+++ b/backport-include/linux/timekeeping.h
@@ -82,10 +82,10 @@ static inline struct timespec64 timespec_to_timespec64(const struct timespec ts)
 static inline void ktime_get_raw_ts64(struct timespec64 *ts)
 {
 #if LINUX_VERSION_IS_LESS(3,19,0)
-	struct timespec64 ts64;
+	struct timespec _ts;
 
-	getrawmonotonic(&ts64);
-	*ts = timespec_to_timespec64(ts64);
+	getrawmonotonic(&_ts);
+	*ts = timespec_to_timespec64(_ts);
 #else
 	return getrawmonotonic64(ts);
 #endif /* < 3.19 */
--- a/drivers/net/wireless/ath/ath10k/ahb.c
+++ b/drivers/net/wireless/ath/ath10k/ahb.c
@@ -191,40 +191,35 @@ static int ath10k_ahb_rst_ctrl_init(struct ath10k *ar)
 
 	dev = &ar_ahb->pdev->dev;
 
-	ar_ahb->core_cold_rst = devm_reset_control_get_exclusive(dev,
-								 "wifi_core_cold");
+	ar_ahb->core_cold_rst = devm_reset_control_get(dev, "wifi_core_cold");
 	if (IS_ERR(ar_ahb->core_cold_rst)) {
 		ath10k_err(ar, "failed to get core cold rst ctrl: %ld\n",
 			   PTR_ERR(ar_ahb->core_cold_rst));
 		return PTR_ERR(ar_ahb->core_cold_rst);
 	}
 
-	ar_ahb->radio_cold_rst = devm_reset_control_get_exclusive(dev,
-								  "wifi_radio_cold");
+	ar_ahb->radio_cold_rst = devm_reset_control_get(dev, "wifi_radio_cold");
 	if (IS_ERR(ar_ahb->radio_cold_rst)) {
 		ath10k_err(ar, "failed to get radio cold rst ctrl: %ld\n",
 			   PTR_ERR(ar_ahb->radio_cold_rst));
 		return PTR_ERR(ar_ahb->radio_cold_rst);
 	}
 
-	ar_ahb->radio_warm_rst = devm_reset_control_get_exclusive(dev,
-								  "wifi_radio_warm");
+	ar_ahb->radio_warm_rst = devm_reset_control_get(dev, "wifi_radio_warm");
 	if (IS_ERR(ar_ahb->radio_warm_rst)) {
 		ath10k_err(ar, "failed to get radio warm rst ctrl: %ld\n",
 			   PTR_ERR(ar_ahb->radio_warm_rst));
 		return PTR_ERR(ar_ahb->radio_warm_rst);
 	}
 
-	ar_ahb->radio_srif_rst = devm_reset_control_get_exclusive(dev,
-								  "wifi_radio_srif");
+	ar_ahb->radio_srif_rst = devm_reset_control_get(dev, "wifi_radio_srif");
 	if (IS_ERR(ar_ahb->radio_srif_rst)) {
 		ath10k_err(ar, "failed to get radio srif rst ctrl: %ld\n",
 			   PTR_ERR(ar_ahb->radio_srif_rst));
 		return PTR_ERR(ar_ahb->radio_srif_rst);
 	}
 
-	ar_ahb->cpu_init_rst = devm_reset_control_get_exclusive(dev,
-								"wifi_cpu_init");
+	ar_ahb->cpu_init_rst = devm_reset_control_get(dev, "wifi_cpu_init");
 	if (IS_ERR(ar_ahb->cpu_init_rst)) {
 		ath10k_err(ar, "failed to get cpu init rst ctrl: %ld\n",
 			   PTR_ERR(ar_ahb->cpu_init_rst));
--- a/drivers/net/wireless/ath/ath10k/ce.c
+++ b/drivers/net/wireless/ath/ath10k/ce.c
@@ -17,6 +17,7 @@
  */
 
 #include "hif.h"
+#include "pci.h"
 #include "ce.h"
 #include "debug.h"
 
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -8392,6 +8392,8 @@ struct ath10k_vif *ath10k_get_arvif(struct ath10k *ar, u32 vdev_id)
 	return arvif_iter.arvif;
 }
 
+#ifdef CONFIG_ACPI
+
 #define WRD_METHOD "WRDD"
 #define WRDD_WIFI  (0x07)
 
@@ -8478,6 +8480,13 @@ static int ath10k_mac_get_wrdd_regulatory(struct ath10k *ar, u16 *rd)
 	*rd |= COUNTRY_ERD_FLAG;
 	return 0;
 }
+#else
+static int ath10k_mac_get_wrdd_regulatory(struct ath10k *ar, u16 *rd)
+{
+	/* Use EEPROM regulatory information when ACPI is not being used */
+	return -1;
+}
+#endif /* CONFIG_ACPI */
 
 static int ath10k_mac_init_rd(struct ath10k *ar)
 {
--- /dev/null
+++ b/include/linux/build_bug.h
@@ -0,0 +1,10 @@
+#ifndef __BP_BUILD_BUG_H
+#define __BP_BUILD_BUG_H
+
+#if LINUX_VERSION_IS_GEQ(4,13,0)
+#include_next <linux/build_bug.h>
+#else /* LINUX_VERSION_IS_GEQ(4,13,0) */
+#include <linux/bug.h>
+#endif /* LINUX_VERSION_IS_GEQ(4,13,0) */
+
+#endif /* __BP_BUILD_BUG_H */
--- a/net/mac80211/debugfs_sta.c
+++ b/net/mac80211/debugfs_sta.c
@@ -1094,14 +1094,14 @@ void ieee80211_rx_h_sta_stats(struct sta_info *sta, struct sk_buff *skb)
 		return;
 
 	/* Not support 5Mhz and 10Mhz currently  */
-	if (status->flag & (RX_FLAG_5MHZ | RX_FLAG_10MHZ))
+	if (status->bw == RATE_INFO_BW_5 || status->bw == RATE_INFO_BW_10)
 		goto out;
 
-	if (status->vht_flag & RX_VHT_FLAG_160MHZ)
+	if (status->bw == RATE_INFO_BW_160)
 		bw_idx = 3;
-	else if (status->vht_flag & RX_VHT_FLAG_80MHZ)
+	else if (status->bw == RATE_INFO_BW_80)
 		bw_idx = 2;
-	else if (status->flag & RX_FLAG_40MHZ)
+	else if (status->bw == RATE_INFO_BW_40)
 		bw_idx = 1;
 	else
 		bw_idx = 0;
@@ -1109,7 +1109,7 @@ void ieee80211_rx_h_sta_stats(struct sta_info *sta, struct sk_buff *skb)
 	i = abs(status->signal);
 	sta->sdata->vif.rx_signal_pkt[i]++;
 
-	if (status->flag & RX_FLAG_HT) {
+	if (status->encoding & RX_ENC_HT) {
 		mcs_idx = status->rate_idx;
 		nss_idx = mcs_idx >> 3;
 
@@ -1123,9 +1123,9 @@ void ieee80211_rx_h_sta_stats(struct sta_info *sta, struct sk_buff *skb)
 		sta->rx_nss_byte[nss_idx] += pkt_len;
 		/* To fit into rate table for HT packets */
 		mcs_idx = mcs_idx % 8;
-	} else if (status->flag & RX_FLAG_VHT) {
+	} else if (status->encoding == RX_ENC_VHT) {
 		mcs_idx = status->rate_idx;
-		nss_idx = status->vht_nss - 1;
+		nss_idx = status->nss - 1;
 
 		if (nss_idx > IEEE80211_NSS_NUM - 1 ||
 		    mcs_idx > (IEEE80211_VHT_MCS_NUM - 1))
@@ -1137,13 +1137,13 @@ void ieee80211_rx_h_sta_stats(struct sta_info *sta, struct sk_buff *skb)
 		sta->rx_nss_byte[nss_idx] += pkt_len;
 	}
 
-	gi_idx = (status->flag & RX_FLAG_SHORT_GI) ? 1 : 0;
+	gi_idx = (status->enc_flags & RX_ENC_FLAG_SHORT_GI) ? 1 : 0;
 	sta->rx_gi_pkt[gi_idx]++;
 	sta->rx_gi_byte[gi_idx] += pkt_len;
 	sta->rx_bw_pkt[bw_idx]++;
 	sta->rx_bw_byte[bw_idx] += pkt_len;
 
-	if (status->flag & (RX_FLAG_HT | RX_FLAG_VHT)) {
+	if (status->encoding == RX_ENC_HT || status->encoding == RX_ENC_VHT) {
 		/* Update Rate table for HT and VHT packets */
 		i = mcs_idx * 8 + 8 * 10 * nss_idx;
 		i += bw_idx * 2 + gi_idx;
@@ -1152,14 +1152,13 @@ void ieee80211_rx_h_sta_stats(struct sta_info *sta, struct sk_buff *skb)
 	} else {
 		struct ieee80211_supported_band *sband;
 		struct ieee80211_sub_if_data *sdata = sta->sdata;
-		enum nl80211_band band = ieee80211_get_sdata_band(sdata);
 		int shift = ieee80211_vif_get_shift(&sta->sdata->vif);
 		u16 brate, legacy_rate;
 
 		if (status->rate_idx > IEEE80211_LEGACY_RATE_NUM - 1)
 			goto out;
 
-		sband = local->hw.wiphy->bands[band];
+		sband = ieee80211_get_sband(sdata);
 		brate = sband->bitrates[status->rate_idx].bitrate;
 		legacy_rate = DIV_ROUND_UP(brate, 1 << shift);
 		i = legacy_rate_to_index(legacy_rate);
--- a/drivers/net/wireless/ath/wil6210/pcie_bus.c
+++ b/drivers/net/wireless/ath/wil6210/pcie_bus.c
@@ -215,7 +215,11 @@ static int wil_if_pcie_enable(struct wil6210_priv *wil)
 	}
 
 	if (n_msi == 3 &&
+#if LINUX_VERSION_IS_GEQ(4,8,0)
 	    pci_alloc_irq_vectors(pdev, n_msi, n_msi, PCI_IRQ_MSI) < n_msi) {
+#else
+	    true) {
+#endif
 		wil_err(wil, "3 MSI mode failed, try 1 MSI\n");
 		n_msi = 1;
 	}
@@ -250,7 +254,11 @@ static int wil_if_pcie_enable(struct wil6210_priv *wil)
 	wil6210_fini_irq(wil, pdev->irq);
  release_vectors:
 	/* safe to call if no allocation */
+#if LINUX_VERSION_IS_GEQ(4,8,0)
 	pci_free_irq_vectors(pdev);
+#else
+	pci_disable_msi(pdev);
+#endif
  stop_master:
 	pci_clear_master(pdev);
 	return rc;
--- a/compat/Makefile
+++ b/compat/Makefile
@@ -36,7 +36,6 @@ compat-$(CPTCFG_KERNEL_4_6) += backport-4.6.o
 compat-$(CPTCFG_KERNEL_4_7) += backport-4.7.o
 compat-$(CPTCFG_KERNEL_4_8) += backport-4.8.o
 compat-$(CPTCFG_KERNEL_4_10) += backport-4.10.o
-compat-$(CPTCFG_KERNEL_4_18) += backport-4.18.o
 compat-$(CPTCFG_KERNEL_4_20) += backport-4.20.o
 
 compat-$(CPTCFG_KERNEL_4_20) += backport-genetlink.o
--- a/include/linux/backport-rhashtable.h
+++ b/include/linux/backport-rhashtable.h
@@ -1161,4 +1161,11 @@ static inline void rhltable_destroy(struct rhltable *hlt)
 	return rhltable_free_and_destroy(hlt, NULL, NULL);
 }
 
+/* Obsolete function, do not use in new code. */
+static inline int rhashtable_walk_init(struct rhashtable *ht,
+                                       struct rhashtable_iter *iter, gfp_t gfp)
+{
+        rhashtable_walk_enter(ht, iter);
+        return 0;
+}
 #endif /* _LINUX_RHASHTABLE_H */
--- a/backport-include/net/netlink.h
+++ b/backport-include/net/netlink.h
@@ -59,10 +59,14 @@ struct backport_nla_policy {
 #define NLA_POLICY_ETH_ADDR		NLA_POLICY_EXACT_LEN(ETH_ALEN)
 #define NLA_POLICY_ETH_ADDR_COMPAT	NLA_POLICY_EXACT_LEN_WARN(ETH_ALEN)
 
-#define NLA_POLICY_NESTED(maxattr, policy) \
+#define _NLA_POLICY_NESTED(maxattr, policy) \
 	{ .type = NLA_NESTED, .validation_data = policy, .len = maxattr }
-#define NLA_POLICY_NESTED_ARRAY(maxattr, policy) \
+#define _NLA_POLICY_NESTED_ARRAY(maxattr, policy) \
 	{ .type = NLA_NESTED_ARRAY, .validation_data = policy, .len = maxattr }
+#define NLA_POLICY_NESTED(policy) \
+	_NLA_POLICY_NESTED(ARRAY_SIZE(policy) - 1, policy)
+#define NLA_POLICY_NESTED_ARRAY(policy) \
+	_NLA_POLICY_NESTED_ARRAY(ARRAY_SIZE(policy) - 1, policy)
 
 #define __NLA_ENSURE(condition) (sizeof(char[1 - 2*!(condition)]) - 1)
 #define NLA_ENSURE_INT_TYPE(tp)				\
