--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -338,6 +338,7 @@ enum ath10k_beacon_state {
 
 struct ath10k_txq {
 	struct list_head pending;
+	unsigned long num_fw_queued;
 };
 
 struct ath10k_vif {
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -3483,6 +3483,8 @@ static void ath10k_mac_txq_push(struct i
 			break;
 		}
 
+		artxq->num_fw_queued++;
+
 		spin_unlock_bh(&ar->htt.tx_lock);
 
 		skb = ieee80211_tx_dequeue(hw, txq);
@@ -3491,6 +3493,7 @@ static void ath10k_mac_txq_push(struct i
 			if (!list_empty(&artxq->pending))
 				list_del_init(&artxq->pending);
 			ath10k_htt_tx_dec_pending(htt, is_mgmt);
+			artxq->num_fw_queued--;
 			spin_unlock_bh(&ar->htt.tx_lock);
 			break;
 		}
@@ -3506,6 +3509,7 @@ static void ath10k_mac_txq_push(struct i
 				    ret);
 			spin_lock_bh(&ar->htt.tx_lock);
 			ath10k_htt_tx_dec_pending(htt, is_mgmt);
+			artxq->num_fw_queued--;
 			spin_unlock_bh(&ar->htt.tx_lock);
 			continue;
 		}
@@ -3874,6 +3878,7 @@ static void ath10k_mac_op_tx(struct ieee
 {
 	struct ath10k *ar = hw->priv;
 	struct ath10k_htt *htt = &ar->htt;
+	struct ath10k_txq *artxq;
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	struct ieee80211_vif *vif = info->control.vif;
 	struct ieee80211_sta *sta = control->sta;
@@ -3893,6 +3898,8 @@ static void ath10k_mac_op_tx(struct ieee
 	else
 		txq = NULL;
 
+	artxq = (void *)txq->drv_priv;
+
 	ath10k_mac_tx_h_fill_cb(ar, txq, skb);
 
 	txmode = ath10k_mac_tx_h_get_txmode(ar, vif, sta, skb);
@@ -3915,6 +3922,9 @@ static void ath10k_mac_op_tx(struct ieee
 			return;
 		}
 
+		if (txq)
+			artxq->num_fw_queued++;
+
 		spin_unlock_bh(&ar->htt.tx_lock);
 	}
 
@@ -3924,6 +3934,8 @@ static void ath10k_mac_op_tx(struct ieee
 		if (is_htt) {
 			spin_lock_bh(&ar->htt.tx_lock);
 			ath10k_htt_tx_dec_pending(htt, is_mgmt);
+			if (txq)
+				artxq->num_fw_queued--;
 			spin_unlock_bh(&ar->htt.tx_lock);
 		}
 		return;
--- a/drivers/net/wireless/ath/ath10k/txrx.c
+++ b/drivers/net/wireless/ath/ath10k/txrx.c
@@ -53,6 +53,7 @@ void ath10k_txrx_tx_unref(struct ath10k_
 			  const struct htt_tx_done *tx_done)
 {
 	struct ath10k *ar = htt->ar;
+	struct ath10k_txq *artxq;
 	struct device *dev = ar->dev;
 	struct ieee80211_tx_info *info;
 	struct ath10k_skb_cb *skb_cb;
@@ -81,6 +82,12 @@ void ath10k_txrx_tx_unref(struct ath10k_
 
 	skb_cb = ATH10K_SKB_CB(msdu);
 
+	if (skb_cb->txq) {
+		artxq = (void *)skb_cb->txq->drv_priv;
+		if (!WARN_ON_ONCE(artxq->num_fw_queued == 0))
+			artxq->num_fw_queued--;
+	}
+
 	if (unlikely(skb_cb->flags & ATH10K_SKB_F_MGMT) &&
 	    ar->hw_params.max_probe_resp_desc_thres)
 		limit_mgmt_desc = true;
