--- a/drivers/net/wireless/ath/ath10k/debug_htt_stats.c
+++ b/drivers/net/wireless/ath/ath10k/debug_htt_stats.c
@@ -719,6 +719,14 @@ static void ath10k_htt_10_4_tx_rate_info_stats_fill(struct htt_10_4_tx_rate_info
 	for (i = 1; i <= ARRAY_SIZE(stats->nss); i++)
 		len += scnprintf(buf + len, buf_len - len, "%dx%d\t%-7d ",
 				 i, i, le32_to_cpu(stats->nss[i - 1]));
+
+	len += scnprintf(buf + len, buf_len - len, "\n\n%30s",
+			 "MU NSS counts : ");
+
+	for (i = 1; i <= ARRAY_SIZE(stats->mu_nss); i++)
+		len += scnprintf(buf + len, buf_len - len, "%dx%d\t%-7d ",
+				 i, i, le32_to_cpu(stats->mu_nss[i - 1]));
+
 	len += scnprintf(buf + len, buf_len - len, "\n\n%30s",
 			 "BW counts (0..3) : ");
 
--- a/drivers/net/wireless/ath/ath10k/debug_htt_stats.h
+++ b/drivers/net/wireless/ath/ath10k/debug_htt_stats.h
@@ -192,6 +192,7 @@ struct htt_10_4_tx_rate_info {
 	__le32 rts_cnt;
 	__le32 ack_rssi;
 	__le32 mu_mcs[10];
+	__le32 mu_nss[2];
 } __packed;
 
 #define DBG_STATS_MAX_HWQ_NUM 10
