diff --git a/drivers/net/wireless/ath/ath10k/core.h b/drivers/net/wireless/ath/ath10k/core.h
index 72db6c495132..9bbfb9bcbf2f 100644
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -98,6 +98,7 @@ struct ath10k_skb_cb {
 	u8 eid;
 	u16 msdu_id;
 	struct ieee80211_txq *txq;
+	struct ieee80211_vif *vif;
 } __packed;
 
 struct ath10k_skb_rxcb {
@@ -618,6 +619,7 @@ enum ath10k_tx_pause_reason {
 struct ath10k {
 	struct ath_common ath_common;
 	struct ieee80211_hw *hw;
+	struct ieee80211_ops *ops;
 	struct device *dev;
 	u8 mac_addr[ETH_ALEN];
 
diff --git a/drivers/net/wireless/ath/ath10k/mac.c b/drivers/net/wireless/ath/ath10k/mac.c
index 70fc64226edf..b48591e7ec5b 100644
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -3329,11 +3329,11 @@ static void ath10k_tx_h_add_p2p_noa_ie(struct ath10k *ar,
 }
 
 void ath10k_mac_tx_h_fill_cb(struct ath10k *ar,
+			     struct ieee80211_vif *vif,
 			     struct ieee80211_txq *txq,
 			     struct sk_buff *skb)
 {
 	struct ieee80211_hdr *hdr = (void *)skb->data;
-	struct ieee80211_vif *vif = txq->vif;
 	struct ath10k_skb_cb *cb = ATH10K_SKB_CB(skb);
 
 	cb->flags = 0;
@@ -3347,6 +3347,7 @@ void ath10k_mac_tx_h_fill_cb(struct ath10k *ar,
 		cb->flags |= ATH10K_SKB_F_QOS;
 
 	cb->txq = txq;
+	cb->vif = vif;
 }
 
 bool ath10k_mac_tx_frm_has_freq(struct ath10k *ar)
@@ -3674,7 +3675,7 @@ int ath10k_mac_txq_push_one(struct ieee80211_hw *hw,
 
 	skb_len = skb->len;
 
-	ath10k_mac_tx_h_fill_cb(ar, txq, skb);
+	ath10k_mac_tx_h_fill_cb(ar, vif, txq, skb);
 
 	txmode = ath10k_mac_tx_h_get_txmode(ar, vif, sta, skb);
 	txpath = ath10k_mac_tx_h_get_txpath(ar, skb, txmode);
@@ -3720,6 +3721,9 @@ void ath10k_mac_txq_push_pending(struct ath10k *ar)
 
 	INIT_LIST_HEAD(&list);
 
+	if (!test_bit(ATH10K_FW_FEATURE_PEER_FLOW_CONTROL, ar->fw_features))
+		return;
+
 	rcu_read_lock();
 	spin_lock_bh(&ar->pending_txqs_lock);
 
@@ -3744,6 +3748,22 @@ void ath10k_mac_txq_push_pending(struct ath10k *ar)
 	rcu_read_unlock();
 }
 
+static void ath10k_mac_vif_cleanup(struct ath10k *ar,
+				   struct ieee80211_vif *vif)
+{
+	struct ath10k_skb_cb *cb;
+	struct sk_buff *msdu;
+	int msdu_id;
+
+	lockdep_assert_held(&ar->htt.tx_lock);
+
+	idr_for_each_entry(&ar->htt.pending_tx, msdu, msdu_id) {
+		cb = ATH10K_SKB_CB(msdu);
+		if (cb->vif == vif)
+			cb->vif = NULL;
+	}
+}
+
 static void ath10k_mac_txq_cleanup(struct ath10k *ar,
 				   struct ieee80211_txq *txq)
 {
@@ -3753,6 +3773,9 @@ static void ath10k_mac_txq_cleanup(struct ath10k *ar,
 
 	lockdep_assert_held(&ar->htt.tx_lock);
 
+	if (!txq)
+		return;
+
 	idr_for_each_entry(&ar->htt.pending_tx, msdu, msdu_id) {
 		cb = ATH10K_SKB_CB(msdu);
 		if (cb->txq == txq)
@@ -4098,7 +4121,7 @@ static void ath10k_mac_op_tx(struct ieee80211_hw *hw,
 
 	artxq = (void *)txq->drv_priv;
 
-	ath10k_mac_tx_h_fill_cb(ar, txq, skb);
+	ath10k_mac_tx_h_fill_cb(ar, vif, txq, skb);
 
 	txmode = ath10k_mac_tx_h_get_txmode(ar, vif, sta, skb);
 	txpath = ath10k_mac_tx_h_get_txpath(ar, skb, txmode);
@@ -5099,18 +5122,17 @@ static void ath10k_remove_interface(struct ieee80211_hw *hw,
 	ar->free_vdev_map |= 1LL << arvif->vdev_id;
 	list_del(&arvif->list);
 
+	spin_lock_bh(&ar->pending_txqs_lock);
+	spin_lock_bh(&ar->htt.tx_lock);
 	if (vif->txq) {
-		spin_lock_bh(&ar->pending_txqs_lock);
-		spin_lock_bh(&ar->htt.tx_lock);
-
 		if (!list_empty(&artxq->pending))
 			list_del_init(&artxq->pending);
 
 		ath10k_mac_txq_cleanup(ar, vif->txq);
-
-		spin_unlock_bh(&ar->htt.tx_lock);
-		spin_unlock_bh(&ar->pending_txqs_lock);
 	}
+	ath10k_mac_vif_cleanup(ar, vif);
+	spin_unlock_bh(&ar->htt.tx_lock);
+	spin_unlock_bh(&ar->pending_txqs_lock);
 
 	if (arvif->vdev_type == WMI_VDEV_TYPE_AP ||
 	    arvif->vdev_type == WMI_VDEV_TYPE_IBSS) {
@@ -7467,20 +7489,29 @@ static const struct ieee80211_channel ath10k_5ghz_channels[] = {
 struct ath10k *ath10k_mac_create(size_t priv_size)
 {
 	struct ieee80211_hw *hw;
+	struct ieee80211_ops *ops;
 	struct ath10k *ar;
 
-	hw = ieee80211_alloc_hw(sizeof(struct ath10k) + priv_size, &ath10k_ops);
-	if (!hw)
+	ops = kmemdup(&ath10k_ops, sizeof(ath10k_ops), GFP_KERNEL);
+	if (!ops)
 		return NULL;
 
+	hw = ieee80211_alloc_hw(sizeof(struct ath10k) + priv_size, ops);
+	if (!hw) {
+		kfree(ar->ops);
+		return NULL;
+	}
+
 	ar = hw->priv;
 	ar->hw = hw;
+	ar->ops = ops;
 
 	return ar;
 }
 
 void ath10k_mac_destroy(struct ath10k *ar)
 {
+	kfree(ar->ops);
 	ieee80211_free_hw(ar->hw);
 }
 
@@ -7710,6 +7741,22 @@ int ath10k_mac_register(struct ath10k *ar)
 	void *channels;
 	int ret;
 
+	/* FIXME: This isn't really nice. The ieee80211_ops structure is passed
+	 * as const when allocating mac80211/ath10k structures.
+	 *
+	 * Thankfully this works because wake_tx_queue (at least at the time of
+	 * this writing) is only used when creating interfaces or station
+	 * entries, i.e. after calling to ieee80211_register_hw().
+	 *
+	 * This prevents peak performance throughput regression on hw/fw combos
+	 * that do not support peer-flow-ctrl. The wake_tx_queue() tx path,
+	 * even when totally optimized in ath10k, still yields ~10% (1100mbps
+	 * -> 1000mbps) degradation on AP148 as of 2015-11-18 on coconut branch
+	 * head.
+	 */
+	if (!test_bit(ATH10K_FW_FEATURE_PEER_FLOW_CONTROL, ar->fw_features))
+		ar->ops->wake_tx_queue = NULL;
+
 	SET_IEEE80211_PERM_ADDR(ar->hw, ar->mac_addr);
 
 	SET_IEEE80211_DEV(ar->hw, ar->dev);
diff --git a/drivers/net/wireless/ath/ath10k/mac.h b/drivers/net/wireless/ath/ath10k/mac.h
index 8d5d565c9df1..ed32ea1607c3 100644
--- a/drivers/net/wireless/ath/ath10k/mac.h
+++ b/drivers/net/wireless/ath/ath10k/mac.h
@@ -78,6 +78,7 @@ u8 ath10k_mac_bitrate_to_idx(const struct ieee80211_supported_band *sband,
 			     u32 bitrate);
 
 void ath10k_mac_tx_h_fill_cb(struct ath10k *ar,
+			     struct ieee80211_vif *vif,
 			     struct ieee80211_txq *txq,
 			     struct sk_buff *skb);
 
-- 
2.1.4

