--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -2462,6 +2462,29 @@ ath10k_peer_assoc_h_vht_limit(u16 tx_mcs
 	return tx_mcs_set;
 }
 
+static u32 get_160m_nss_from_maxrate(int rate)
+{
+	u32 nss;
+
+	switch(rate) {
+		case 780:
+			nss = 1;
+			break;
+		case 1560:
+			nss = 2;
+			break;
+		case 2106:
+			nss = 3; /* not support MCS9 from spec*/
+			break;
+		case 3120:
+			nss = 4;
+			break;
+		default:
+			nss = 0; /* don't know*/
+	}
+
+	return nss;
+}
 static void ath10k_peer_assoc_h_vht(struct ath10k *ar,
 				    struct ieee80211_vif *vif,
 				    struct ieee80211_sta *sta,
@@ -2469,12 +2492,14 @@ static void ath10k_peer_assoc_h_vht(stru
 {
 	const struct ieee80211_sta_vht_cap *vht_cap = &sta->vht_cap;
 	struct ath10k_vif *arvif = (void *)vif->drv_priv;
+	struct ath10k_hw_params *hw = &ar->hw_params;
 	struct cfg80211_chan_def def;
 	enum nl80211_band band;
 	const u16 *vht_mcs_mask;
 	u8 ampdu_factor;
 	u8 max_nss, vht_mcs;
 	int i;
+	u32 peer_rx_nss_160;
 
 	if (WARN_ON(ath10k_mac_vif_chan(vif, &def)))
 		return;
@@ -2535,22 +2560,24 @@ static void ath10k_peer_assoc_h_vht(stru
 	arg->peer_vht_rates.tx_mcs_set = ath10k_peer_assoc_h_vht_limit(
 		__le16_to_cpu(vht_cap->vht_mcs.tx_mcs_map), vht_mcs_mask);
 
-	ath10k_dbg(ar, ATH10K_DBG_MAC, "mac vht peer %pM max_mpdu %d flags 0x%x\n",
-		   sta->addr, arg->peer_max_mpdu, arg->peer_flags);
+	/* Configure bandwidth-NSS mapping to FW for the chip's tx chains setting on 160Mhz bw */
+	if ((arg->peer_phymode == MODE_11AC_VHT160) || (arg->peer_phymode == MODE_11AC_VHT80_80)) {
+		peer_rx_nss_160 = get_160m_nss_from_maxrate(arg->peer_vht_rates.rx_max_rate);
 
-	if (arg->peer_vht_rates.rx_max_rate &&
-	    (sta->vht_cap.cap & IEEE80211_VHT_CAP_SUPP_CHAN_WIDTH_MASK)) {
-		switch (arg->peer_vht_rates.rx_max_rate) {
-		case 1560:
-			/* Must be 2x2 at 160Mhz is all it can do. */
-			arg->peer_bw_rxnss_override = 2;
-			break;
-		case 780:
-			/* Can only do 1x1 at 160Mhz (Long Guard Interval) */
-			arg->peer_bw_rxnss_override = 1;
-			break;
+		peer_rx_nss_160 = (peer_rx_nss_160 == 0) ?
+			arg->peer_num_spatial_streams : min(arg->peer_num_spatial_streams, peer_rx_nss_160);
+
+		peer_rx_nss_160 = min(peer_rx_nss_160, get_160m_nss_from_maxrate(hw->vht160_mcs_tx_highest));
+
+		arg->peer_bw_rxnss_override = WMI_PEER_BW_NSS_FWCONF_160(peer_rx_nss_160);
+
+		if (arg->peer_phymode == MODE_11AC_VHT80_80) {
+			arg->peer_bw_rxnss_override |= WMI_PEER_BW_NSS_FWCONF_80_80(peer_rx_nss_160);
 		}
 	}
+
+	ath10k_dbg(ar, ATH10K_DBG_MAC, "mac vht peer %pM max_mpdu %d flags 0x%x peer_bw_rxnss_override 0x%x,\n",
+			   sta->addr, arg->peer_max_mpdu, arg->peer_flags, arg->peer_bw_rxnss_override);
 }
 
 static void ath10k_peer_assoc_h_qos(struct ath10k *ar,
@@ -2702,9 +2729,9 @@ static int ath10k_peer_assoc_prepare(str
 	ath10k_peer_assoc_h_crypto(ar, vif, sta, arg);
 	ath10k_peer_assoc_h_rates(ar, vif, sta, arg);
 	ath10k_peer_assoc_h_ht(ar, vif, sta, arg);
+	ath10k_peer_assoc_h_phymode(ar, vif, sta, arg);
 	ath10k_peer_assoc_h_vht(ar, vif, sta, arg);
 	ath10k_peer_assoc_h_qos(ar, vif, sta, arg);
-	ath10k_peer_assoc_h_phymode(ar, vif, sta, arg);
 
 	return 0;
 }
@@ -4624,8 +4651,6 @@ static struct ieee80211_sta_vht_cap ath1
 
 	vht_cap.vht_supported = 1;
 	vht_cap.cap = ar->vht_cap_info;
-	/* Disabling VHT160 and VHT80+80 modes */
-	vht_cap.cap &= ~IEEE80211_VHT_CAP_SUPP_CHAN_WIDTH_MASK;
 
 	if (ar->vht_cap_info & (IEEE80211_VHT_CAP_SU_BEAMFORMEE_CAPABLE |
 				IEEE80211_VHT_CAP_MU_BEAMFORMEE_CAPABLE)) {
@@ -8688,7 +8713,9 @@ static const struct ieee80211_iface_comb
 		.radar_detect_widths =	BIT(NL80211_CHAN_WIDTH_20_NOHT) |
 					BIT(NL80211_CHAN_WIDTH_20) |
 					BIT(NL80211_CHAN_WIDTH_40) |
-					BIT(NL80211_CHAN_WIDTH_80),
+					BIT(NL80211_CHAN_WIDTH_80) |
+					BIT(NL80211_CHAN_WIDTH_80P80) |
+					BIT(NL80211_CHAN_WIDTH_160),
 #endif
 	},
 };
@@ -8706,7 +8733,9 @@ ieee80211_iface_combination ath10k_10_4_
 		.radar_detect_widths =  BIT(NL80211_CHAN_WIDTH_20_NOHT) |
 					BIT(NL80211_CHAN_WIDTH_20) |
 					BIT(NL80211_CHAN_WIDTH_40) |
-					BIT(NL80211_CHAN_WIDTH_80),
+					BIT(NL80211_CHAN_WIDTH_80) |
+					BIT(NL80211_CHAN_WIDTH_80P80) |
+					BIT(NL80211_CHAN_WIDTH_160),
 #endif
 	},
 };
--- a/drivers/net/wireless/ath/ath10k/wmi.c
+++ b/drivers/net/wireless/ath/ath10k/wmi.c
@@ -1732,12 +1732,22 @@ void ath10k_wmi_put_wmi_channel(struct w
 	if (arg->chan_radar)
 		flags |= WMI_CHAN_FLAG_DFS;
 
+	ch->band_center_freq2 = 0;
 	ch->mhz = __cpu_to_le32(arg->freq);
 	ch->band_center_freq1 = __cpu_to_le32(arg->band_center_freq1);
 	if (arg->mode == MODE_11AC_VHT80_80)
 		ch->band_center_freq2 = __cpu_to_le32(arg->band_center_freq2);
-	else
-		ch->band_center_freq2 = 0;
+
+	/* Follow FW setting */
+	if (arg->mode == MODE_11AC_VHT160)
+	{
+		if (arg->freq > arg->band_center_freq1)
+			ch->band_center_freq1 = __cpu_to_le32(arg->band_center_freq1 + 40);
+		else
+			ch->band_center_freq1 = __cpu_to_le32(arg->band_center_freq1 - 40);
+
+		ch->band_center_freq2 = __cpu_to_le32(arg->band_center_freq1);
+	}
 	ch->min_power = arg->min_power;
 	ch->max_power = arg->max_power;
 	ch->reg_power = arg->max_reg_power;
@@ -7651,9 +7661,7 @@ ath10k_wmi_peer_assoc_fill_10_4(struct a
 
 	ath10k_wmi_peer_assoc_fill_10_2(ar, buf, arg);
 	if (arg->peer_bw_rxnss_override)
-		cmd->peer_bw_rxnss_override =
-			__cpu_to_le32((arg->peer_bw_rxnss_override - 1) |
-				      BIT(PEER_BW_RXNSS_OVERRIDE_OFFSET));
+		cmd->peer_bw_rxnss_override = __cpu_to_le32(arg->peer_bw_rxnss_override);
 	else
 		cmd->peer_bw_rxnss_override = 0;
 }
--- a/drivers/net/wireless/ath/ath10k/wmi.h
+++ b/drivers/net/wireless/ath/ath10k/wmi.h
@@ -6577,7 +6577,15 @@ struct wmi_10_2_peer_assoc_complete_cmd
 	__le32 info0; /* WMI_PEER_ASSOC_INFO0_ */
 } __packed;
 
-#define PEER_BW_RXNSS_OVERRIDE_OFFSET  31
+/* NSS Mapping to FW */
+#define WMI_PEER_BW_NSS_FWCONF_MAP_ENABLE             (1 << 31)
+#define WMI_PEER_BW_NSS_FWCONF_MAP_160MHZ_LSB         (0)
+#define WMI_PEER_BW_NSS_FWCONF_MAP_160MHZ_MASK        (0x00000007)
+#define WMI_PEER_BW_NSS_FWCONF_MAP_80_80MHZ_LSB       (3)
+#define WMI_PEER_BW_NSS_FWCONF_MAP_80_80MHZ_MASK      (0x00000038)
+#define WMI_PEER_BW_NSS_FWCONF_MAP_MASK               (0x0000003F)
+#define WMI_PEER_BW_NSS_FWCONF_160(x)          		  (WMI_PEER_BW_NSS_FWCONF_MAP_ENABLE | SM((x - 1), WMI_PEER_BW_NSS_FWCONF_MAP_160MHZ))
+#define WMI_PEER_BW_NSS_FWCONF_80_80(x)        		  (WMI_PEER_BW_NSS_FWCONF_MAP_ENABLE | SM((x - 1), WMI_PEER_BW_NSS_FWCONF_MAP_80_80MHZ))
 
 struct wmi_10_4_peer_assoc_complete_cmd {
 	struct wmi_10_2_peer_assoc_complete_cmd cmd;
