--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -6547,6 +6547,7 @@ static int ath10k_sta_state(struct ieee8
 			arsta->tx_stats = kzalloc(sizeof(*arsta->tx_stats),
 							 GFP_KERNEL);
 			if (!arsta->tx_stats) {
+				ath10k_mac_dec_num_stations(arvif, sta);
 				ret = -ENOMEM;
 				goto exit;
 			}
@@ -6557,7 +6558,8 @@ static int ath10k_sta_state(struct ieee8
 		if (ret) {
 			ath10k_warn(ar, "failed to add peer %pM for vdev %d when adding a new sta: %i\n",
 				    sta->addr, arvif->vdev_id, ret);
-			kfree(arsta->tx_stats);
+			if (ar->debug.enable_extd_tx_stats)
+				kfree(arsta->tx_stats);
 			ath10k_mac_dec_num_stations(arvif, sta);
 			for (i = 0; i < ARRAY_SIZE(sta->txq); i++)
 				ath10k_mac_txq_unref(ar, sta->txq[i]);
@@ -6572,7 +6574,8 @@ static int ath10k_sta_state(struct ieee8
 			ath10k_warn(ar, "failed to lookup peer %pM on vdev %i\n",
 				    vif->addr, arvif->vdev_id);
 			spin_unlock_bh(&ar->data_lock);
-			kfree(arsta->tx_stats);
+			if (ar->debug.enable_extd_tx_stats)
+				kfree(arsta->tx_stats);
 			ath10k_peer_delete(ar, arvif->vdev_id, sta->addr);
 			ath10k_mac_dec_num_stations(arvif, sta);
 			kfree(arsta->tx_stats);
@@ -6593,7 +6596,8 @@ static int ath10k_sta_state(struct ieee8
 		if (ret) {
 			ath10k_warn(ar, "failed to update fw tdls state on vdev %i: %i\n",
 				    arvif->vdev_id, ret);
-			kfree(arsta->tx_stats);
+			if (ar->debug.enable_extd_tx_stats)
+				kfree(arsta->tx_stats);
 			ath10k_peer_delete(ar, arvif->vdev_id,
 					   sta->addr);
 			ath10k_mac_dec_num_stations(arvif, sta);
@@ -6607,7 +6611,8 @@ static int ath10k_sta_state(struct ieee8
 			ath10k_warn(ar,
 				    "failed to update tdls peer %pM for vdev %d when adding a new sta: %i\n",
 				    sta->addr, arvif->vdev_id, ret);
-			kfree(arsta->tx_stats);
+			if (ar->debug.enable_extd_tx_stats)
+				kfree(arsta->tx_stats);
 			ath10k_peer_delete(ar, arvif->vdev_id, sta->addr);
 			ath10k_mac_dec_num_stations(arvif, sta);
 			kfree(arsta->tx_stats);
