--- a/drivers/net/wireless/ath/ath10k/txrx.c
+++ b/drivers/net/wireless/ath/ath10k/txrx.c
@@ -113,6 +113,9 @@ int ath10k_txrx_tx_unref(struct ath10k_htt *htt,
 	    (info->flags & IEEE80211_TX_CTL_NO_ACK))
 		info->flags |= IEEE80211_TX_STAT_NOACK_TRANSMITTED;
 
+	/* we don't provide rate info */
+	info->status.rates[0].idx = -1;
+
 	if (tx_done->status == HTT_TX_COMPL_STATE_DISCARD) {
 		if (info->flags & IEEE80211_TX_CTL_NO_ACK)
 			info->flags &= ~IEEE80211_TX_STAT_NOACK_TRANSMITTED;
