diff --git a/drivers/net/wireless/ath/ath10k/core.h b/drivers/net/wireless/ath/ath10k/core.h
index 413f84661232..3b8d052b339a 100644
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -332,6 +332,7 @@ enum ath10k_beacon_state {
 
 struct ath10k_txq {
 	struct list_head pending;
+	unsigned long num_fw_queued;
 };
 
 struct ath10k_vif {
diff --git a/drivers/net/wireless/ath/ath10k/mac.c b/drivers/net/wireless/ath/ath10k/mac.c
index 80b18f427ed4..3c46610f6617 100644
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -3475,6 +3475,8 @@ static void ath10k_mac_txq_push(struct ieee80211_hw *hw,
 			break;
 		}
 
+		artxq->num_fw_queued++;
+
 		spin_unlock_bh(&ar->htt.tx_lock);
 
 		skb = ieee80211_tx_dequeue(hw, txq);
@@ -3483,6 +3485,7 @@ static void ath10k_mac_txq_push(struct ieee80211_hw *hw,
 			if (!list_empty(&artxq->pending))
 				list_del_init(&artxq->pending);
 			ath10k_htt_tx_dec_pending(htt, is_mgmt);
+			artxq->num_fw_queued--;
 			spin_unlock_bh(&ar->htt.tx_lock);
 			break;
 		}
@@ -3498,6 +3501,7 @@ static void ath10k_mac_txq_push(struct ieee80211_hw *hw,
 				    ret);
 			spin_lock_bh(&ar->htt.tx_lock);
 			ath10k_htt_tx_dec_pending(htt, is_mgmt);
+			artxq->num_fw_queued--;
 			spin_unlock_bh(&ar->htt.tx_lock);
 			continue;
 		}
@@ -3866,6 +3870,7 @@ static void ath10k_mac_op_tx(struct ieee80211_hw *hw,
 {
 	struct ath10k *ar = hw->priv;
 	struct ath10k_htt *htt = &ar->htt;
+	struct ath10k_txq *artxq;
 	struct ieee80211_tx_info *info = IEEE80211_SKB_CB(skb);
 	struct ieee80211_vif *vif = info->control.vif;
 	struct ieee80211_sta *sta = control->sta;
@@ -3885,6 +3890,8 @@ static void ath10k_mac_op_tx(struct ieee80211_hw *hw,
 	else
 		txq = NULL;
 
+	artxq = (void *)txq->drv_priv;
+
 	ath10k_mac_tx_h_fill_cb(ar, txq, skb);
 
 	txmode = ath10k_mac_tx_h_get_txmode(ar, vif, sta, skb);
@@ -3907,6 +3914,9 @@ static void ath10k_mac_op_tx(struct ieee80211_hw *hw,
 			return;
 		}
 
+		if (txq)
+			artxq->num_fw_queued++;
+
 		spin_unlock_bh(&ar->htt.tx_lock);
 	}
 
@@ -3916,6 +3926,8 @@ static void ath10k_mac_op_tx(struct ieee80211_hw *hw,
 		if (is_htt) {
 			spin_lock_bh(&ar->htt.tx_lock);
 			ath10k_htt_tx_dec_pending(htt, is_mgmt);
+			if (txq)
+				artxq->num_fw_queued--;
 			spin_unlock_bh(&ar->htt.tx_lock);
 		}
 		return;
diff --git a/drivers/net/wireless/ath/ath10k/txrx.c b/drivers/net/wireless/ath/ath10k/txrx.c
index 118586ece20e..ff4ef0adb1fe 100644
--- a/drivers/net/wireless/ath/ath10k/txrx.c
+++ b/drivers/net/wireless/ath/ath10k/txrx.c
@@ -53,6 +53,7 @@ void ath10k_txrx_tx_unref(struct ath10k_htt *htt,
 			  const struct htt_tx_done *tx_done)
 {
 	struct ath10k *ar = htt->ar;
+	struct ath10k_txq *artxq;
 	struct device *dev = ar->dev;
 	struct ieee80211_tx_info *info;
 	struct ath10k_skb_cb *skb_cb;
@@ -81,6 +82,12 @@ void ath10k_txrx_tx_unref(struct ath10k_htt *htt,
 
 	skb_cb = ATH10K_SKB_CB(msdu);
 
+	if (skb_cb->txq) {
+		artxq = (void *)skb_cb->txq->drv_priv;
+		if (!WARN_ON_ONCE(artxq->num_fw_queued == 0))
+			artxq->num_fw_queued--;
+	}
+
 	if (unlikely(skb_cb->flags & ATH10K_SKB_F_MGMT) &&
 	    ar->hw_params.max_probe_resp_desc_thres)
 		limit_mgmt_desc = true;
-- 
2.1.4

