--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -9164,11 +9164,6 @@ int ath10k_mac_register(struct ath10k *a
 		wiphy_ext_feature_set(ar->hw->wiphy,
 				      NL80211_EXT_FEATURE_ACK_SIGNAL_SUPPORT);
 
-	if (ath10k_peer_stats_enabled(ar) ||
-	    test_bit(WMI_SERVICE_REPORT_AIRTIME, ar->wmi.svc_map))
-		wiphy_ext_feature_set(ar->hw->wiphy,
-				      NL80211_EXT_FEATURE_AIRTIME_FAIRNESS);
-
 	if (test_bit(WMI_SERVICE_RTT_RESPONDER_ROLE, ar->wmi.svc_map))
 		wiphy_ext_feature_set(ar->hw->wiphy,
 				      NL80211_EXT_FEATURE_ENABLE_FTM_RESPONDER);
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -3710,7 +3710,7 @@ struct ieee80211_txq *ieee80211_next_txq
 }
 EXPORT_SYMBOL(ieee80211_next_txq);
 
-static void __ieee80211_insert_txq(struct rb_root_cached *root,
+static void __ieee80211_insert_txq(struct rb_root_cached *root, struct ieee80211_hw *hw,
                                    struct txq_info *txqi, u8 ac)
 {
 	struct rb_node **new = &root->rb_root.rb_node;
@@ -3723,8 +3723,15 @@ static void __ieee80211_insert_txq(struc
 		__txqi = rb_entry(parent, struct txq_info, schedule_order);
 
 		if (!txqi->txq.sta) {
-			/* new txqi has no sta - insert to the left */
-			new = &parent->rb_left;
+			if (wiphy_ext_feature_isset(hw->wiphy,
+				    NL80211_EXT_FEATURE_AIRTIME_FAIRNESS))
+				/* new txqi has no sta - insert to the left */
+				new = &parent->rb_left;
+			else {
+				new = &parent->rb_right;
+				leftmost = false;
+			}
+
 		} else if (!__txqi->txq.sta) {
 			/* existing txqi has no sta - insert to the right */
 			new = &parent->rb_right;
@@ -3737,7 +3744,9 @@ static void __ieee80211_insert_txq(struc
 												struct sta_info,
 												sta);
 
-			if (new_sta->airtime[ac].v_t <= old_sta->airtime[ac].v_t)
+			if (wiphy_ext_feature_isset(hw->wiphy,
+						NL80211_EXT_FEATURE_AIRTIME_FAIRNESS) &&
+				new_sta->airtime[ac].v_t <= old_sta->airtime[ac].v_t)
 				new = &parent->rb_left;
 			else {
 				new = &parent->rb_right;
@@ -3779,7 +3788,7 @@ void ieee80211_schedule_txq(struct ieee8
 
 	}
 
-	__ieee80211_insert_txq(&local->active_txqs[ac], txqi, ac);
+	__ieee80211_insert_txq(&local->active_txqs[ac], hw, txqi, ac);
 
 out:
 	spin_unlock_bh(&local->active_txq_lock[ac]);
@@ -3797,7 +3806,7 @@ void ieee80211_resort_txq(struct ieee802
 		rb_erase_cached(&txqi->schedule_order,
 	                       &local->active_txqs[ac]);
 		RB_CLEAR_NODE(&txqi->schedule_order);
-		__ieee80211_insert_txq(&local->active_txqs[ac], txqi, ac);
+		__ieee80211_insert_txq(&local->active_txqs[ac], hw, txqi, ac);
 	}
 }
 
