--- a/drivers/net/wireless/ath/ath10k/txrx.c
+++ b/drivers/net/wireless/ath/ath10k/txrx.c
@@ -122,9 +122,11 @@ int ath10k_txrx_tx_unref(struct ath10k_htt *htt,
 	if (tx_done->status == HTT_TX_COMPL_STATE_NOACK)
 		info->flags &= ~IEEE80211_TX_STAT_ACK;
 
-	if (tx_done->ack_rssi && tx_done->ack_rssi != ATH10K_INVALID_RSSI)
+	if (tx_done->ack_rssi && tx_done->ack_rssi != ATH10K_INVALID_RSSI) {
 		info->status.ack_signal = ATH10K_DEFAULT_NOISE_FLOOR +
 						tx_done->ack_rssi;
+		info->status.is_valid_ack_signal = true;
+	}
 
 	if ((tx_done->status == HTT_TX_COMPL_STATE_ACK) &&
 	    (info->flags & IEEE80211_TX_CTL_NO_ACK))
--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -954,6 +954,7 @@ struct ieee80211_tx_info {
 			u8 ampdu_len;
 			u8 antenna;
 			u16 tx_time;
+			bool is_valid_ack_signal;
 			void *status_driver_data[19 / sizeof(void *)];
 		} status;
 		struct {
--- a/net/mac80211/status.c
+++ b/net/mac80211/status.c
@@ -191,9 +191,12 @@ static void ieee80211_frame_acked(struct sta_info *sta, struct sk_buff *skb)
 
 	if (ieee80211_hw_check(&local->hw, REPORTS_TX_ACK_STATUS)) {
 		sta->status_stats.last_ack = jiffies;
-		sta->status_stats.last_ack_signal = (s8) txinfo->status.ack_signal;
-		ewma_avg_signal_add(&sta->avg_ack_signal,
-				    -(txinfo->status.ack_signal));
+		if (txinfo->status.is_valid_ack_signal) {
+			sta->status_stats.last_ack_signal =
+					    (s8) txinfo->status.ack_signal;
+			ewma_avg_signal_add(&sta->avg_ack_signal,
+					    -(txinfo->status.ack_signal));
+		}
 	}
 
 	if (ieee80211_is_data_qos(mgmt->frame_control)) {
@@ -1034,6 +1037,12 @@ void ieee80211_tx_status_8023(struct ieee80211_hw *hw,
 			ieee80211_sta_reset_conn_monitor(sdata);
 
 		sta->status_stats.last_ack = jiffies;
+		if (info->status.is_valid_ack_signal) {
+			sta->status_stats.last_ack_signal =
+			        (s8) info->status.ack_signal;
+			ewma_avg_signal_add(&sta->avg_ack_signal,
+				-(info->status.ack_signal));
+		}
 		if (info->flags & IEEE80211_TX_STAT_ACK) {
 			if (sta->status_stats.lost_packets)
 				sta->status_stats.lost_packets = 0;
-- 
2.7.4

