--- a/drivers/net/wireless/ath/ath10k/htt.h
+++ b/drivers/net/wireless/ath/ath10k/htt.h
@@ -1345,7 +1345,8 @@ enum htt_q_depth_type {
 	HTT_Q_DEPTH_TYPE_MSDUS = 1,
 };
 
-#define HTT_TX_Q_STATE_NUM_PEERS		32 /* FIXME: how to get more? */
+#define HTT_TX_Q_STATE_NUM_PEERS	(TARGET_10_4_NUM_QCACHE_PEERS_MAX + \
+					TARGET_10_4_NUM_VDEVS)
 #define HTT_TX_Q_STATE_NUM_TIDS			8
 #define HTT_TX_Q_STATE_ENTRY_SIZE		1
 #define HTT_TX_Q_STATE_ENTRY_MULTIPLIER		0
