Index: backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/core.h
===================================================================
--- backports-20171004-3.14.77-c09dbd7.orig/drivers/net/wireless/ath/ath10k/core.h	2017-12-21 22:18:27.425965310 +0530
+++ backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/core.h	2017-12-21 22:20:31.305454246 +0530
@@ -429,6 +429,45 @@
 };
 
 
+enum ath10k_pkt_rx_err {
+	ATH10K_PKT_RX_ERR_FCS,
+	ATH10K_PKT_RX_ERR_TKIP,
+	ATH10K_PKT_RX_ERR_CRYPT,
+	ATH10K_PKT_RX_ERR_PEER_IDX_INVAL,
+	ATH10K_PKT_RX_ERR_MAX,
+};
+
+enum ath10k_vow_stats_ampdu {
+	ATH10K_VOW_STATS_AMPDU_10,
+	ATH10K_VOW_STATS_AMPDU_20,
+	ATH10K_VOW_STATS_AMPDU_30,
+	ATH10K_VOW_STATS_AMPDU_40,
+	ATH10K_VOW_STATS_AMPDU_50,
+	ATH10K_VOW_STATS_AMPDU_60,
+	ATH10K_VOW_STATS_AMPDU_MORE,
+	ATH10K_VOW_STATS_AMPDU_NUM_MAX,
+};
+
+enum ath10k_vow_stats_amsdu {
+	ATH10K_VOW_STATS_AMSDU_1,
+	ATH10K_VOW_STATS_AMSDU_2,
+	ATH10K_VOW_STATS_AMSDU_3,
+	ATH10K_VOW_STATS_AMSDU_4,
+	ATH10K_VOW_STATS_AMSDU_MORE,
+	ATH10K_VOW_STATS_AMSDU_NUM_MAX,
+};
+
+struct ath10k_vow_stats {
+	unsigned long int rx_pkt_from_fw;
+	unsigned long int rx_pkt_unchained;
+	unsigned long int rx_pkt_drop_chained;
+	unsigned long int rx_pkt_drop_filter;
+	unsigned long int rx_pkt_err[ATH10K_PKT_RX_ERR_MAX];
+	unsigned long int rx_pkt_to_mac;
+	unsigned long int rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_NUM_MAX];
+	unsigned long int rx_pkt_amsdu[ATH10K_VOW_STATS_AMSDU_NUM_MAX];
+};
+
 struct ath10k_sta {
 	struct ath10k_vif *arvif;
 
@@ -454,6 +493,9 @@
 	u8 tpc;
 	u32 peer_ps_state;
 	enum nl80211_cqm_txrate_threshold_event last_event;
+
+	/* Protected with ar->data_lock */
+	struct ath10k_vow_stats vow_stats;
 };
 
 #define ATH10K_VDEV_SETUP_TIMEOUT_HZ (5 * HZ)
@@ -1123,6 +1165,8 @@
 	enum ath10k_dt_btcoex_support_flag btcoex_support;
 	int btcoex_gpio_pin;
 
+	bool vow_stats_enabled;
+
 	/* must be last */
 	u8 drv_priv[0] __aligned(sizeof(void *));
 };
@@ -1152,6 +1196,11 @@
 	return true;
 }
 
+static inline bool ath10k_vow_stats_update_required(struct ath10k *ar, u8 tid)
+{
+	return ar->vow_stats_enabled && (tid == 4 || tid == 5);
+}
+
 struct ath10k *ath10k_core_create(size_t priv_size, struct device *dev,
 				  enum ath10k_bus bus,
 				  enum ath10k_hw_rev hw_rev,
Index: backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/debug.c
===================================================================
--- backports-20171004-3.14.77-c09dbd7.orig/drivers/net/wireless/ath/ath10k/debug.c	2017-12-21 22:18:27.425965310 +0530
+++ backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/debug.c	2017-12-21 22:18:27.421965326 +0530
@@ -2976,6 +2976,9 @@
 	debugfs_create_file("fw_test", 0600, ar->debug.debugfs_phy, ar,
 			    &fops_fw_test);
 
+	debugfs_create_bool("enable_vow_stats", 0200, ar->debug.debugfs_phy,
+			    &ar->vow_stats_enabled);
+
 	return 0;
 }
 
Index: backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/debugfs_sta.c
===================================================================
--- backports-20171004-3.14.77-c09dbd7.orig/drivers/net/wireless/ath/ath10k/debugfs_sta.c	2017-12-21 22:18:27.425965310 +0530
+++ backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/debugfs_sta.c	2017-12-21 22:18:27.421965326 +0530
@@ -861,6 +861,148 @@
 	.llseek = default_llseek,
 };
 
+static char *get_err_str(enum ath10k_pkt_rx_err i)
+{
+	switch (i) {
+	case ATH10K_PKT_RX_ERR_FCS:
+		return "fcs_err";
+	case ATH10K_PKT_RX_ERR_TKIP:
+		return "tkip_err";
+	case ATH10K_PKT_RX_ERR_CRYPT:
+		return "crypt_err";
+	case ATH10K_PKT_RX_ERR_PEER_IDX_INVAL:
+		return "peer_idx_inval";
+	case ATH10K_PKT_RX_ERR_MAX:
+		return "unknown";
+	}
+
+	return "unknown";
+}
+
+static char *get_num_ampdu_subfrm_str(enum ath10k_vow_stats_ampdu i)
+{
+	switch (i) {
+	case ATH10K_VOW_STATS_AMPDU_10:
+		return "upto 10";
+	case ATH10K_VOW_STATS_AMPDU_20:
+		return "11-20";
+	case ATH10K_VOW_STATS_AMPDU_30:
+		return "21-30";
+	case ATH10K_VOW_STATS_AMPDU_40:
+		return "31-40";
+	case ATH10K_VOW_STATS_AMPDU_50:
+		return "41-50";
+	case ATH10K_VOW_STATS_AMPDU_60:
+		return "51-60";
+	case ATH10K_VOW_STATS_AMPDU_MORE:
+		return ">60";
+	case ATH10K_VOW_STATS_AMPDU_NUM_MAX:
+		return "0";
+	}
+
+	return "0";
+}
+
+static char *get_num_amsdu_subfrm_str(enum ath10k_vow_stats_amsdu i)
+{
+	switch (i) {
+	case ATH10K_VOW_STATS_AMSDU_1:
+		return "1";
+	case ATH10K_VOW_STATS_AMSDU_2:
+		return "2";
+	case ATH10K_VOW_STATS_AMSDU_3:
+		return "3";
+	case ATH10K_VOW_STATS_AMSDU_4:
+		return "4";
+	case ATH10K_VOW_STATS_AMSDU_MORE:
+		return ">4";
+	case ATH10K_VOW_STATS_AMSDU_NUM_MAX:
+		return "0";
+	}
+
+	return "0";
+}
+
+static ssize_t ath10k_dbg_sta_read_vow_stats(struct file *file,
+					     char __user *user_buf,
+					     size_t count, loff_t *ppos)
+{
+	struct ieee80211_sta *sta = file->private_data;
+	struct ath10k_sta *arsta = (struct ath10k_sta *)sta->drv_priv;
+	struct ath10k *ar = arsta->arvif->ar;
+	struct ath10k_vow_stats *stats;
+	size_t len = 0, buf_len = 2048;
+	char *buf;
+	int i;
+	ssize_t ret;
+
+	buf = kzalloc(buf_len, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
+	mutex_lock(&ar->conf_mutex);
+
+	spin_lock_bh(&ar->data_lock);
+
+	stats = &arsta->vow_stats;
+
+	len += scnprintf(buf + len, buf_len - len,
+			 "\t\tAth10k Rx Video pkt stats\n");
+	len += scnprintf(buf + len, buf_len - len,
+			 "\t\t=========================\n");
+	len += scnprintf(buf + len, buf_len - len,
+			 "MSDUs from FW\t\t\t\t\t%lu\n",
+			 stats->rx_pkt_from_fw);
+	len += scnprintf(buf + len, buf_len - len,
+			 "MSDUs unchained\t\t\t\t\t%lu\n",
+			 stats->rx_pkt_unchained);
+	len += scnprintf(buf + len, buf_len - len,
+			 "MSDUs locally dropped:chained\t\t\t%lu\n",
+			 stats->rx_pkt_drop_chained);
+	len += scnprintf(buf + len, buf_len - len,
+			 "MSDUs locally dropped:filtered\t\t\t%lu\n",
+			 stats->rx_pkt_drop_filter);
+	len += scnprintf(buf + len, buf_len - len,
+			 "MSDUs to mac80211\t\t\t\t%lu\n",
+			 stats->rx_pkt_to_mac);
+	for (i = 0; i < ATH10K_PKT_RX_ERR_MAX; i++) {
+		len += scnprintf(buf + len, buf_len - len,
+				 "MSDUs with error:%s\t\t\t%lu\n",
+			 get_err_str(i), stats->rx_pkt_err[i]);
+	}
+
+	len += scnprintf(buf + len, buf_len - len, "\n");
+	for (i = 0; i < ATH10K_VOW_STATS_AMPDU_NUM_MAX; i++) {
+		len += scnprintf(buf + len, buf_len - len,
+				 "A-MPDU num subframes %s\t\t\t%lu\n",
+				 get_num_ampdu_subfrm_str(i), stats->rx_pkt_ampdu[i]);
+	}
+
+	len += scnprintf(buf + len, buf_len - len, "\n");
+	for (i = 0; i < ATH10K_VOW_STATS_AMSDU_NUM_MAX; i++) {
+		len += scnprintf(buf + len, buf_len - len,
+				 "A-MSDU num subframes %s\t\t\t\t%lu\n",
+				 get_num_amsdu_subfrm_str(i), stats->rx_pkt_amsdu[i]);
+	}
+
+	spin_unlock_bh(&ar->data_lock);
+
+	ret = simple_read_from_buffer(user_buf, count, ppos, buf, len);
+
+	kfree(buf);
+
+	mutex_unlock(&ar->conf_mutex);
+
+	return ret;
+}
+
+static const struct file_operations fops_vow_stats_dump = {
+	.open = simple_open,
+	.read = ath10k_dbg_sta_read_vow_stats,
+	.owner = THIS_MODULE,
+	.llseek = default_llseek,
+};
+
 void ath10k_sta_add_debugfs(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			    struct ieee80211_sta *sta, struct dentry *dir)
 {
@@ -876,4 +1018,6 @@
 			    &fops_set_tpc);
 	debugfs_create_file("peer_tid_log", S_IRUSR, dir, sta, &fops_peer_tid_log);
 	debugfs_create_file("peer_ps_state", S_IRUSR, dir, sta, &fops_peer_ps_state);
+	debugfs_create_file("dump_vow_stats", 0400, dir, sta,
+			    &fops_vow_stats_dump);
 }
Index: backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/htt_rx.c
===================================================================
--- backports-20171004-3.14.77-c09dbd7.orig/drivers/net/wireless/ath/ath10k/htt_rx.c	2017-12-21 22:18:27.425965310 +0530
+++ backports-20171004-3.14.77-c09dbd7/drivers/net/wireless/ath/ath10k/htt_rx.c	2017-12-21 22:21:34.849191933 +0530
@@ -282,6 +282,8 @@
 	ath10k_dbg_dump(ar, ATH10K_DBG_HTT_DUMP, NULL, "htt rx netbuf pop: ",
 			msdu->data, msdu->len + skb_tailroom(msdu));
 
+	__net_timestamp(msdu);
+
 	return msdu;
 }
 
@@ -1382,7 +1384,9 @@
 static void ath10k_htt_rx_h_mpdu(struct ath10k *ar,
 				 struct sk_buff_head *amsdu,
 				 struct ieee80211_rx_status *status,
-				 bool fill_crypt_header)
+				 bool fill_crypt_header,
+				 u8 *rx_hdr,
+				 enum ath10k_pkt_rx_err *err)
 {
 	struct sk_buff *first;
 	struct sk_buff *last;
@@ -1418,6 +1422,9 @@
 	hdr = (void *)rxd->rx_hdr_status;
 	memcpy(first_hdr, hdr, RX_HTT_HDR_STATUS_LEN);
 
+	if (rx_hdr)
+		memcpy(rx_hdr, hdr, RX_HTT_HDR_STATUS_LEN);
+
 	/* Each A-MSDU subframe will use the original header as the base and be
 	 * reported as a separate MSDU so strip the A-MSDU bit from QoS Ctl.
 	 */
@@ -1461,6 +1468,17 @@
 	if (has_tkip_err)
 		status->flag |= RX_FLAG_MMIC_ERROR;
 
+	if (err) {
+		if (has_fcs_err)
+			*err = ATH10K_PKT_RX_ERR_FCS;
+		else if (has_tkip_err)
+			*err = ATH10K_PKT_RX_ERR_TKIP;
+		else if (has_crypto_err)
+			*err = ATH10K_PKT_RX_ERR_CRYPT;
+		else if (has_peer_idx_invalid)
+			*err = ATH10K_PKT_RX_ERR_PEER_IDX_INVAL;
+	}
+
 	/* Firmware reports all necessary management frames via WMI already.
 	 * They are not reported to monitor interfaces at all so pass the ones
 	 * coming via HTT to monitor interfaces instead. This simplifies
@@ -1531,11 +1549,13 @@
 	}
 }
 
-static int ath10k_unchain_msdu(struct sk_buff_head *amsdu)
+static int ath10k_unchain_msdu(struct sk_buff_head *amsdu,
+			       unsigned long int *unchain_cnt)
 {
 	struct sk_buff *skb, *first;
 	int space;
 	int total_len = 0;
+	int amsdu_len = skb_queue_len(amsdu);
 
 	/* TODO:  Might could optimize this by using
 	 * skb_try_coalesce or similar method to
@@ -1571,11 +1591,16 @@
 	}
 
 	__skb_queue_head(amsdu, first);
+
+	*unchain_cnt += amsdu_len - 1;
+
 	return 0;
 }
 
 static void ath10k_htt_rx_h_unchain(struct ath10k *ar,
-				    struct sk_buff_head *amsdu)
+				    struct sk_buff_head *amsdu,
+				    unsigned long int *drop_cnt,
+				    unsigned long int *unchain_cnt)
 {
 	struct sk_buff *first;
 	struct htt_rx_desc *rxd;
@@ -1593,11 +1618,12 @@
 	 */
 	if (decap != RX_MSDU_DECAP_RAW ||
 	    skb_queue_len(amsdu) != 1 + rxd->frag_info.ring2_more_count) {
+		*drop_cnt += skb_queue_len(amsdu);
 		__skb_queue_purge(amsdu);
 		return;
 	}
 
-	ath10k_unchain_msdu(amsdu);
+	ath10k_unchain_msdu(amsdu, unchain_cnt);
 }
 
 static bool ath10k_htt_rx_amsdu_allowed(struct ath10k *ar,
@@ -1623,7 +1649,8 @@
 
 static void ath10k_htt_rx_h_filter(struct ath10k *ar,
 				   struct sk_buff_head *amsdu,
-				   struct ieee80211_rx_status *rx_status)
+				   struct ieee80211_rx_status *rx_status,
+				   unsigned long int *drop_cnt)
 {
 	if (skb_queue_empty(amsdu))
 		return;
@@ -1631,15 +1658,70 @@
 	if (ath10k_htt_rx_amsdu_allowed(ar, amsdu, rx_status))
 		return;
 
+	if (drop_cnt)
+		*drop_cnt += skb_queue_len(amsdu);
+
 	__skb_queue_purge(amsdu);
 }
 
+static void ath10k_rx_vow_stats_ampdu_update(struct ath10k *ar,
+					     struct ath10k_sta *arsta,
+					     u32 mpdu_count)
+{
+	struct ath10k_vow_stats *stats = &arsta->vow_stats;
+
+	if (mpdu_count <= 10)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_10]++;
+	else if (mpdu_count <= 20)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_20]++;
+	else if (mpdu_count <= 30)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_30]++;
+	else if (mpdu_count <= 40)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_40]++;
+	else if (mpdu_count <= 50)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_50]++;
+	else if (mpdu_count <= 60)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_60]++;
+	else if (mpdu_count > 60)
+		stats->rx_pkt_ampdu[ATH10K_VOW_STATS_AMPDU_MORE]++;
+}
+
+static void ath10k_rx_vow_stats_amsdu_update(struct ath10k *ar,
+					     struct ath10k_sta *arsta,
+					     u32 msdu_count)
+{
+	struct ath10k_vow_stats *stats = &arsta->vow_stats;
+
+	if (msdu_count == 1)
+		stats->rx_pkt_amsdu[ATH10K_VOW_STATS_AMSDU_1]++;
+	else if (msdu_count == 2)
+		stats->rx_pkt_amsdu[ATH10K_VOW_STATS_AMSDU_2]++;
+	else if (msdu_count == 3)
+		stats->rx_pkt_amsdu[ATH10K_VOW_STATS_AMSDU_3]++;
+	else if (msdu_count == 4)
+		stats->rx_pkt_amsdu[ATH10K_VOW_STATS_AMSDU_4]++;
+	else if (msdu_count > 4)
+		stats->rx_pkt_amsdu[ATH10K_VOW_STATS_AMSDU_MORE]++;
+}
+
 static int ath10k_htt_rx_handle_amsdu(struct ath10k_htt *htt)
 {
 	struct ath10k *ar = htt->ar;
 	struct ieee80211_rx_status *rx_status = &htt->rx_status;
 	struct sk_buff_head amsdu;
 	int ret, num_msdus;
+	struct ieee80211_hdr *hdr;
+	struct ieee80211_sta *sta;
+	struct ath10k_sta *arsta;
+	unsigned long int drop_cnt = 0;
+	unsigned long int unchain_cnt = 0;
+	unsigned long int drop_cnt_filter = 0;
+	unsigned long int msdus_to_mac = 0;
+	enum ath10k_pkt_rx_err err = ATH10K_PKT_RX_ERR_MAX;
+	struct ath10k_vow_stats *stats;
+	u8 first_hdr[RX_HTT_HDR_STATUS_LEN];
+	u8 tid = 0;
+	bool is_vow = false;
 
 	__skb_queue_head_init(&amsdu);
 
@@ -1667,12 +1749,52 @@
 
 	/* only for ret = 1 indicates chained msdus */
 	if (ret > 0)
-		ath10k_htt_rx_h_unchain(ar, &amsdu);
+		ath10k_htt_rx_h_unchain(ar, &amsdu, &drop_cnt, &unchain_cnt);
+
+	ath10k_htt_rx_h_filter(ar, &amsdu, rx_status, &drop_cnt_filter);
+	ath10k_htt_rx_h_mpdu(ar, &amsdu, rx_status, true, first_hdr, &err);
+	msdus_to_mac = skb_queue_len(&amsdu);
+
+	hdr = (struct ieee80211_hdr *)first_hdr;
+	if (ieee80211_is_data_qos(hdr->frame_control))
+		tid = *ieee80211_get_qos_ctl(hdr) & IEEE80211_QOS_CTL_TID_MASK;
+
+	if (ath10k_vow_stats_update_required(ar, tid)) {
+		rx_status->flag |= RX_FLAG_IS_VOW;
+		is_vow = true;
+	} else {
+		rx_status->flag &= ~RX_FLAG_IS_VOW;
+	}
 
-	ath10k_htt_rx_h_filter(ar, &amsdu, rx_status);
-	ath10k_htt_rx_h_mpdu(ar, &amsdu, rx_status, true);
 	ath10k_htt_rx_h_deliver(ar, &amsdu, rx_status);
 
+	if (!is_vow)
+		goto exit;
+
+	rcu_read_lock();
+
+	sta = ieee80211_find_sta_by_ifaddr(ar->hw, hdr->addr2, NULL);
+	if (!sta) {
+		rcu_read_unlock();
+		goto exit;
+	}
+
+	arsta = (struct ath10k_sta *)sta->drv_priv;
+
+	spin_lock_bh(&ar->data_lock);
+	stats = &arsta->vow_stats;
+	stats->rx_pkt_from_fw += num_msdus;
+	stats->rx_pkt_unchained += unchain_cnt;
+	stats->rx_pkt_drop_chained += drop_cnt;
+	stats->rx_pkt_drop_filter += drop_cnt_filter;
+	if (err != ATH10K_PKT_RX_ERR_MAX)
+		stats->rx_pkt_err[err] += msdus_to_mac;
+	stats->rx_pkt_to_mac += msdus_to_mac;
+	ath10k_rx_vow_stats_amsdu_update(ar, arsta, num_msdus);
+	spin_unlock_bh(&ar->data_lock);
+
+	rcu_read_unlock();
+exit:
 	return num_msdus;
 }
 
@@ -1683,9 +1805,16 @@
 	struct htt_rx_indication_mpdu_range *mpdu_ranges;
 	int num_mpdu_ranges;
 	int i, mpdu_count = 0;
+	struct ath10k_sta *arsta;
+	struct ath10k_peer *peer;
+	u16 peer_id;
+	u8 tid;
 
 	num_mpdu_ranges = MS(__le32_to_cpu(rx->hdr.info1),
 			     HTT_RX_INDICATION_INFO1_NUM_MPDU_RANGES);
+	peer_id = __le16_to_cpu(rx->hdr.peer_id);
+	tid =  MS(rx->hdr.info0, HTT_RX_INDICATION_INFO0_EXT_TID);
+
 	mpdu_ranges = htt_rx_ind_get_mpdu_ranges(rx);
 
 	ath10k_dbg_dump(ar, ATH10K_DBG_HTT_DUMP, NULL, "htt rx ind: ",
@@ -1697,6 +1826,26 @@
 		mpdu_count += mpdu_ranges[i].mpdu_count;
 
 	atomic_add(mpdu_count, &htt->num_mpdus_ready);
+
+	if (!ath10k_vow_stats_update_required(ar, tid))
+		return;
+
+	rcu_read_lock();
+	spin_lock_bh(&ar->data_lock);
+
+	peer = ath10k_peer_find_by_id(ar, peer_id);
+	if (!peer)
+		goto out;
+
+	arsta = (struct ath10k_sta *)peer->sta->drv_priv;
+
+	for (i = 0; i < num_mpdu_ranges; i++)
+		ath10k_rx_vow_stats_ampdu_update(ar, arsta,
+						 mpdu_ranges[i].mpdu_count);
+
+out:
+	spin_unlock_bh(&ar->data_lock);
+	rcu_read_unlock();
 }
 
 static void ath10k_htt_rx_tx_compl_ind(struct ath10k *ar,
@@ -2009,8 +2158,9 @@
 			ath10k_pktlog_rx(ar, &amsdu);
 			budget_left -= skb_queue_len(&amsdu);
 			ath10k_htt_rx_h_ppdu(ar, &amsdu, status, vdev_id);
-			ath10k_htt_rx_h_filter(ar, &amsdu, status);
-			ath10k_htt_rx_h_mpdu(ar, &amsdu, status, false);
+			ath10k_htt_rx_h_filter(ar, &amsdu, status, NULL);
+			ath10k_htt_rx_h_mpdu(ar, &amsdu, status, false, NULL,
+					     NULL);
 			ath10k_htt_rx_h_deliver(ar, &amsdu, status);
 			break;
 		case -EAGAIN:
Index: backports-20171004-3.14.77-c09dbd7/include/net/mac80211.h
===================================================================
--- backports-20171004-3.14.77-c09dbd7.orig/include/net/mac80211.h	2017-12-21 22:18:27.425965310 +0530
+++ backports-20171004-3.14.77-c09dbd7/include/net/mac80211.h	2017-12-21 22:18:27.421965326 +0530
@@ -1144,6 +1144,7 @@
 	RX_FLAG_MIC_STRIPPED		= BIT(21),
 	RX_FLAG_ALLOW_SAME_PN		= BIT(22),
 	RX_FLAG_ICV_STRIPPED		= BIT(23),
+	RX_FLAG_IS_VOW			= BIT(24),
 };
 
 /**
