--- a/drivers/net/wireless/ath/ath10k/htt_tx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_tx.c
@@ -83,7 +83,7 @@ static void __ath10k_htt_tx_txq_recalc(s
 	ar->htt.tx_q_state.vaddr->map[tid][idx] &= ~bit;
 	ar->htt.tx_q_state.vaddr->map[tid][idx] |= count ? bit : 0;
 
-	ath10k_dbg(ar, ATH10K_DBG_MAC, "mac txq state update peer_id %hu tid %hhu count %hhu\n",
+	ath10k_dbg(ar, ATH10K_DBG_HTT, "mac txq state update peer_id %hu tid %hhu count %hhu\n",
 		   peer_id, tid, count);
 }
 
@@ -101,7 +101,7 @@ static void __ath10k_htt_tx_txq_sync(str
 	seq++;
 	ar->htt.tx_q_state.vaddr->seq = cpu_to_le32(seq);
 
-	ath10k_dbg(ar, ATH10K_DBG_MAC, "mac txq state update commit seq %u\n",
+	ath10k_dbg(ar, ATH10K_DBG_HTT, "mac txq state update commit seq %u\n",
 		   seq);
 
 	size = sizeof(*ar->htt.tx_q_state.vaddr);
