--- a/drivers/net/wireless/ath/ath10k/ce.c
+++ b/drivers/net/wireless/ath/ath10k/ce.c
@@ -1134,9 +1134,9 @@ void ath10k_ce_free_pipe(struct ath10k *
 void ath10k_ce_dump_registers(struct ath10k *ar)
 {
 	struct ath10k_pci *ar_pci = ath10k_pci_priv(ar);
-	struct ath10k_pci_pipe *pci_tx;
+	struct ath10k_pci_pipe *pci_pipe;
 	struct ath10k_ce_pipe *ce_state;
-	struct ath10k_ce_ring *src_ring;
+	struct ath10k_ce_ring *ce_ring;
 	struct ce_desc *base;
 	struct ce_desc *src_desc;
 	u32 ce_id, start, addr;
@@ -1158,20 +1158,34 @@ void ath10k_ce_dump_registers(struct ath
 			   ath10k_ce_dest_ring_read_index_get(ar, addr));
 	}
 
+	/* Dump CE5 indices when firmware crashes, this information might be
+	 * helpful in debugging firmware asserts because of CE5 full, CE5 is a
+	 * special case where CE buffers are reused using dma sync API unlike
+	 * other CE's.
+	 */
+	pci_pipe = &ar_pci->pipe_info[5];
+	ce_state = pci_pipe->ce_hdl;
+	ce_ring = ce_state->dest_ring;
+
+	write_index = ce_ring->write_index;
+	sw_index = ce_ring->sw_index;
+
+	ath10k_err(ar, "CE5 write_index %u sw_index %u\n", write_index, sw_index);
+
 	/* We are interested in Copy Engine 4 Host to target HTT */
-	pci_tx = &ar_pci->pipe_info[CE4_HOST_TARGET_HTT];
-	ce_state = pci_tx->ce_hdl;
-	src_ring = ce_state->src_ring;
-
-	write_index = src_ring->write_index;
-	sw_index = src_ring->sw_index;
-	hw_index = src_ring->hw_index;
-	nentries_mask = src_ring->nentries_mask;
+	pci_pipe = &ar_pci->pipe_info[CE4_HOST_TARGET_HTT];
+	ce_state = pci_pipe->ce_hdl;
+	ce_ring = ce_state->src_ring;
+
+	write_index = ce_ring->write_index;
+	sw_index = ce_ring->sw_index;
+	hw_index = ce_ring->hw_index;
+	nentries_mask = ce_ring->nentries_mask;
 
 	ath10k_err(ar, "write_index %u sw_index %u hw_index %u nentries_mask 0x%08x\n",
 		   write_index, sw_index, hw_index, nentries_mask);
 
-	base = src_ring->base_addr_owner_space;
+	base = ce_ring->base_addr_owner_space;
 	start = (write_index  - CE_DESC_DUMP_COUNT) & (nentries_mask);
 
 	ath10k_err(ar, "DMA addr: nbytes: meta data: byte swap: gather:\n");
