--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -1640,7 +1640,7 @@ static void ath10k_tpc_stats_print(struc
 
 	for (i = 0; i < tpc_stats->rate_max; i++) {
 		*len += scnprintf(buf + *len, buf_len - *len,
-				  "%8d %s 0x%2x %s\n", i,
+				  "\n%8d %s 0x%2x %s\n", i,
 				  pream_str[tpc_stats->tpc_table[j].pream_idx[i]],
 				  tpc_stats->tpc_table[j].rate_code[i],
 				  tpc_stats->tpc_table[j].tpc_value[i]);
--- a/drivers/net/wireless/ath/ath10k/wmi.c
+++ b/drivers/net/wireless/ath/ath10k/wmi.c
@@ -4836,12 +4836,12 @@ void ath10k_wmi_event_pdev_tpc_config(st
 	if (num_tx_chain > WMI_TPC_TX_N_CHAIN) {
 		ath10k_warn(ar, "number of tx chain is %d greater than TPC configured tx chain %d\n",
 			    num_tx_chain, WMI_TPC_TX_N_CHAIN);
-		return;
+		goto exit;
 	}
 
 	tpc_stats = kzalloc(sizeof(*tpc_stats), GFP_ATOMIC);
 	if (!tpc_stats)
-		return;
+		goto exit;
 
 	ath10k_wmi_tpc_config_get_rate_code(rate_code, pream_table,
 					    num_tx_chain);
@@ -4882,6 +4882,9 @@ void ath10k_wmi_event_pdev_tpc_config(st
 		   __le32_to_cpu(ev->twice_max_rd_power) / 2,
 		   __le32_to_cpu(ev->num_tx_chain),
 		   __le32_to_cpu(ev->rate_max));
+
+exit:
+	complete(&ar->debug.tpc_complete);
 }
 
 static u8
@@ -5086,10 +5089,17 @@ void ath10k_wmi_event_tpc_final_table(st
 
 	tpc_stats = kzalloc(sizeof(*tpc_stats), GFP_ATOMIC);
 	if (!tpc_stats)
-		return;
+		goto exit;
 
 	num_tx_chain = __le32_to_cpu(ev->num_tx_chain);
 
+	if (num_tx_chain > WMI_TPC_TX_N_CHAIN) {
+		ath10k_warn(ar, "number of tx chain is %d greater than TPC"
+			    " configured tx chain %d\n", num_tx_chain,
+			    WMI_TPC_TX_N_CHAIN);
+		goto exit;
+	}
+
 	ath10k_wmi_tpc_config_get_rate_code(rate_code, pream_table,
 					    num_tx_chain);
 
@@ -5129,6 +5139,9 @@ void ath10k_wmi_event_tpc_final_table(st
 		   __le32_to_cpu(ev->twice_max_rd_power) / 2,
 		   __le32_to_cpu(ev->num_tx_chain),
 		   __le32_to_cpu(ev->rate_max));
+
+exit:
+	complete(&ar->debug.tpc_complete);
 }
 
 static void
