--- a/drivers/net/wireless/ath/ath10k/debugfs_sta.c
+++ b/drivers/net/wireless/ath/ath10k/debugfs_sta.c
@@ -442,6 +442,224 @@ static const struct file_operations fops
 	.llseek = default_llseek,
 };
 
+static ssize_t ath10k_dbg_sta_dump_tx_stats(struct file *file,
+					       char __user *user_buf,
+					       size_t count, loff_t *ppos)
+{
+
+	struct ieee80211_sta *sta = file->private_data;
+	struct ath10k_sta *arsta = (struct ath10k_sta *)sta->drv_priv;
+	char *buf;
+	int len = 0, i, retval = 0, size = 2240;
+
+	buf = kzalloc(size, GFP_KERNEL);
+	if (buf == NULL)
+		return -ENOMEM;
+
+	len += scnprintf(buf + len, size - len, "LEGACY RATE PACKETS:\t");
+	len += scnprintf(buf + len, size - len, "1Mbps\t2Mbps\t5.5Mbps\t11Mbps"
+			"\t6Mbps\t9Mbps\t12Mbps\t18Mbps\t24Mbps\t36Mbps\t48Mbps"
+			"\t54Mbps\n");
+	len += scnprintf(buf + len, size - len, "Success packets:\t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_pkts_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry packets:  \t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_pkts_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed packets: \t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_pkts_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nTotal packets:  \t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_pkts_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU packets:  \t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_pkts_legacy_rates[i]);
+
+	len += scnprintf(buf + len, size - len, "\nLEGACY RATE BYTES:\n");
+	len += scnprintf(buf + len, size - len, "Success bytes:\t\t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_bytes_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry bytes:  \t\t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_bytes_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed bytes: \t\t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_bytes_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nTotal bytes:  \t\t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_bytes_legacy_rates[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU bytes:  \t\t");
+	for (i = 0; i < LEGACY_RATE_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_bytes_legacy_rates[i]);
+
+	len += scnprintf(buf + len, size - len, "\nMCS RATE PACKETS: \t");
+	len += scnprintf(buf + len, size - len, "MCS-0\tMCS-1\tMCS-2\tMCS-3"
+			"\tMCS-4\tMCS-5\tMCS-6\tMCS-7\tMCS-8\tMCS-9\n");
+	len += scnprintf(buf + len, size - len, "Success packets:\t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_pkts_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry packets:  \t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_pkts_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed packets: \t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_pkts_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nTotal packets:  \t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_pkts_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU packets:  \t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_pkts_mcs[i]);
+
+	len += scnprintf(buf + len, size - len, "\nMCS RATE BYTES: \n");
+	len += scnprintf(buf + len, size - len, "Success bytes:\t\t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_bytes_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry bytes:  \t\t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_bytes_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed bytes: \t\t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_bytes_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nTotal bytes:  \t\t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_bytes_mcs[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU bytes:  \t\t");
+	for (i = 0; i < VHT_MCS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_bytes_mcs[i]);
+
+	len += scnprintf(buf + len, size - len,
+			"\nBW PACKETS:  \t\t20Mhz \t40Mhz \t80Mhz \t160Mhz \n");
+	len += scnprintf(buf + len, size - len, "Success packets:\t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_pkts_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry packets:  \t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_pkts_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed packets: \t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_pkts_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nTotal packets:  \t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_pkts_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU packets:  \t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_pkts_bw[i]);
+
+	len += scnprintf(buf + len, size - len, "\nBW BYTES: \n");
+	len += scnprintf(buf + len, size - len, "Success bytes:\t\t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_bytes_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry bytes:  \t\t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_bytes_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed bytes: \t\t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_bytes_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nTotal bytes:  \t\t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_bytes_bw[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU bytes:  \t\t");
+	for (i = 0; i < VHT_BW_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_bytes_bw[i]);
+
+	len += scnprintf(buf + len, size - len,
+			"\nNSS PACKETS: \t\t1x1 \t2x2 \t3x3 \t 4x4\n");
+	len += scnprintf(buf + len, size - len, "Success packets:\t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_pkts_nss[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry packets:  \t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_pkts_nss[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed packets: \t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_pkts_nss[i]);
+	len += scnprintf(buf + len, size - len, "\ntotal packets:  \t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_pkts_nss[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU packets:  \t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_pkts_nss[i]);
+
+	len += scnprintf(buf + len, size - len, "\nNSS BYTES: \n");
+	len += scnprintf(buf + len, size - len, "Success bytes:\t\t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.succ_bytes_nss[i]);
+	len += scnprintf(buf + len, size - len, "\nRetry bytes:  \t\t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.retry_bytes_nss[i]);
+	len += scnprintf(buf + len, size - len, "\nFailed bytes: \t\t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.fail_bytes_nss[i]);
+	len += scnprintf(buf + len, size - len, "\ntotal bytes:  \t\t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.total_bytes_nss[i]);
+	len += scnprintf(buf + len, size - len, "\nAMPDU bytes:  \t\t");
+	for (i = 0; i < VHT_NSS_NUM; i++)
+		len += scnprintf(buf + len, size - len, "%llu\t",
+				 arsta->tx_stats.ampdu_bytes_nss[i]);
+
+	len += scnprintf(buf + len, size - len,
+			"\nTX duration:\t %llu usecs\n", arsta->tx_stats.tx_duration);
+
+	len += scnprintf(buf + len, size - len,
+			"BA fails:\t %llu\n", arsta->tx_stats.ba_fails);
+
+	if (len > size)
+		len = size;
+	retval = simple_read_from_buffer(user_buf, count, ppos, buf, len);
+	kfree(buf);
+
+	return retval;
+}
+
+static const struct file_operations fops_tx_stats = {
+	.read = ath10k_dbg_sta_dump_tx_stats,
+	.open = simple_open,
+	.owner = THIS_MODULE,
+	.llseek = default_llseek,
+};
+
 void ath10k_sta_add_debugfs(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 			    struct ieee80211_sta *sta, struct dentry *dir)
 {
@@ -452,4 +670,6 @@ void ath10k_sta_add_debugfs(struct ieee8
 	debugfs_create_file("delba", S_IWUSR, dir, sta, &fops_delba);
 	debugfs_create_file("rx_duration", S_IRUGO, dir, sta,
 			    &fops_rx_duration);
+	debugfs_create_file("tx_stats", S_IRUGO, dir, sta,
+			    &fops_tx_stats);
 }
