diff --git a/drivers/net/wireless/ath/ath10k/core.c b/drivers/net/wireless/ath/ath10k/core.c
index 1deafef..975e9f7 100644
--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -2767,7 +2767,8 @@ int ath10k_core_start(struct ath10k *ar, enum ath10k_firmware_mode mode,
 			     ar->wmi.svc_map))
 			val |= WMI_10_4_TDLS_UAPSD_BUFFER_STA;
 
-		if (test_bit(WMI_SERVICE_REPORT_AIRTIME, ar->wmi.svc_map))
+		if (test_bit(WMI_SERVICE_REPORT_AIRTIME, ar->wmi.svc_map) &&
+			(ar->enable_atf_sch))
 			val |= WMI_10_4_REPORT_AIRTIME;
 
 		status = ath10k_mac_ext_resource_config(ar, val);
diff --git a/drivers/net/wireless/ath/ath10k/htt_rx.c b/drivers/net/wireless/ath/ath10k/htt_rx.c
index edf2a7f..4b528e1 100644
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -2318,7 +2318,7 @@ static void ath10k_htt_rx_tx_fetch_ind(struct ath10k *ar, struct sk_buff *skb)
 	u8 tid;
 	int ret;
 	int i;
-	bool may_tx;
+	bool may_tx = true;
 
 	ath10k_dbg(ar, ATH10K_DBG_HTT, "htt rx tx fetch ind\n");
 
@@ -2391,8 +2391,11 @@ static void ath10k_htt_rx_tx_fetch_ind(struct ath10k *ar, struct sk_buff *skb)
 		num_msdus = 0;
 		num_bytes = 0;
 
-		ieee80211_txq_schedule_start(hw, txq->ac);
-		may_tx = ieee80211_txq_may_transmit(hw, txq);
+		if (ar->enable_atf_sch) {
+			ieee80211_txq_schedule_start(hw, txq->ac);
+			may_tx = ieee80211_txq_may_transmit(hw, txq);
+		}
+
 		while (num_msdus < max_num_msdus &&
 		       num_bytes < max_num_bytes) {
 			if (!may_tx)
@@ -2405,8 +2408,11 @@ static void ath10k_htt_rx_tx_fetch_ind(struct ath10k *ar, struct sk_buff *skb)
 			num_msdus++;
 			num_bytes += ret;
 		}
-		ieee80211_return_txq(hw, txq);
-		ieee80211_txq_schedule_end(hw, txq->ac);
+
+		if (ar->enable_atf_sch) {
+			ieee80211_return_txq(hw, txq);
+			ieee80211_txq_schedule_end(hw, txq->ac);
+		}
 
 		record->num_msdus = cpu_to_le16(num_msdus);
 		record->num_bytes = cpu_to_le32(num_bytes);
diff --git a/drivers/net/wireless/ath/ath10k/mac.c b/drivers/net/wireless/ath/ath10k/mac.c
index 55daec0..1636b29 100644
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -4089,7 +4089,7 @@ static u32 ath10k_mac_update_airtime(struct ath10k *ar,
 	u32 pktlen, overhead;
 	u32 airtime = 0;
 
-	if (!txq || !txq->sta)
+	if (!txq || !txq->sta || !ar->enable_atf_sch)
 		return airtime;
 
 	if (test_bit(WMI_SERVICE_REPORT_AIRTIME, ar->wmi.svc_map))
