--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -2266,7 +2266,6 @@ struct ath10k *ath10k_core_create(size_t
 
 	mutex_init(&ar->conf_mutex);
 	spin_lock_init(&ar->data_lock);
-	spin_lock_init(&ar->pending_txqs_lock);
 
 	INIT_LIST_HEAD(&ar->pending_txqs);
 	INIT_LIST_HEAD(&ar->peers);
--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -819,7 +819,6 @@ struct ath10k {
 
 	/* protects shared structure data */
 	spinlock_t data_lock;
-	spinlock_t pending_txqs_lock;
 
 	struct list_head pending_txqs;
 	struct list_head arvifs;
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -2055,7 +2055,6 @@ static void ath10k_htt_rx_tx_fetch_ind(s
 		return;
 	}
 
-	spin_lock_bh(&ar->pending_txqs_lock);
 	rcu_read_lock();
 
 	for (i = 0; i < num_records; i++) {
@@ -2108,7 +2107,6 @@ static void ath10k_htt_rx_tx_fetch_ind(s
 	}
 
 	rcu_read_unlock();
-	spin_unlock_bh(&ar->pending_txqs_lock);
 
 	resp_ids = ath10k_htt_get_tx_fetch_ind_resp_ids(&resp->tx_fetch_ind);
 	ath10k_htt_rx_tx_fetch_resp_id_confirm(ar, resp_ids, num_resp_ids);
--- a/drivers/net/wireless/ath/ath10k/mac.c
+++ b/drivers/net/wireless/ath/ath10k/mac.c
@@ -3630,23 +3630,15 @@ int ath10k_mac_txq_push_one(struct ieee8
 	int ret;
 	int skb_len;
 
-	lockdep_assert_held(&ar->pending_txqs_lock);
-
 	spin_lock_bh(&ar->htt.tx_lock);
 
 	if (!htt_requested && ath10k_mac_tx_push_threshold_reached(ar, artxq)) {
-		if (list_empty(&artxq->pending) && txq->qdepth > 0)
-			list_move_tail(&artxq->pending,
-				       &ar->pending_txqs);
 		spin_unlock_bh(&ar->htt.tx_lock);
 		return -EBUSY;
 	}
 
 	ret = ath10k_htt_tx_inc_pending(htt, is_mgmt, is_presp);
 	if (ret) {
-		if (list_empty(&artxq->pending) && txq->qdepth > 0)
-			list_move_tail(&artxq->pending,
-				       &ar->pending_txqs);
 		spin_unlock_bh(&ar->htt.tx_lock);
 		return -EBUSY;
 	}
@@ -3687,57 +3679,70 @@ int ath10k_mac_txq_push_one(struct ieee8
 	return skb_len;
 }
 
-static void ath10k_mac_txq_push(struct ieee80211_hw *hw,
-				struct ieee80211_txq *txq)
+static int ath10k_mac_txq_push(struct ieee80211_hw *hw,
+			       struct ieee80211_txq *txq)
 {
-	struct ath10k *ar = hw->priv;
 	bool htt_requested = false;
 	int ret;
 
-	lockdep_assert_held(&ar->pending_txqs_lock);
-
 	for (;;) {
 		ret = ath10k_mac_txq_push_one(hw, txq, htt_requested);
 		if (ret < 0)
-			break;
+			return ret;
 	}
 
-	ath10k_mac_txq_state_update(hw, txq);
+	return 0;
 }
 
 void ath10k_mac_txq_push_pending(struct ath10k *ar)
 {
 	struct ieee80211_hw *hw = ar->hw;
 	struct ieee80211_txq *txq;
-	struct ath10k_txq *artxq, *tmp;
-	struct list_head list;
-
-	INIT_LIST_HEAD(&list);
+	struct ath10k_txq *artxq;
+	int ret;
+	int n_iters;
 
 	if (!test_bit(ATH10K_FW_FEATURE_PEER_FLOW_CONTROL, ar->fw_features))
 		return;
 
 	rcu_read_lock();
-	spin_lock_bh(&ar->pending_txqs_lock);
+	spin_lock_bh(&ar->htt.tx_lock);
 
-	/* FIXME: This can probably be simplified/optimized */
+	if (ar->htt.tx_q_state.mode == HTT_TX_MODE_SWITCH_PUSH_PULL &&
+	    ar->htt.tx_q_state.threshold == 0)
+		goto skip;
 
-	list_for_each_entry_safe(artxq, tmp, &ar->pending_txqs, pending) {
+	/* The artxq->pending can be re-injected by other code flows so to
+	 * avoid vicious loop limit the number of iterations.
+	 */
+	n_iters = 0;
+	list_for_each_entry(artxq, &ar->pending_txqs, pending)
+		n_iters++;
+
+	while (!list_empty(&ar->pending_txqs) && n_iters-- > 0) {
+		artxq = list_first_entry(&ar->pending_txqs,
+					 struct ath10k_txq,
+					 pending);
 		txq = container_of((void *)artxq, struct ieee80211_txq,
 				   drv_priv);
-		list_move_tail(&artxq->pending, &list);
-	}
 
-	list_for_each_entry_safe(artxq, tmp, &list, pending) {
-		txq = container_of((void *)artxq, struct ieee80211_txq,
-				   drv_priv);
-		list_del_init(&artxq->pending);
-		ath10k_mac_txq_push(hw, txq);
+		spin_unlock_bh(&ar->htt.tx_lock);
+		ret = ath10k_mac_txq_push(hw, txq);
+		spin_lock_bh(&ar->htt.tx_lock);
+
+		__ath10k_mac_txq_state_update(hw, txq);
+
+		if (txq->qdepth > 0)
+			list_move_tail(&artxq->pending, &ar->pending_txqs);
+
+		if (ret == -EBUSY)
+			break;
 	}
 
-	ath10k_mac_txq_state_update_commit(ar);
+	__ath10k_mac_txq_state_update_commit(ar);
 
-	spin_unlock_bh(&ar->pending_txqs_lock);
+skip:
+	spin_unlock_bh(&ar->htt.tx_lock);
 	rcu_read_unlock();
 }
 
@@ -4160,13 +4165,17 @@ static void ath10k_mac_op_wake_tx_queue(
 					struct ieee80211_txq *txq)
 {
 	struct ath10k *ar = hw->priv;
-
-	spin_lock_bh(&ar->pending_txqs_lock);
+	struct ath10k_txq *artxq = (void *)txq->drv_priv;
 
 	ath10k_mac_txq_push(hw, txq);
-	ath10k_mac_txq_state_update_commit(ar);
 
-	spin_unlock_bh(&ar->pending_txqs_lock);
+	spin_lock_bh(&ar->htt.tx_lock);
+	if (txq->qdepth > 0 && list_empty(&artxq->pending))
+		list_add_tail(&artxq->pending, &ar->pending_txqs);
+
+	__ath10k_mac_txq_state_update(hw, txq);
+	__ath10k_mac_txq_state_update_commit(ar);
+	spin_unlock_bh(&ar->htt.tx_lock);
 }
 
 /* Must not be called with conf_mutex held as workers can use that also. */
@@ -5131,7 +5140,6 @@ static void ath10k_remove_interface(stru
 	ar->free_vdev_map |= 1LL << arvif->vdev_id;
 	list_del(&arvif->list);
 
-	spin_lock_bh(&ar->pending_txqs_lock);
 	spin_lock_bh(&ar->htt.tx_lock);
 	if (vif->txq) {
 		if (!list_empty(&artxq->pending))
@@ -5141,7 +5149,6 @@ static void ath10k_remove_interface(stru
 	}
 	ath10k_mac_vif_cleanup(ar, vif);
 	spin_unlock_bh(&ar->htt.tx_lock);
-	spin_unlock_bh(&ar->pending_txqs_lock);
 
 	if (arvif->vdev_type == WMI_VDEV_TYPE_AP ||
 	    arvif->vdev_type == WMI_VDEV_TYPE_IBSS) {
@@ -6049,7 +6056,6 @@ static int ath10k_sta_state(struct ieee8
 
 		ath10k_mac_dec_num_stations(arvif, sta);
 
-		spin_lock_bh(&ar->pending_txqs_lock);
 		spin_lock_bh(&ar->htt.tx_lock);
 
 		for (i = 0; i < ARRAY_SIZE(sta->txq); i++) {
@@ -6065,7 +6071,6 @@ static int ath10k_sta_state(struct ieee8
 		}
 
 		spin_unlock_bh(&ar->htt.tx_lock);
-		spin_unlock_bh(&ar->pending_txqs_lock);
 
 		spin_lock_bh(&ar->data_lock);
 		for (i = 0; i < ARRAY_SIZE(ar->peer_map); i++) {
