diff --git a/drivers/net/wireless/ath/ath10k/htt_rx.c b/drivers/net/wireless/ath/ath10k/htt_rx.c
index d7d1f09..0f6b26e 100644
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -2742,7 +2742,6 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	}
 
 	arsta->txrate.bw = ath10k_bw_to_mac80211_bw(txrate->bw);
-	arsta->last_tx_bitrate = cfg80211_calculate_bitrate(&arsta->txrate);
 	arsta->tx_info.status.rates[0].count = 1;
 
 	switch (txrate->flags) {
@@ -2766,6 +2765,7 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 			arsta->tx_info.status.rates[0].flags |=
 					IEEE80211_TX_RC_SHORT_GI;
 		arsta->tx_info.status.rates[0].flags |= IEEE80211_TX_RC_MCS;
+		arsta->txrate.flags = RATE_INFO_FLAGS_MCS;
 		break;
 	case WMI_RATE_PREAMBLE_VHT:
 		ieee80211_rate_set_vht(&arsta->tx_info.status.rates[0],
@@ -2774,6 +2774,7 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 			arsta->tx_info.status.rates[0].flags |=
 						IEEE80211_TX_RC_SHORT_GI;
 		arsta->tx_info.status.rates[0].flags |= IEEE80211_TX_RC_VHT_MCS;
+		arsta->txrate.flags = RATE_INFO_FLAGS_VHT_MCS;
 		break;
  	}
 
@@ -2796,11 +2797,12 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	if (peer_stats->succ_pkts)
 		arsta->tx_info.flags = IEEE80211_TX_STAT_ACK;
 
-	tx_bitrate = cfg80211_calculate_bitrate(&arsta->txrate);
+	arsta->last_tx_bitrate = cfg80211_calculate_bitrate(&arsta->txrate);
 
-	if (arsta->arvif->txrate_high && (tx_bitrate < arsta->arvif->txrate_low ||
-	    tx_bitrate > arsta->arvif->txrate_high))
-		ath10k_mac_cqm_tx_rate_notify(arsta, sta->addr, tx_bitrate);
+	if (arsta->arvif->txrate_high &&
+	    (arsta->last_tx_bitrate < arsta->arvif->txrate_low ||
+	     arsta->last_tx_bitrate > arsta->arvif->txrate_high))
+		ath10k_mac_cqm_tx_rate_notify(arsta, sta->addr, arsta->last_tx_bitrate);
 
 	if (ar->debug.enable_extd_tx_stats)
 		ath10k_accumulate_per_peer_tx_stats(ar, arsta, peer_stats);
