Index: backports-20190311-3.14.77-d492b001/drivers/net/wireless/ath/ath10k/mac.c
===================================================================
--- backports-20190311-3.14.77-d492b001.orig/drivers/net/wireless/ath/ath10k/mac.c	2020-01-30 11:04:41.835436163 +0530
+++ backports-20190311-3.14.77-d492b001/drivers/net/wireless/ath/ath10k/mac.c	2020-01-30 11:05:55.651436007 +0530
@@ -4608,17 +4608,19 @@
 	return 0;
 }
 
-static void ath10k_check_chain_mask(struct ath10k *ar, u32 cm, const char *dbg)
+static bool ath10k_check_chain_mask(struct ath10k *ar, u32 cm, const char *dbg)
 {
 	/* It is not clear that allowing gaps in chainmask
 	 * is helpful.  Probably it will not do what user
 	 * is hoping for, so warn in that case.
 	 */
-	if (cm == 15 || cm == 7 || cm == 3 || cm == 1 || cm == 0)
-		return;
-
-	ath10k_warn(ar, "mac %s antenna chainmask may be invalid: 0x%x.  Suggested values: 15, 7, 3, 1 or 0.\n",
-		    dbg, cm);
+	if (cm == 15 || cm == 7 || cm == 3 || cm == 1 || cm == 0) {
+		return true;
+	} else {
+		ath10k_warn(ar, "mac %s antenna chainmask is invalid: 0x%x.  Suggested values: 15, 7, 3, 1 or 0.\n",
+			    dbg, cm);
+		return false;
+	}
 }
 
 static int ath10k_mac_get_vht_cap_bf_sts(struct ath10k *ar)
@@ -4803,11 +4805,15 @@
 static int __ath10k_set_antenna(struct ath10k *ar, u32 tx_ant, u32 rx_ant)
 {
 	int ret;
+	bool is_valid_tx_chain_mask, is_valid_rx_chain_mask;
 
 	lockdep_assert_held(&ar->conf_mutex);
 
-	ath10k_check_chain_mask(ar, tx_ant, "tx");
-	ath10k_check_chain_mask(ar, rx_ant, "rx");
+	is_valid_tx_chain_mask = ath10k_check_chain_mask(ar, tx_ant, "tx");
+	is_valid_rx_chain_mask = ath10k_check_chain_mask(ar, rx_ant, "rx");
+
+	if (!is_valid_tx_chain_mask || !is_valid_rx_chain_mask)
+		return -EINVAL;
 
 	ar->cfg_tx_chainmask = tx_ant;
 	ar->cfg_rx_chainmask = rx_ant;
