--- a/drivers/net/wireless/ath/ath10k/core.h
+++ b/drivers/net/wireless/ath/ath10k/core.h
@@ -1075,6 +1075,7 @@ struct ath10k {
 	unsigned int filter_flags;
 	unsigned long dev_flags;
 	bool dfs_block_radar_events;
+	bool dfs_block_radar_confirmation;
 
 	/* protected by conf_mutex */
 	bool radar_enabled;
--- a/drivers/net/wireless/ath/ath10k/debug.c
+++ b/drivers/net/wireless/ath/ath10k/debug.c
@@ -2879,6 +2879,10 @@ int ath10k_debug_register(struct ath10k
 				    ar->debug.debugfs_phy,
 				    &ar->dfs_block_radar_events);
 
+		debugfs_create_bool("dfs_block_radar_confirmation", 0200,
+				    ar->debug.debugfs_phy,
+				    &ar->dfs_block_radar_confirmation);
+
 		debugfs_create_file("dfs_stats", 0400, ar->debug.debugfs_phy, ar,
 				    &fops_dfs_stats);
 	}
--- a/drivers/net/wireless/ath/ath10k/wmi.c
+++ b/drivers/net/wireless/ath/ath10k/wmi.c
@@ -4059,6 +4059,11 @@ static void ath10k_dfs_radar_report(stru
 
 	if ((test_bit(WMI_SERVICE_HOST_DFS_CHECK_SUPPORT, ar->wmi.svc_map)) &&
 	    ar->dfs_detector->region == NL80211_DFS_FCC) {
+		if (ar->dfs_block_radar_confirmation) {
+			ath10k_info(ar, "DFS Radar detected, but confirmation not sent to FW as requested\n");
+			return;
+		}
+
 		/* Consecutive radar indications need not be
 		 * sent to the firmware until we get confirmation
 		 * for the previous detected radar.
