diff --git a/drivers/net/wireless/ath/ath10k/core.c b/drivers/net/wireless/ath/ath10k/core.c
index 8cbe5a4..7694633 100644
--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -849,6 +849,9 @@ static int ath10k_core_get_board_id_from_otp(struct ath10k *ar)
 
 	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_FLASH_MTD)
 		board_id_param = BMI_PARAM_GET_FLASH_BOARD_ID;
+	else if (ar->cal_mode == ATH10K_PRE_CAL_MODE_FILE &&
+		ar->hw_params.id == QCA9888_HW_2_0_DEV_VERSION)
+		board_id_param = BMI_PARAM_GET_FLASH_BOARD_ID;
 	else
 		board_id_param = BMI_PARAM_GET_EEPROM_BOARD_ID;
 
