diff --git a/drivers/net/wireless/ath/ath10k/core.c b/drivers/net/wireless/ath/ath10k/core.c
index 7694633..a1061ad 100644
--- a/drivers/net/wireless/ath/ath10k/core.c
+++ b/drivers/net/wireless/ath/ath10k/core.c
@@ -917,6 +917,9 @@ static int ath10k_download_and_run_otp(struct ath10k *ar)
 	if (ar->cal_mode == ATH10K_PRE_CAL_MODE_FLASH_MTD)
 		bmi_otp_exe_param = BMI_PARAM_FLASH_SECTION_ALL;
 
+	if (QCA_REV_9888(ar) && ar->cal_mode == ATH10K_PRE_CAL_MODE_FILE)
+		bmi_otp_exe_param = BMI_PARAM_FLASH_SECTION_ALL;
+
 	ret = ath10k_bmi_execute(ar, address, bmi_otp_exe_param, &result);
 	if (ret) {
 		ath10k_err(ar, "could not execute otp (%d)\n", ret);
