From bfeeff271980b8d34ff505fcd401386580be661a Mon Sep 17 00:00:00 2001
From: Jiani Liu <quic_jianil@quicinc.com>
Date: Thu, 6 Apr 2023 17:16:14 +0800
Subject: [PATCH] ath9k: resume tx after NoA period

Change-Id: If69e7dfcaa3e00b2157d4d7882541abfde8be45a

Signed-off-by: Miaoqing Pan <miaoqing@codeaurora.org>
Signed-off-by: Jiani Liu <quic_jianil@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath9k/xmit.c b/drivers/net/wireless/ath/ath9k/xmit.c
index 2c97786..d6e5abc 100644
--- a/drivers/net/wireless/ath/ath9k/xmit.c
+++ b/drivers/net/wireless/ath/ath9k/xmit.c
@@ -133,6 +133,12 @@ void ath9k_wake_tx_queue(struct ieee80211_hw *hw, struct ieee80211_txq *queue)
 		tid->tidno);
 
 	ath_txq_lock(sc, txq);
+
+	if (tid->an->sleeping) {
+		ath_txq_unlock(sc, txq);
+		return;
+	}
+
 	ath_txq_schedule(sc, txq);
 	ath_txq_unlock(sc, txq);
 }
-- 
2.7.4

