From 1514ab663a32afb5b6968572a3ad1a36c4812771 Mon Sep 17 00:00:00 2001
From: Jiani Liu <quic_jianil@quicinc.com>
Date: Thu, 6 Apr 2023 16:27:56 +0800
Subject: [PATCH] ath9k: adjust btcoex duty cycle

Adjust btcoex duty cycle.

Signed-off-by: Miaoqing Pan <miaoqing@codeaurora.org>
Signed-off-by: Jiani Liu <quic_jianil@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath9k/btcoex.h b/drivers/net/wireless/ath/ath9k/btcoex.h
index 1bdfa84..2213d3f 100644
--- a/drivers/net/wireless/ath/ath9k/btcoex.h
+++ b/drivers/net/wireless/ath/ath9k/btcoex.h
@@ -28,7 +28,7 @@
 #define ATH_BTPRIORITY_GPIO_9300     8
 
 #define ATH_BTCOEX_DEF_BT_PERIOD  45
-#define ATH_BTCOEX_DEF_DUTY_CYCLE 55
+#define ATH_BTCOEX_DEF_DUTY_CYCLE 70
 #define ATH_BTCOEX_BTSCAN_DUTY_CYCLE 90
 #define ATH_BTCOEX_BMISS_THRESH   50
 
-- 
2.7.4

