From aca3e5d115b51af4783c284f2373ff99c4f2cf12 Mon Sep 17 00:00:00 2001
From: Jiani Liu <quic_jianil@quicinc.com>
Date: Thu, 6 Apr 2023 16:08:06 +0800
Subject: [PATCH] ath9k: Extra platform leds

Add extra platform leds.

Signed-off-by: Jiani Liu <quic_jianil@quicinc.com>

diff --git a/drivers/net/wireless/ath/ath9k/gpio.c b/drivers/net/wireless/ath/ath9k/gpio.c
index aefed89..36966e6 100644
--- a/drivers/net/wireless/ath/ath9k/gpio.c
+++ b/drivers/net/wireless/ath/ath9k/gpio.c
@@ -129,10 +129,30 @@ void ath_deinit_leds(struct ath_softc *sc)
 	}
 }
 
+static int ath_create_platform_led(struct ath_softc *sc,
+					const struct gpio_led *gpio)
+{
+	struct ath_led *led;
+	int ret;
+
+	led = kzalloc(sizeof(*led), GFP_KERNEL);
+	if (!led)
+		return -ENOMEM;
+
+	led->gpio = gpio;
+	ret = ath_add_led(sc, led);
+	if (ret < 0)
+		kfree(led);
+
+	return ret;
+}
+
 void ath_init_leds(struct ath_softc *sc)
 {
+	struct ath9k_platform_data *pdata = sc->dev->platform_data;
 	char led_name[32];
 	const char *trigger;
+	int i;
 
 	INIT_LIST_HEAD(&sc->leds);
 
@@ -150,6 +170,12 @@ void ath_init_leds(struct ath_softc *sc)
 		trigger = ieee80211_get_radio_led_name(sc->hw);
 
 	ath_create_gpio_led(sc, sc->sc_ah->led_pin, led_name, trigger, 1);
+
+	if (!pdata)
+		return;
+
+	for (i = 0; i < pdata->num_leds; i++)
+		ath_create_platform_led(sc, &pdata->leds[i]);
 }
 #endif
 
-- 
2.7.4

