diff --git a/drivers/net/wireless/ath/ath10k/debug.h b/drivers/net/wireless/ath/ath10k/debug.h
index 30b5654..e9ff1ed 100644
--- a/drivers/net/wireless/ath/ath10k/debug.h
+++ b/drivers/net/wireless/ath/ath10k/debug.h
@@ -35,6 +35,7 @@ enum ath10k_debug_mask {
 	ATH10K_DBG_SNOC		= 0x00100000,
 	ATH10K_DBG_QMI		= 0x00200000,
 	ATH10K_DBG_STA		= 0x00400000,
+	ATH10K_DBG_FWLOG	= 0x00800000,
 	ATH10K_DBG_SMART_ANT    = 0x01000000,
 	ATH10K_DBG_ANY		= 0xffffffff,
 };
diff --git a/drivers/net/wireless/ath/ath10k/fwlog.c b/drivers/net/wireless/ath/ath10k/fwlog.c
index 2da91aa..c4b7584 100644
--- a/drivers/net/wireless/ath/ath10k/fwlog.c
+++ b/drivers/net/wireless/ath/ath10k/fwlog.c
@@ -1461,8 +1461,8 @@ static void ath10k_fwlog_parse_msg(struct ath10k *ar, u8 *data, int len) {
 	len -= sizeof(dropped);
 
 	if (dropped > 0)
-		printk(FW_DBGLOG_PRINT_PREFIX "%d log buffers are dropped \n",
-		       dropped);
+		ath10k_dbg(ar, ATH10K_DBG_FWLOG, FW_DBGLOG_PRINT_PREFIX
+			   "%d log buffers are dropped \n\n", dropped);
 
 	count = 0;
 	buffer = (u32 *)data;
