diff --git a/drivers/net/wireless/ath/ath10k/htt_rx.c b/drivers/net/wireless/ath/ath10k/htt_rx.c
index 90d320e..dc2e60c 100644
--- a/drivers/net/wireless/ath/ath10k/htt_rx.c
+++ b/drivers/net/wireless/ath/ath10k/htt_rx.c
@@ -3488,7 +3488,7 @@ ath10k_update_per_peer_tx_stats(struct ath10k *ar,
 	u8 rate = 0, sgi;
 	s8 rate_idx = 0;
 	bool skip_auto_rate;
-	struct rate_info txrate;
+	struct rate_info txrate = {};
 
 	lockdep_assert_held(&ar->data_lock);
 
