From b840df3c2805c9d2a871f2b213a369b0726e08a7 Mon Sep 17 00:00:00 2001
From: Jiani Liu <quic_jianil@quicinc.com>
Date: Fri, 7 Apr 2023 15:00:05 +0800
Subject: [PATCH] mac80211: fix wext get tx rate failed

Change-Id: I0c3a520add5694ea6c87c5840ea65ff2de5b4f2e
Signed-off-by: Miaoqing Pan <miaoqing@codeaurora.org>
Signed-off-by: Jiani Liu <quic_jianil@quicinc.com>

diff --git a/net/wireless/wext-compat.c b/net/wireless/wext-compat.c
index a8320dc..9c1c905 100644
--- a/net/wireless/wext-compat.c
+++ b/net/wireless/wext-compat.c
@@ -1302,6 +1302,7 @@ static int cfg80211_wext_giwrate(struct net_device *dev,
 	if (err)
 		return err;
 
+	sinfo.filled &= ~BIT(NL80211_STA_INFO_TX_BITRATE);
 	wiphy_lock(&rdev->wiphy);
 	err = rdev_get_station(rdev, dev, addr, &sinfo);
 	wiphy_unlock(&rdev->wiphy);
-- 
2.7.4

